// @strudel/webaudio/dist/index.mjs
import * as k from "@strudel/core";
import { Pattern as y, clamp as q } from "@strudel/core";
import { setLogger as C, doughTrigger as O, superdough as b, getAudioContext as R, analysers as v, getAnalyzerData as S } from "superdough";
export * from "superdough";
import { getTheme as W, getDrawContext as T } from "@strudel/draw";
const { Pattern: z, logger: B, repl: D } = k;
C(B);
const P = (e) => (e.ensureObjectValue(), e.value), I = (e, t, o, a) => b(P(t), e - o, t.duration / a, a), M = (e, t, o, a, s) => b(P(e), s ? `=${s}` : t, o);
z.prototype.webaudio = function() {
  return this.onTrigger(I);
};
function G(e = {}) {
  return e = {
    getTime: () => R().currentTime,
    defaultOutput: M,
    ...e
  }, D(e);
}
z.prototype.dough = function() {
  return this.onTrigger(O, 1);
};
function j(e, {
  align: t = !0,
  color: o = "white",
  thickness: a = 3,
  scale: s = 0.25,
  pos: f = 0.75,
  trigger: g = 0,
  ctx: r = T(),
  id: w = 1
} = {}) {
  r.lineWidth = a, r.strokeStyle = o;
  let h = r.canvas;
  if (!e) {
    r.beginPath();
    let i = f * h.height;
    r.moveTo(0, i), r.lineTo(h.width, i), r.stroke();
    return;
  }
  const u = S("time", w);
  r.beginPath();
  const p = e.frequencyBinCount;
  let d = t ? Array.from(u).findIndex((i, l, c) => l && c[l - 1] > -g && i <= -g) : 0;
  d = Math.max(d, 0);
  const m = h.width * 1 / p;
  let n = 0;
  for (let i = d; i < p; i++) {
    const l = u[i] + 1, c = (f - s * (l - 1)) * h.height;
    i === 0 ? r.moveTo(n, c) : r.lineTo(n, c), n += m;
  }
  r.stroke();
}
function F(e, { color: t = "white", scale: o = 0.25, pos: a = 0.75, lean: s = 0.5, min: f = -150, max: g = 0, ctx: r = T(), id: w = 1 } = {}) {
  if (!e) {
    r.beginPath();
    let n = a * u.height;
    r.moveTo(0, n), r.lineTo(u.width, n), r.stroke();
    return;
  }
  const h = S("frequency", w), u = r.canvas;
  r.fillStyle = t;
  const p = e.frequencyBinCount, d = u.width * 1 / p;
  let m = 0;
  for (let n = 0; n < p; n++) {
    const l = q((h[n] - f) / (g - f), 0, 1) * o, c = l * u.height, $ = (a - l * s) * u.height;
    r.fillRect(m, $, Math.max(d, 1), c), m += d;
  }
}
function A(e = 0, t = "0,0,0", o = T()) {
  e ? (o.fillStyle = `rgba(${t},${1 - e})`, o.fillRect(0, 0, o.canvas.width, o.canvas.height)) : o.clearRect(0, 0, o.canvas.width, o.canvas.height);
}
y.prototype.fscope = function(e = {}) {
  let t = e.id ?? 1;
  return this.analyze(t).draw(
    () => {
      A(e.smear, "0,0,0", e.ctx), v[t] && F(v[t], e);
    },
    { id: t }
  );
};
y.prototype.tscope = function(e = {}) {
  let t = e.id ?? 1;
  return this.analyze(t).draw(
    (o) => {
      e.color = o[0]?.value?.color || W().foreground, e.color, A(e.smear, "0,0,0", e.ctx), j(v[t], e);
    },
    { id: t }
  );
};
y.prototype.scope = y.prototype.tscope;
export {
  F as drawFrequencyScope,
  j as drawTimeScope,
  M as webaudioOutput,
  I as webaudioOutputTrigger,
  G as webaudioRepl
};
