// @strudel/mini/dist/index.mjs
import * as w from "@strudel/core";
function sr(r, n) {
  function e() {
    this.constructor = r;
  }
  e.prototype = n.prototype, r.prototype = new e();
}
function Y(r, n, e, E) {
  var C = Error.call(this, r);
  return Object.setPrototypeOf && Object.setPrototypeOf(C, Y.prototype), C.expected = n, C.found = e, C.location = E, C.name = "SyntaxError", C;
}
sr(Y, Error);
function fu(r, n, e) {
  return e = e || " ", r.length > n ? r : (n -= r.length, e += e.repeat(n), r + e.slice(0, n));
}
Y.prototype.format = function(r) {
  var n = "Error: " + this.message;
  if (this.location) {
    var e = null, E;
    for (E = 0; E < r.length; E++)
      if (r[E].source === this.location.source) {
        e = r[E].text.split(/\r\n|\n|\r/g);
        break;
      }
    var C = this.location.start, c = this.location.source && typeof this.location.source.offset == "function" ? this.location.source.offset(C) : C, g = this.location.source + ":" + c.line + ":" + c.column;
    if (e) {
      var $ = this.location.end, v = fu("", c.line.toString().length, " "), D = e[C.line - 1], a = C.line === $.line ? $.column : D.length + 1, F = a - C.column || 1;
      n += `
 --> ` + g + `
` + v + ` |
` + c.line + " | " + D + `
` + v + " | " + fu("", C.column - 1, " ") + fu("", F, "^");
    } else
      n += `
 at ` + g;
  }
  return n;
};
Y.buildMessage = function(r, n) {
  var e = {
    literal: function(D) {
      return '"' + C(D.text) + '"';
    },
    class: function(D) {
      var a = D.parts.map(function(F) {
        return Array.isArray(F) ? c(F[0]) + "-" + c(F[1]) : c(F);
      });
      return "[" + (D.inverted ? "^" : "") + a.join("") + "]";
    },
    any: function() {
      return "any character";
    },
    end: function() {
      return "end of input";
    },
    other: function(D) {
      return D.description;
    }
  };
  function E(D) {
    return D.charCodeAt(0).toString(16).toUpperCase();
  }
  function C(D) {
    return D.replace(/\\/g, "\\\\").replace(/"/g, '\\"').replace(/\0/g, "\\0").replace(/\t/g, "\\t").replace(/\n/g, "\\n").replace(/\r/g, "\\r").replace(/[\x00-\x0F]/g, function(a) {
      return "\\x0" + E(a);
    }).replace(/[\x10-\x1F\x7F-\x9F]/g, function(a) {
      return "\\x" + E(a);
    });
  }
  function c(D) {
    return D.replace(/\\/g, "\\\\").replace(/\]/g, "\\]").replace(/\^/g, "\\^").replace(/-/g, "\\-").replace(/\0/g, "\\0").replace(/\t/g, "\\t").replace(/\n/g, "\\n").replace(/\r/g, "\\r").replace(/[\x00-\x0F]/g, function(a) {
      return "\\x0" + E(a);
    }).replace(/[\x10-\x1F\x7F-\x9F]/g, function(a) {
      return "\\x" + E(a);
    });
  }
  function g(D) {
    return e[D.type](D);
  }
  function $(D) {
    var a = D.map(g), F, y;
    if (a.sort(), a.length > 0) {
      for (F = 1, y = 1; F < a.length; F++)
        a[F - 1] !== a[F] && (a[y] = a[F], y++);
      a.length = y;
    }
    switch (a.length) {
      case 1:
        return a[0];
      case 2:
        return a[0] + " or " + a[1];
      default:
        return a.slice(0, -1).join(", ") + ", or " + a[a.length - 1];
    }
  }
  function v(D) {
    return D ? '"' + C(D) + '"' : "end of input";
  }
  return "Expected " + $(r) + " but " + v(n) + " found.";
};
function nr(r, n) {
  n = n !== void 0 ? n : {};
  var e = {}, E = n.grammarSource, C = { start: Wu }, c = Wu, g = ".", $ = "-", v = "+", D = "0", a = ",", F = "|", y = '"', x = "'", q = "#", j = "^", Bu = "_", Du = "[", lu = "]", A0 = "{", t0 = "}", s0 = "%", n0 = "<", i0 = ">", E0 = "@", a0 = "!", C0 = "(", F0 = ")", f0 = "/", o0 = "*", c0 = "?", B0 = ":", pu = "..", gu = "struct", hu = "target", vu = "euclid", du = "slow", $u = "rotL", mu = "rotR", _u = "fast", wu = "scale", yu = "//", bu = "cat", D0 = "$", xu = "setcps", Nu = "setbpm", Mu = "hush", l0 = /^[1-9]/, p0 = /^[eE]/, g0 = /^[0-9]/, ku = /^[ \n\r\t\xA0]/, h0 = /^[0-9~]/, Lu = /^[^\n]/, v0 = /^[a-z\xB5\xDF-\xF6\xF8-\xFF\u0101\u0103\u0105\u0107\u0109\u010B\u010D\u010F\u0111\u0113\u0115\u0117\u0119\u011B\u011D\u011F\u0121\u0123\u0125\u0127\u0129\u012B\u012D\u012F\u0131\u0133\u0135\u0137-\u0138\u013A\u013C\u013E\u0140\u0142\u0144\u0146\u0148-\u0149\u014B\u014D\u014F\u0151\u0153\u0155\u0157\u0159\u015B\u015D\u015F\u0161\u0163\u0165\u0167\u0169\u016B\u016D\u016F\u0171\u0173\u0175\u0177\u017A\u017C\u017E-\u0180\u0183\u0185\u0188\u018C-\u018D\u0192\u0195\u0199-\u019B\u019E\u01A1\u01A3\u01A5\u01A8\u01AA-\u01AB\u01AD\u01B0\u01B4\u01B6\u01B9-\u01BA\u01BD-\u01BF\u01C6\u01C9\u01CC\u01CE\u01D0\u01D2\u01D4\u01D6\u01D8\u01DA\u01DC-\u01DD\u01DF\u01E1\u01E3\u01E5\u01E7\u01E9\u01EB\u01ED\u01EF-\u01F0\u01F3\u01F5\u01F9\u01FB\u01FD\u01FF\u0201\u0203\u0205\u0207\u0209\u020B\u020D\u020F\u0211\u0213\u0215\u0217\u0219\u021B\u021D\u021F\u0221\u0223\u0225\u0227\u0229\u022B\u022D\u022F\u0231\u0233-\u0239\u023C\u023F-\u0240\u0242\u0247\u0249\u024B\u024D\u024F-\u0293\u0295-\u02AF\u0371\u0373\u0377\u037B-\u037D\u0390\u03AC-\u03CE\u03D0-\u03D1\u03D5-\u03D7\u03D9\u03DB\u03DD\u03DF\u03E1\u03E3\u03E5\u03E7\u03E9\u03EB\u03ED\u03EF-\u03F3\u03F5\u03F8\u03FB-\u03FC\u0430-\u045F\u0461\u0463\u0465\u0467\u0469\u046B\u046D\u046F\u0471\u0473\u0475\u0477\u0479\u047B\u047D\u047F\u0481\u048B\u048D\u048F\u0491\u0493\u0495\u0497\u0499\u049B\u049D\u049F\u04A1\u04A3\u04A5\u04A7\u04A9\u04AB\u04AD\u04AF\u04B1\u04B3\u04B5\u04B7\u04B9\u04BB\u04BD\u04BF\u04C2\u04C4\u04C6\u04C8\u04CA\u04CC\u04CE-\u04CF\u04D1\u04D3\u04D5\u04D7\u04D9\u04DB\u04DD\u04DF\u04E1\u04E3\u04E5\u04E7\u04E9\u04EB\u04ED\u04EF\u04F1\u04F3\u04F5\u04F7\u04F9\u04FB\u04FD\u04FF\u0501\u0503\u0505\u0507\u0509\u050B\u050D\u050F\u0511\u0513\u0515\u0517\u0519\u051B\u051D\u051F\u0521\u0523\u0525\u0527\u0529\u052B\u052D\u052F\u0560-\u0588\u10D0-\u10FA\u10FD-\u10FF\u13F8-\u13FD\u1C80-\u1C88\u1D00-\u1D2B\u1D6B-\u1D77\u1D79-\u1D9A\u1E01\u1E03\u1E05\u1E07\u1E09\u1E0B\u1E0D\u1E0F\u1E11\u1E13\u1E15\u1E17\u1E19\u1E1B\u1E1D\u1E1F\u1E21\u1E23\u1E25\u1E27\u1E29\u1E2B\u1E2D\u1E2F\u1E31\u1E33\u1E35\u1E37\u1E39\u1E3B\u1E3D\u1E3F\u1E41\u1E43\u1E45\u1E47\u1E49\u1E4B\u1E4D\u1E4F\u1E51\u1E53\u1E55\u1E57\u1E59\u1E5B\u1E5D\u1E5F\u1E61\u1E63\u1E65\u1E67\u1E69\u1E6B\u1E6D\u1E6F\u1E71\u1E73\u1E75\u1E77\u1E79\u1E7B\u1E7D\u1E7F\u1E81\u1E83\u1E85\u1E87\u1E89\u1E8B\u1E8D\u1E8F\u1E91\u1E93\u1E95-\u1E9D\u1E9F\u1EA1\u1EA3\u1EA5\u1EA7\u1EA9\u1EAB\u1EAD\u1EAF\u1EB1\u1EB3\u1EB5\u1EB7\u1EB9\u1EBB\u1EBD\u1EBF\u1EC1\u1EC3\u1EC5\u1EC7\u1EC9\u1ECB\u1ECD\u1ECF\u1ED1\u1ED3\u1ED5\u1ED7\u1ED9\u1EDB\u1EDD\u1EDF\u1EE1\u1EE3\u1EE5\u1EE7\u1EE9\u1EEB\u1EED\u1EEF\u1EF1\u1EF3\u1EF5\u1EF7\u1EF9\u1EFB\u1EFD\u1EFF-\u1F07\u1F10-\u1F15\u1F20-\u1F27\u1F30-\u1F37\u1F40-\u1F45\u1F50-\u1F57\u1F60-\u1F67\u1F70-\u1F7D\u1F80-\u1F87\u1F90-\u1F97\u1FA0-\u1FA7\u1FB0-\u1FB4\u1FB6-\u1FB7\u1FBE\u1FC2-\u1FC4\u1FC6-\u1FC7\u1FD0-\u1FD3\u1FD6-\u1FD7\u1FE0-\u1FE7\u1FF2-\u1FF4\u1FF6-\u1FF7\u210A\u210E-\u210F\u2113\u212F\u2134\u2139\u213C-\u213D\u2146-\u2149\u214E\u2184\u2C30-\u2C5E\u2C61\u2C65-\u2C66\u2C68\u2C6A\u2C6C\u2C71\u2C73-\u2C74\u2C76-\u2C7B\u2C81\u2C83\u2C85\u2C87\u2C89\u2C8B\u2C8D\u2C8F\u2C91\u2C93\u2C95\u2C97\u2C99\u2C9B\u2C9D\u2C9F\u2CA1\u2CA3\u2CA5\u2CA7\u2CA9\u2CAB\u2CAD\u2CAF\u2CB1\u2CB3\u2CB5\u2CB7\u2CB9\u2CBB\u2CBD\u2CBF\u2CC1\u2CC3\u2CC5\u2CC7\u2CC9\u2CCB\u2CCD\u2CCF\u2CD1\u2CD3\u2CD5\u2CD7\u2CD9\u2CDB\u2CDD\u2CDF\u2CE1\u2CE3-\u2CE4\u2CEC\u2CEE\u2CF3\u2D00-\u2D25\u2D27\u2D2D\uA641\uA643\uA645\uA647\uA649\uA64B\uA64D\uA64F\uA651\uA653\uA655\uA657\uA659\uA65B\uA65D\uA65F\uA661\uA663\uA665\uA667\uA669\uA66B\uA66D\uA681\uA683\uA685\uA687\uA689\uA68B\uA68D\uA68F\uA691\uA693\uA695\uA697\uA699\uA69B\uA723\uA725\uA727\uA729\uA72B\uA72D\uA72F-\uA731\uA733\uA735\uA737\uA739\uA73B\uA73D\uA73F\uA741\uA743\uA745\uA747\uA749\uA74B\uA74D\uA74F\uA751\uA753\uA755\uA757\uA759\uA75B\uA75D\uA75F\uA761\uA763\uA765\uA767\uA769\uA76B\uA76D\uA76F\uA771-\uA778\uA77A\uA77C\uA77F\uA781\uA783\uA785\uA787\uA78C\uA78E\uA791\uA793-\uA795\uA797\uA799\uA79B\uA79D\uA79F\uA7A1\uA7A3\uA7A5\uA7A7\uA7A9\uA7AF\uA7B5\uA7B7\uA7B9\uA7FA\uAB30-\uAB5A\uAB60-\uAB65\uAB70-\uABBF\uFB00-\uFB06\uFB13-\uFB17\uFF41-\uFF5A]/, d0 = /^[\u02B0-\u02C1\u02C6-\u02D1\u02E0-\u02E4\u02EC\u02EE\u0374\u037A\u0559\u0640\u06E5-\u06E6\u07F4-\u07F5\u07FA\u081A\u0824\u0828\u0971\u0E46\u0EC6\u10FC\u17D7\u1843\u1AA7\u1C78-\u1C7D\u1D2C-\u1D6A\u1D78\u1D9B-\u1DBF\u2071\u207F\u2090-\u209C\u2C7C-\u2C7D\u2D6F\u2E2F\u3005\u3031-\u3035\u303B\u309D-\u309E\u30FC-\u30FE\uA015\uA4F8-\uA4FD\uA60C\uA67F\uA69C-\uA69D\uA717-\uA71F\uA770\uA788\uA7F8-\uA7F9\uA9CF\uA9E6\uAA70\uAADD\uAAF3-\uAAF4\uAB5C-\uAB5F\uFF70\uFF9E-\uFF9F]/, $0 = /^[\xAA\xBA\u01BB\u01C0-\u01C3\u0294\u05D0-\u05EA\u05EF-\u05F2\u0620-\u063F\u0641-\u064A\u066E-\u066F\u0671-\u06D3\u06D5\u06EE-\u06EF\u06FA-\u06FC\u06FF\u0710\u0712-\u072F\u074D-\u07A5\u07B1\u07CA-\u07EA\u0800-\u0815\u0840-\u0858\u0860-\u086A\u08A0-\u08B4\u08B6-\u08BD\u0904-\u0939\u093D\u0950\u0958-\u0961\u0972-\u0980\u0985-\u098C\u098F-\u0990\u0993-\u09A8\u09AA-\u09B0\u09B2\u09B6-\u09B9\u09BD\u09CE\u09DC-\u09DD\u09DF-\u09E1\u09F0-\u09F1\u09FC\u0A05-\u0A0A\u0A0F-\u0A10\u0A13-\u0A28\u0A2A-\u0A30\u0A32-\u0A33\u0A35-\u0A36\u0A38-\u0A39\u0A59-\u0A5C\u0A5E\u0A72-\u0A74\u0A85-\u0A8D\u0A8F-\u0A91\u0A93-\u0AA8\u0AAA-\u0AB0\u0AB2-\u0AB3\u0AB5-\u0AB9\u0ABD\u0AD0\u0AE0-\u0AE1\u0AF9\u0B05-\u0B0C\u0B0F-\u0B10\u0B13-\u0B28\u0B2A-\u0B30\u0B32-\u0B33\u0B35-\u0B39\u0B3D\u0B5C-\u0B5D\u0B5F-\u0B61\u0B71\u0B83\u0B85-\u0B8A\u0B8E-\u0B90\u0B92-\u0B95\u0B99-\u0B9A\u0B9C\u0B9E-\u0B9F\u0BA3-\u0BA4\u0BA8-\u0BAA\u0BAE-\u0BB9\u0BD0\u0C05-\u0C0C\u0C0E-\u0C10\u0C12-\u0C28\u0C2A-\u0C39\u0C3D\u0C58-\u0C5A\u0C60-\u0C61\u0C80\u0C85-\u0C8C\u0C8E-\u0C90\u0C92-\u0CA8\u0CAA-\u0CB3\u0CB5-\u0CB9\u0CBD\u0CDE\u0CE0-\u0CE1\u0CF1-\u0CF2\u0D05-\u0D0C\u0D0E-\u0D10\u0D12-\u0D3A\u0D3D\u0D4E\u0D54-\u0D56\u0D5F-\u0D61\u0D7A-\u0D7F\u0D85-\u0D96\u0D9A-\u0DB1\u0DB3-\u0DBB\u0DBD\u0DC0-\u0DC6\u0E01-\u0E30\u0E32-\u0E33\u0E40-\u0E45\u0E81-\u0E82\u0E84\u0E87-\u0E88\u0E8A\u0E8D\u0E94-\u0E97\u0E99-\u0E9F\u0EA1-\u0EA3\u0EA5\u0EA7\u0EAA-\u0EAB\u0EAD-\u0EB0\u0EB2-\u0EB3\u0EBD\u0EC0-\u0EC4\u0EDC-\u0EDF\u0F00\u0F40-\u0F47\u0F49-\u0F6C\u0F88-\u0F8C\u1000-\u102A\u103F\u1050-\u1055\u105A-\u105D\u1061\u1065-\u1066\u106E-\u1070\u1075-\u1081\u108E\u1100-\u1248\u124A-\u124D\u1250-\u1256\u1258\u125A-\u125D\u1260-\u1288\u128A-\u128D\u1290-\u12B0\u12B2-\u12B5\u12B8-\u12BE\u12C0\u12C2-\u12C5\u12C8-\u12D6\u12D8-\u1310\u1312-\u1315\u1318-\u135A\u1380-\u138F\u1401-\u166C\u166F-\u167F\u1681-\u169A\u16A0-\u16EA\u16F1-\u16F8\u1700-\u170C\u170E-\u1711\u1720-\u1731\u1740-\u1751\u1760-\u176C\u176E-\u1770\u1780-\u17B3\u17DC\u1820-\u1842\u1844-\u1878\u1880-\u1884\u1887-\u18A8\u18AA\u18B0-\u18F5\u1900-\u191E\u1950-\u196D\u1970-\u1974\u1980-\u19AB\u19B0-\u19C9\u1A00-\u1A16\u1A20-\u1A54\u1B05-\u1B33\u1B45-\u1B4B\u1B83-\u1BA0\u1BAE-\u1BAF\u1BBA-\u1BE5\u1C00-\u1C23\u1C4D-\u1C4F\u1C5A-\u1C77\u1CE9-\u1CEC\u1CEE-\u1CF1\u1CF5-\u1CF6\u2135-\u2138\u2D30-\u2D67\u2D80-\u2D96\u2DA0-\u2DA6\u2DA8-\u2DAE\u2DB0-\u2DB6\u2DB8-\u2DBE\u2DC0-\u2DC6\u2DC8-\u2DCE\u2DD0-\u2DD6\u2DD8-\u2DDE\u3006\u303C\u3041-\u3096\u309F\u30A1-\u30FA\u30FF\u3105-\u312F\u3131-\u318E\u31A0-\u31BA\u31F0-\u31FF\u3400-\u4DB5\u4E00-\u9FEF\uA000-\uA014\uA016-\uA48C\uA4D0-\uA4F7\uA500-\uA60B\uA610-\uA61F\uA62A-\uA62B\uA66E\uA6A0-\uA6E5\uA78F\uA7F7\uA7FB-\uA801\uA803-\uA805\uA807-\uA80A\uA80C-\uA822\uA840-\uA873\uA882-\uA8B3\uA8F2-\uA8F7\uA8FB\uA8FD-\uA8FE\uA90A-\uA925\uA930-\uA946\uA960-\uA97C\uA984-\uA9B2\uA9E0-\uA9E4\uA9E7-\uA9EF\uA9FA-\uA9FE\uAA00-\uAA28\uAA40-\uAA42\uAA44-\uAA4B\uAA60-\uAA6F\uAA71-\uAA76\uAA7A\uAA7E-\uAAAF\uAAB1\uAAB5-\uAAB6\uAAB9-\uAABD\uAAC0\uAAC2\uAADB-\uAADC\uAAE0-\uAAEA\uAAF2\uAB01-\uAB06\uAB09-\uAB0E\uAB11-\uAB16\uAB20-\uAB26\uAB28-\uAB2E\uABC0-\uABE2\uAC00-\uD7A3\uD7B0-\uD7C6\uD7CB-\uD7FB\uF900-\uFA6D\uFA70-\uFAD9\uFB1D\uFB1F-\uFB28\uFB2A-\uFB36\uFB38-\uFB3C\uFB3E\uFB40-\uFB41\uFB43-\uFB44\uFB46-\uFBB1\uFBD3-\uFD3D\uFD50-\uFD8F\uFD92-\uFDC7\uFDF0-\uFDFB\uFE70-\uFE74\uFE76-\uFEFC\uFF66-\uFF6F\uFF71-\uFF9D\uFFA0-\uFFBE\uFFC2-\uFFC7\uFFCA-\uFFCF\uFFD2-\uFFD7\uFFDA-\uFFDC]/, m0 = /^[\u01C5\u01C8\u01CB\u01F2\u1F88-\u1F8F\u1F98-\u1F9F\u1FA8-\u1FAF\u1FBC\u1FCC\u1FFC]/, _0 = /^[A-Z\xC0-\xD6\xD8-\xDE\u0100\u0102\u0104\u0106\u0108\u010A\u010C\u010E\u0110\u0112\u0114\u0116\u0118\u011A\u011C\u011E\u0120\u0122\u0124\u0126\u0128\u012A\u012C\u012E\u0130\u0132\u0134\u0136\u0139\u013B\u013D\u013F\u0141\u0143\u0145\u0147\u014A\u014C\u014E\u0150\u0152\u0154\u0156\u0158\u015A\u015C\u015E\u0160\u0162\u0164\u0166\u0168\u016A\u016C\u016E\u0170\u0172\u0174\u0176\u0178-\u0179\u017B\u017D\u0181-\u0182\u0184\u0186-\u0187\u0189-\u018B\u018E-\u0191\u0193-\u0194\u0196-\u0198\u019C-\u019D\u019F-\u01A0\u01A2\u01A4\u01A6-\u01A7\u01A9\u01AC\u01AE-\u01AF\u01B1-\u01B3\u01B5\u01B7-\u01B8\u01BC\u01C4\u01C7\u01CA\u01CD\u01CF\u01D1\u01D3\u01D5\u01D7\u01D9\u01DB\u01DE\u01E0\u01E2\u01E4\u01E6\u01E8\u01EA\u01EC\u01EE\u01F1\u01F4\u01F6-\u01F8\u01FA\u01FC\u01FE\u0200\u0202\u0204\u0206\u0208\u020A\u020C\u020E\u0210\u0212\u0214\u0216\u0218\u021A\u021C\u021E\u0220\u0222\u0224\u0226\u0228\u022A\u022C\u022E\u0230\u0232\u023A-\u023B\u023D-\u023E\u0241\u0243-\u0246\u0248\u024A\u024C\u024E\u0370\u0372\u0376\u037F\u0386\u0388-\u038A\u038C\u038E-\u038F\u0391-\u03A1\u03A3-\u03AB\u03CF\u03D2-\u03D4\u03D8\u03DA\u03DC\u03DE\u03E0\u03E2\u03E4\u03E6\u03E8\u03EA\u03EC\u03EE\u03F4\u03F7\u03F9-\u03FA\u03FD-\u042F\u0460\u0462\u0464\u0466\u0468\u046A\u046C\u046E\u0470\u0472\u0474\u0476\u0478\u047A\u047C\u047E\u0480\u048A\u048C\u048E\u0490\u0492\u0494\u0496\u0498\u049A\u049C\u049E\u04A0\u04A2\u04A4\u04A6\u04A8\u04AA\u04AC\u04AE\u04B0\u04B2\u04B4\u04B6\u04B8\u04BA\u04BC\u04BE\u04C0-\u04C1\u04C3\u04C5\u04C7\u04C9\u04CB\u04CD\u04D0\u04D2\u04D4\u04D6\u04D8\u04DA\u04DC\u04DE\u04E0\u04E2\u04E4\u04E6\u04E8\u04EA\u04EC\u04EE\u04F0\u04F2\u04F4\u04F6\u04F8\u04FA\u04FC\u04FE\u0500\u0502\u0504\u0506\u0508\u050A\u050C\u050E\u0510\u0512\u0514\u0516\u0518\u051A\u051C\u051E\u0520\u0522\u0524\u0526\u0528\u052A\u052C\u052E\u0531-\u0556\u10A0-\u10C5\u10C7\u10CD\u13A0-\u13F5\u1C90-\u1CBA\u1CBD-\u1CBF\u1E00\u1E02\u1E04\u1E06\u1E08\u1E0A\u1E0C\u1E0E\u1E10\u1E12\u1E14\u1E16\u1E18\u1E1A\u1E1C\u1E1E\u1E20\u1E22\u1E24\u1E26\u1E28\u1E2A\u1E2C\u1E2E\u1E30\u1E32\u1E34\u1E36\u1E38\u1E3A\u1E3C\u1E3E\u1E40\u1E42\u1E44\u1E46\u1E48\u1E4A\u1E4C\u1E4E\u1E50\u1E52\u1E54\u1E56\u1E58\u1E5A\u1E5C\u1E5E\u1E60\u1E62\u1E64\u1E66\u1E68\u1E6A\u1E6C\u1E6E\u1E70\u1E72\u1E74\u1E76\u1E78\u1E7A\u1E7C\u1E7E\u1E80\u1E82\u1E84\u1E86\u1E88\u1E8A\u1E8C\u1E8E\u1E90\u1E92\u1E94\u1E9E\u1EA0\u1EA2\u1EA4\u1EA6\u1EA8\u1EAA\u1EAC\u1EAE\u1EB0\u1EB2\u1EB4\u1EB6\u1EB8\u1EBA\u1EBC\u1EBE\u1EC0\u1EC2\u1EC4\u1EC6\u1EC8\u1ECA\u1ECC\u1ECE\u1ED0\u1ED2\u1ED4\u1ED6\u1ED8\u1EDA\u1EDC\u1EDE\u1EE0\u1EE2\u1EE4\u1EE6\u1EE8\u1EEA\u1EEC\u1EEE\u1EF0\u1EF2\u1EF4\u1EF6\u1EF8\u1EFA\u1EFC\u1EFE\u1F08-\u1F0F\u1F18-\u1F1D\u1F28-\u1F2F\u1F38-\u1F3F\u1F48-\u1F4D\u1F59\u1F5B\u1F5D\u1F5F\u1F68-\u1F6F\u1FB8-\u1FBB\u1FC8-\u1FCB\u1FD8-\u1FDB\u1FE8-\u1FEC\u1FF8-\u1FFB\u2102\u2107\u210B-\u210D\u2110-\u2112\u2115\u2119-\u211D\u2124\u2126\u2128\u212A-\u212D\u2130-\u2133\u213E-\u213F\u2145\u2183\u2C00-\u2C2E\u2C60\u2C62-\u2C64\u2C67\u2C69\u2C6B\u2C6D-\u2C70\u2C72\u2C75\u2C7E-\u2C80\u2C82\u2C84\u2C86\u2C88\u2C8A\u2C8C\u2C8E\u2C90\u2C92\u2C94\u2C96\u2C98\u2C9A\u2C9C\u2C9E\u2CA0\u2CA2\u2CA4\u2CA6\u2CA8\u2CAA\u2CAC\u2CAE\u2CB0\u2CB2\u2CB4\u2CB6\u2CB8\u2CBA\u2CBC\u2CBE\u2CC0\u2CC2\u2CC4\u2CC6\u2CC8\u2CCA\u2CCC\u2CCE\u2CD0\u2CD2\u2CD4\u2CD6\u2CD8\u2CDA\u2CDC\u2CDE\u2CE0\u2CE2\u2CEB\u2CED\u2CF2\uA640\uA642\uA644\uA646\uA648\uA64A\uA64C\uA64E\uA650\uA652\uA654\uA656\uA658\uA65A\uA65C\uA65E\uA660\uA662\uA664\uA666\uA668\uA66A\uA66C\uA680\uA682\uA684\uA686\uA688\uA68A\uA68C\uA68E\uA690\uA692\uA694\uA696\uA698\uA69A\uA722\uA724\uA726\uA728\uA72A\uA72C\uA72E\uA732\uA734\uA736\uA738\uA73A\uA73C\uA73E\uA740\uA742\uA744\uA746\uA748\uA74A\uA74C\uA74E\uA750\uA752\uA754\uA756\uA758\uA75A\uA75C\uA75E\uA760\uA762\uA764\uA766\uA768\uA76A\uA76C\uA76E\uA779\uA77B\uA77D-\uA77E\uA780\uA782\uA784\uA786\uA78B\uA78D\uA790\uA792\uA796\uA798\uA79A\uA79C\uA79E\uA7A0\uA7A2\uA7A4\uA7A6\uA7A8\uA7AA-\uA7AE\uA7B0-\uA7B4\uA7B6\uA7B8\uFF21-\uFF3A]/, w0 = /^[\u16EE-\u16F0\u2160-\u2182\u2185-\u2188\u3007\u3021-\u3029\u3038-\u303A\uA6E6-\uA6EF]/, y0 = iu("number"), su = d(".", !1), b0 = M([["1", "9"]], !1, !1), x0 = M(["e", "E"], !1, !1), Pu = d("-", !1), N0 = d("+", !1), M0 = d("0", !1), k0 = M([["0", "9"]], !1, !1), L0 = iu("whitespace"), qu = M([" ", `
`, "\r", "	", "В "], !1, !1), P0 = d(",", !1), q0 = d("|", !1), j0 = d('"', !1), S0 = d("'", !1), R0 = iu('a letter, a number, "-", "#", ".", "^", "_"'), I0 = M([["0", "9"], "~"], !1, !1), O0 = d("#", !1), ju = d("^", !1), Su = d("_", !1), Ru = d("[", !1), Iu = d("]", !1), z0 = d("{", !1), Z0 = d("}", !1), U0 = d("%", !1), W0 = d("<", !1), X0 = d(">", !1), G0 = d("@", !1), T0 = d("!", !1), Y0 = d("(", !1), H0 = d(")", !1), J0 = d("/", !1), K0 = d("*", !1), Q0 = d("?", !1), V0 = d(":", !1), u1 = d("..", !1), e1 = d("struct", !1), r1 = d("target", !1), A1 = d("euclid", !1), t1 = d("slow", !1), s1 = d("rotL", !1), n1 = d("rotR", !1), i1 = d("fast", !1), E1 = d("scale", !1), a1 = d("//", !1), Ou = M([`
`], !0, !1), C1 = d("cat", !1), F1 = d("$", !1), f1 = d("setcps", !1), o1 = d("setbpm", !1), c1 = d("hush", !1), B1 = M([["a", "z"], "Вө", ["Гҹ", "Г¶"], ["Гё", "Гҝ"], "ДҒ", "Дғ", "Д…", "ДҮ", "Дү", "ДӢ", "ДҚ", "ДҸ", "Д‘", "Д“", "Д•", "Д—", "Дҷ", "Дӣ", "Дқ", "Дҹ", "ДЎ", "ДЈ", "ДҘ", "Д§", "Д©", "Д«", "Дӯ", "ДҜ", "Дұ", "Ді", "Дө", ["Д·", "Дё"], "Дә", "Дј", "Дҫ", "ЕҖ", "ЕӮ", "Е„", "ЕҶ", ["ЕҲ", "Еү"], "ЕӢ", "ЕҚ", "ЕҸ", "Е‘", "Е“", "Е•", "Е—", "Еҷ", "Еӣ", "Еқ", "Еҹ", "ЕЎ", "ЕЈ", "ЕҘ", "Е§", "Е©", "Е«", "Еӯ", "ЕҜ", "Еұ", "Еі", "Еө", "Е·", "Еә", "Еј", ["Еҫ", "ЖҖ"], "Жғ", "Ж…", "ЖҲ", ["ЖҢ", "ЖҚ"], "Ж’", "Ж•", ["Жҷ", "Жӣ"], "Жһ", "ЖЎ", "ЖЈ", "ЖҘ", "ЖЁ", ["ЖӘ", "Ж«"], "Жӯ", "Ж°", "Жҙ", "Ж¶", ["Ж№", "Жә"], ["ЖҪ", "Жҝ"], "ЗҶ", "Зү", "ЗҢ", "ЗҺ", "Зҗ", "З’", "З”", "З–", "Зҳ", "Зҡ", ["Зң", "Зқ"], "Зҹ", "ЗЎ", "ЗЈ", "ЗҘ", "З§", "З©", "З«", "Зӯ", ["ЗҜ", "З°"], "Зі", "Зө", "З№", "З»", "ЗҪ", "Зҝ", "ИҒ", "Иғ", "И…", "ИҮ", "Иү", "ИӢ", "ИҚ", "ИҸ", "И‘", "И“", "И•", "И—", "Иҷ", "Иӣ", "Иқ", "Иҹ", "ИЎ", "ИЈ", "ИҘ", "И§", "И©", "И«", "Иӯ", "ИҜ", "Иұ", ["Иі", "И№"], "Иј", ["Иҝ", "ЙҖ"], "ЙӮ", "ЙҮ", "Йү", "ЙӢ", "ЙҚ", ["ЙҸ", "К“"], ["К•", "КҜ"], "Нұ", "Ні", "Н·", ["Н»", "НҪ"], "Оҗ", ["О¬", "ПҺ"], ["Пҗ", "П‘"], ["П•", "П—"], "Пҷ", "Пӣ", "Пқ", "Пҹ", "ПЎ", "ПЈ", "ПҘ", "П§", "П©", "П«", "Пӯ", ["ПҜ", "Пі"], "Пө", "Пё", ["П»", "Пј"], ["Р°", "Сҹ"], "СЎ", "СЈ", "СҘ", "С§", "С©", "С«", "Сӯ", "СҜ", "Сұ", "Сі", "Сө", "С·", "С№", "С»", "СҪ", "Сҝ", "ТҒ", "ТӢ", "ТҚ", "ТҸ", "Т‘", "Т“", "Т•", "Т—", "Тҷ", "Тӣ", "Тқ", "Тҹ", "ТЎ", "ТЈ", "ТҘ", "Т§", "Т©", "Т«", "Тӯ", "ТҜ", "Тұ", "Ті", "Тө", "Т·", "Т№", "Т»", "ТҪ", "Тҝ", "УӮ", "У„", "УҶ", "УҲ", "УҠ", "УҢ", ["УҺ", "УҸ"], "У‘", "У“", "У•", "У—", "Уҷ", "Уӣ", "Уқ", "Уҹ", "УЎ", "УЈ", "УҘ", "У§", "У©", "У«", "Уӯ", "УҜ", "Уұ", "Уі", "Уө", "У·", "У№", "У»", "УҪ", "Уҝ", "ФҒ", "Фғ", "Ф…", "ФҮ", "Фү", "ФӢ", "ФҚ", "ФҸ", "Ф‘", "Ф“", "Ф•", "Ф—", "Фҷ", "Фӣ", "Фқ", "Фҹ", "ФЎ", "ФЈ", "ФҘ", "Ф§", "Ф©", "Ф«", "Фӯ", "ФҜ", ["Х ", "ЦҲ"], ["бғҗ", "бғә"], ["бғҪ", "бғҝ"], ["бҸё", "бҸҪ"], ["бІҖ", "бІҲ"], ["бҙҖ", "бҙ«"], ["бө«", "бө·"], ["бө№", "б¶ҡ"], "бёҒ", "бёғ", "бё…", "бёҮ", "бёү", "бёӢ", "бёҚ", "бёҸ", "бё‘", "бё“", "бё•", "бё—", "бёҷ", "бёӣ", "бёқ", "бёҹ", "бёЎ", "бёЈ", "бёҘ", "бё§", "бё©", "бё«", "бёӯ", "бёҜ", "бёұ", "бёі", "бёө", "бё·", "бё№", "бё»", "бёҪ", "бёҝ", "б№Ғ", "б№ғ", "б№…", "б№Ү", "б№ү", "б№Ӣ", "б№Қ", "б№Ҹ", "б№‘", "б№“", "б№•", "б№—", "б№ҷ", "б№ӣ", "б№қ", "б№ҹ", "б№Ў", "б№Ј", "б№Ҙ", "б№§", "б№©", "б№«", "б№ӯ", "б№Ҝ", "б№ұ", "б№і", "б№ө", "б№·", "б№№", "б№»", "б№Ҫ", "б№ҝ", "бәҒ", "бәғ", "бә…", "бәҮ", "бәү", "бәӢ", "бәҚ", "бәҸ", "бә‘", "бә“", ["бә•", "бәқ"], "бәҹ", "бәЎ", "бәЈ", "бәҘ", "бә§", "бә©", "бә«", "бәӯ", "бәҜ", "бәұ", "бәі", "бәө", "бә·", "бә№", "бә»", "бәҪ", "бәҝ", "б»Ғ", "б»ғ", "б»…", "б»Ү", "б»ү", "б»Ӣ", "б»Қ", "б»Ҹ", "б»‘", "б»“", "б»•", "б»—", "б»ҷ", "б»ӣ", "б»қ", "б»ҹ", "б»Ў", "б»Ј", "б»Ҙ", "б»§", "б»©", "б»«", "б»ӯ", "б»Ҝ", "б»ұ", "б»і", "б»ө", "б»·", "б»№", "б»»", "б»Ҫ", ["б»ҝ", "бјҮ"], ["бјҗ", "бј•"], ["бј ", "бј§"], ["бј°", "бј·"], ["бҪҖ", "бҪ…"], ["бҪҗ", "бҪ—"], ["бҪ ", "бҪ§"], ["бҪ°", "бҪҪ"], ["бҫҖ", "бҫҮ"], ["бҫҗ", "бҫ—"], ["бҫ ", "бҫ§"], ["бҫ°", "бҫҙ"], ["бҫ¶", "бҫ·"], "бҫҫ", ["бҝӮ", "бҝ„"], ["бҝҶ", "бҝҮ"], ["бҝҗ", "бҝ“"], ["бҝ–", "бҝ—"], ["бҝ ", "бҝ§"], ["бҝІ", "бҝҙ"], ["бҝ¶", "бҝ·"], "в„Ҡ", ["в„Һ", "в„Ҹ"], "в„“", "в„Ҝ", "в„ҙ", "в„№", ["в„ј", "в„Ҫ"], ["в…Ҷ", "в…ү"], "в…Һ", "вҶ„", ["в°°", "вұһ"], "вұЎ", ["вұҘ", "вұҰ"], "вұЁ", "вұӘ", "вұ¬", "вұұ", ["вұі", "вұҙ"], ["вұ¶", "вұ»"], "вІҒ", "вІғ", "вІ…", "вІҮ", "вІү", "вІӢ", "вІҚ", "вІҸ", "вІ‘", "вІ“", "вІ•", "вІ—", "вІҷ", "вІӣ", "вІқ", "вІҹ", "вІЎ", "вІЈ", "вІҘ", "вІ§", "вІ©", "вІ«", "вІӯ", "вІҜ", "вІұ", "вІі", "вІө", "вІ·", "вІ№", "вІ»", "вІҪ", "вІҝ", "віҒ", "віғ", "ві…", "віҮ", "віү", "віӢ", "віҚ", "віҸ", "ві‘", "ві“", "ві•", "ві—", "віҷ", "віӣ", "віқ", "віҹ", "віЎ", ["віЈ", "віӨ"], "ві¬", "ві®", "віі", ["вҙҖ", "вҙҘ"], "вҙ§", "вҙӯ", "кҷҒ", "кҷғ", "кҷ…", "кҷҮ", "кҷү", "кҷӢ", "кҷҚ", "кҷҸ", "кҷ‘", "кҷ“", "кҷ•", "кҷ—", "кҷҷ", "кҷӣ", "кҷқ", "кҷҹ", "кҷЎ", "кҷЈ", "кҷҘ", "кҷ§", "кҷ©", "кҷ«", "кҷӯ", "кҡҒ", "кҡғ", "кҡ…", "кҡҮ", "кҡү", "кҡӢ", "кҡҚ", "кҡҸ", "кҡ‘", "кҡ“", "кҡ•", "кҡ—", "кҡҷ", "кҡӣ", "кңЈ", "кңҘ", "кң§", "кң©", "кң«", "кңӯ", ["кңҜ", "кңұ"], "кңі", "кңө", "кң·", "кң№", "кң»", "кңҪ", "кңҝ", "кқҒ", "кқғ", "кқ…", "кқҮ", "кқү", "кқӢ", "кқҚ", "кқҸ", "кқ‘", "кқ“", "кқ•", "кқ—", "кқҷ", "кқӣ", "кққ", "кқҹ", "кқЎ", "кқЈ", "кқҘ", "кқ§", "кқ©", "кқ«", "кқӯ", "кқҜ", ["кқұ", "кқё"], "кқә", "кқј", "кқҝ", "кһҒ", "кһғ", "кһ…", "кһҮ", "кһҢ", "кһҺ", "кһ‘", ["кһ“", "кһ•"], "кһ—", "кһҷ", "кһӣ", "кһқ", "кһҹ", "кһЎ", "кһЈ", "кһҘ", "кһ§", "кһ©", "кһҜ", "кһө", "кһ·", "кһ№", "кҹә", ["к¬°", "кӯҡ"], ["кӯ ", "кӯҘ"], ["кӯ°", "к®ҝ"], ["п¬Җ", "п¬Ҷ"], ["п¬“", "п¬—"], ["пҪҒ", "пҪҡ"]], !1, !1), D1 = M([["К°", "ЛҒ"], ["ЛҶ", "Л‘"], ["Л ", "ЛӨ"], "Л¬", "Л®", "Нҙ", "Нә", "Хҷ", "ЩҖ", ["ЫҘ", "ЫҰ"], ["Яҙ", "Яө"], "Яә", "а ҡ", "а Ө", "а Ё", "аҘұ", "а№Ҷ", "а»Ҷ", "бғј", "бҹ—", "бЎғ", "бӘ§", ["бұё", "бұҪ"], ["бҙ¬", "бөӘ"], "бөё", ["б¶ӣ", "б¶ҝ"], "вҒұ", "вҒҝ", ["вӮҗ", "вӮң"], ["вұј", "вұҪ"], "вөҜ", "вёҜ", "гҖ…", ["гҖұ", "гҖө"], "гҖ»", ["гӮқ", "гӮһ"], ["гғј", "гғҫ"], "кҖ•", ["к“ё", "к“Ҫ"], "кҳҢ", "кҷҝ", ["кҡң", "кҡқ"], ["кң—", "кңҹ"], "кқ°", "кһҲ", ["кҹё", "кҹ№"], "к§Ҹ", "к§Ұ", "к©°", "к«қ", ["к«і", "к«ҙ"], ["кӯң", "кӯҹ"], "пҪ°", ["пҫһ", "пҫҹ"]], !1, !1), l1 = M(["ВӘ", "Вә", "Ж»", ["ЗҖ", "Зғ"], "К”", ["Чҗ", "ЧӘ"], ["ЧҜ", "ЧІ"], ["Ш ", "Шҝ"], ["ЩҒ", "ЩҠ"], ["Щ®", "ЩҜ"], ["Щұ", "Ы“"], "Ы•", ["Ы®", "ЫҜ"], ["Ыә", "Ыј"], "Ыҝ", "Ьҗ", ["Ь’", "ЬҜ"], ["ЭҚ", "ЮҘ"], "Юұ", ["ЯҠ", "ЯӘ"], ["а Җ", "а •"], ["аЎҖ", "аЎҳ"], ["аЎ ", "аЎӘ"], ["аў ", "аўҙ"], ["аў¶", "аўҪ"], ["аӨ„", "аӨ№"], "аӨҪ", "аҘҗ", ["аҘҳ", "аҘЎ"], ["аҘІ", "аҰҖ"], ["аҰ…", "аҰҢ"], ["аҰҸ", "аҰҗ"], ["аҰ“", "аҰЁ"], ["аҰӘ", "аҰ°"], "аҰІ", ["аҰ¶", "аҰ№"], "аҰҪ", "а§Һ", ["а§ң", "а§қ"], ["а§ҹ", "а§Ў"], ["а§°", "а§ұ"], "а§ј", ["аЁ…", "аЁҠ"], ["аЁҸ", "аЁҗ"], ["аЁ“", "аЁЁ"], ["аЁӘ", "аЁ°"], ["аЁІ", "аЁі"], ["аЁө", "аЁ¶"], ["аЁё", "аЁ№"], ["а©ҷ", "а©ң"], "а©һ", ["а©І", "а©ҙ"], ["аӘ…", "аӘҚ"], ["аӘҸ", "аӘ‘"], ["аӘ“", "аӘЁ"], ["аӘӘ", "аӘ°"], ["аӘІ", "аӘі"], ["аӘө", "аӘ№"], "аӘҪ", "а«җ", ["а« ", "а«Ў"], "а«№", ["а¬…", "а¬Ң"], ["а¬Ҹ", "а¬җ"], ["а¬“", "а¬Ё"], ["а¬Ә", "а¬°"], ["а¬І", "а¬і"], ["а¬ө", "а¬№"], "а¬Ҫ", ["аӯң", "аӯқ"], ["аӯҹ", "аӯЎ"], "аӯұ", "а®ғ", ["а®…", "а®Ҡ"], ["а®Һ", "а®җ"], ["а®’", "а®•"], ["а®ҷ", "а®ҡ"], "а®ң", ["а®һ", "а®ҹ"], ["а®Ј", "а®Ө"], ["а®Ё", "а®Ә"], ["а®®", "а®№"], "аҜҗ", ["а°…", "а°Ң"], ["а°Һ", "а°җ"], ["а°’", "а°Ё"], ["а°Ә", "а°№"], "а°Ҫ", ["аұҳ", "аұҡ"], ["аұ ", "аұЎ"], "аІҖ", ["аІ…", "аІҢ"], ["аІҺ", "аІҗ"], ["аІ’", "аІЁ"], ["аІӘ", "аІі"], ["аІө", "аІ№"], "аІҪ", "аіһ", ["аі ", "аіЎ"], ["аіұ", "аіІ"], ["аҙ…", "аҙҢ"], ["аҙҺ", "аҙҗ"], ["аҙ’", "аҙә"], "аҙҪ", "аөҺ", ["аө”", "аө–"], ["аөҹ", "аөЎ"], ["аөә", "аөҝ"], ["а¶…", "а¶–"], ["а¶ҡ", "а¶ұ"], ["а¶і", "а¶»"], "а¶Ҫ", ["а·Җ", "а·Ҷ"], ["аёҒ", "аё°"], ["аёІ", "аёі"], ["а№Җ", "а№…"], ["аәҒ", "аәӮ"], "аә„", ["аәҮ", "аәҲ"], "аәҠ", "аәҚ", ["аә”", "аә—"], ["аәҷ", "аәҹ"], ["аәЎ", "аәЈ"], "аәҘ", "аә§", ["аәӘ", "аә«"], ["аәӯ", "аә°"], ["аәІ", "аәі"], "аәҪ", ["а»Җ", "а»„"], ["а»ң", "а»ҹ"], "ајҖ", ["аҪҖ", "аҪҮ"], ["аҪү", "аҪ¬"], ["аҫҲ", "аҫҢ"], ["бҖҖ", "бҖӘ"], "бҖҝ", ["бҒҗ", "бҒ•"], ["бҒҡ", "бҒқ"], "бҒЎ", ["бҒҘ", "бҒҰ"], ["бҒ®", "бҒ°"], ["бҒө", "бӮҒ"], "бӮҺ", ["б„Җ", "бүҲ"], ["бүҠ", "бүҚ"], ["бүҗ", "бү–"], "бүҳ", ["бүҡ", "бүқ"], ["бү ", "бҠҲ"], ["бҠҠ", "бҠҚ"], ["бҠҗ", "бҠ°"], ["бҠІ", "бҠө"], ["бҠё", "бҠҫ"], "бӢҖ", ["бӢӮ", "бӢ…"], ["бӢҲ", "бӢ–"], ["бӢҳ", "бҢҗ"], ["бҢ’", "бҢ•"], ["бҢҳ", "бҚҡ"], ["бҺҖ", "бҺҸ"], ["бҗҒ", "бҷ¬"], ["бҷҜ", "бҷҝ"], ["бҡҒ", "бҡҡ"], ["бҡ ", "бӣӘ"], ["бӣұ", "бӣё"], ["бңҖ", "бңҢ"], ["бңҺ", "бң‘"], ["бң ", "бңұ"], ["бқҖ", "бқ‘"], ["бқ ", "бқ¬"], ["бқ®", "бқ°"], ["бһҖ", "бһі"], "бҹң", ["б  ", "бЎӮ"], ["бЎ„", "бЎё"], ["бўҖ", "бў„"], ["бўҮ", "бўЁ"], "бўӘ", ["бў°", "бЈө"], ["бӨҖ", "бӨһ"], ["бҘҗ", "бҘӯ"], ["бҘ°", "бҘҙ"], ["бҰҖ", "бҰ«"], ["бҰ°", "б§ү"], ["бЁҖ", "бЁ–"], ["бЁ ", "б©”"], ["б¬…", "б¬і"], ["бӯ…", "бӯӢ"], ["б®ғ", "б® "], ["б®®", "б®Ҝ"], ["б®ә", "бҜҘ"], ["б°Җ", "б°Ј"], ["бұҚ", "бұҸ"], ["бұҡ", "бұ·"], ["бі©", "бі¬"], ["бі®", "біұ"], ["біө", "бі¶"], ["в„ө", "в„ё"], ["вҙ°", "вө§"], ["в¶Җ", "в¶–"], ["в¶ ", "в¶Ұ"], ["в¶Ё", "в¶®"], ["в¶°", "в¶¶"], ["в¶ё", "в¶ҫ"], ["в·Җ", "в·Ҷ"], ["в·Ҳ", "в·Һ"], ["в·җ", "в·–"], ["в·ҳ", "в·һ"], "гҖҶ", "гҖј", ["гҒҒ", "гӮ–"], "гӮҹ", ["гӮЎ", "гғә"], "гғҝ", ["г„…", "г„Ҝ"], ["г„ұ", "гҶҺ"], ["гҶ ", "гҶә"], ["гҮ°", "гҮҝ"], ["гҗҖ", "д¶ө"], ["дёҖ", "йҝҜ"], ["кҖҖ", "кҖ”"], ["кҖ–", "к’Ң"], ["к“җ", "к“·"], ["к”Җ", "кҳӢ"], ["кҳҗ", "кҳҹ"], ["кҳӘ", "кҳ«"], "кҷ®", ["кҡ ", "кӣҘ"], "кһҸ", "кҹ·", ["кҹ»", "к Ғ"], ["к ғ", "к …"], ["к Ү", "к Ҡ"], ["к Ң", "к ў"], ["кЎҖ", "кЎі"], ["кўӮ", "кўі"], ["кЈІ", "кЈ·"], "кЈ»", ["кЈҪ", "кЈҫ"], ["кӨҠ", "кӨҘ"], ["кӨ°", "кҘҶ"], ["кҘ ", "кҘј"], ["кҰ„", "кҰІ"], ["к§ ", "к§Ө"], ["к§§", "к§Ҝ"], ["к§ә", "к§ҫ"], ["кЁҖ", "кЁЁ"], ["к©Җ", "к©Ӯ"], ["к©„", "к©Ӣ"], ["к© ", "к©Ҝ"], ["к©ұ", "к©¶"], "к©ә", ["к©ҫ", "кӘҜ"], "кӘұ", ["кӘө", "кӘ¶"], ["кӘ№", "кӘҪ"], "к«Җ", "к«Ӯ", ["к«ӣ", "к«ң"], ["к« ", "к«Ә"], "к«І", ["к¬Ғ", "к¬Ҷ"], ["к¬ү", "к¬Һ"], ["к¬‘", "к¬–"], ["к¬ ", "к¬Ұ"], ["к¬Ё", "к¬®"], ["кҜҖ", "кҜў"], ["к°Җ", "нһЈ"], ["нһ°", "нҹҶ"], ["нҹӢ", "нҹ»"], ["пӨҖ", "п©ӯ"], ["п©°", "п«ҷ"], "п¬қ", ["п¬ҹ", "п¬Ё"], ["п¬Ә", "п¬¶"], ["п¬ё", "п¬ј"], "п¬ҫ", ["пӯҖ", "пӯҒ"], ["пӯғ", "пӯ„"], ["пӯҶ", "п®ұ"], ["пҜ“", "пҙҪ"], ["пөҗ", "п¶Ҹ"], ["п¶’", "п·Ү"], ["п·°", "п·»"], ["п№°", "п№ҙ"], ["п№¶", "п»ј"], ["пҪҰ", "пҪҜ"], ["пҪұ", "пҫқ"], ["пҫ ", "пҫҫ"], ["пҝӮ", "пҝҮ"], ["пҝҠ", "пҝҸ"], ["пҝ’", "пҝ—"], ["пҝҡ", "пҝң"]], !1, !1), p1 = M(["З…", "ЗҲ", "ЗӢ", "ЗІ", ["бҫҲ", "бҫҸ"], ["бҫҳ", "бҫҹ"], ["бҫЁ", "бҫҜ"], "бҫј", "бҝҢ", "бҝј"], !1, !1), g1 = M([["A", "Z"], ["ГҖ", "Г–"], ["Гҳ", "Гһ"], "ДҖ", "ДӮ", "Д„", "ДҶ", "ДҲ", "ДҠ", "ДҢ", "ДҺ", "Дҗ", "Д’", "Д”", "Д–", "Дҳ", "Дҡ", "Дң", "Дһ", "Д ", "Дў", "ДӨ", "ДҰ", "ДЁ", "ДӘ", "Д¬", "Д®", "Д°", "ДІ", "Дҙ", "Д¶", "Д№", "Д»", "ДҪ", "Дҝ", "ЕҒ", "Еғ", "Е…", "ЕҮ", "ЕҠ", "ЕҢ", "ЕҺ", "Еҗ", "Е’", "Е”", "Е–", "Еҳ", "Еҡ", "Ең", "Еһ", "Е ", "Еў", "ЕӨ", "ЕҰ", "ЕЁ", "ЕӘ", "Е¬", "Е®", "Е°", "ЕІ", "Еҙ", "Е¶", ["Её", "Е№"], "Е»", "ЕҪ", ["ЖҒ", "ЖӮ"], "Ж„", ["ЖҶ", "ЖҮ"], ["Жү", "ЖӢ"], ["ЖҺ", "Ж‘"], ["Ж“", "Ж”"], ["Ж–", "Жҳ"], ["Жң", "Жқ"], ["Жҹ", "Ж "], "Жў", "ЖӨ", ["ЖҰ", "Ж§"], "Ж©", "Ж¬", ["Ж®", "ЖҜ"], ["Жұ", "Жі"], "Жө", ["Ж·", "Жё"], "Жј", "З„", "ЗҮ", "ЗҠ", "ЗҚ", "ЗҸ", "З‘", "З“", "З•", "З—", "Зҷ", "Зӣ", "Зһ", "З ", "Зў", "ЗӨ", "ЗҰ", "ЗЁ", "ЗӘ", "З¬", "З®", "Зұ", "Зҙ", ["З¶", "Зё"], "Зә", "Зј", "Зҫ", "ИҖ", "ИӮ", "И„", "ИҶ", "ИҲ", "ИҠ", "ИҢ", "ИҺ", "Иҗ", "И’", "И”", "И–", "Иҳ", "Иҡ", "Иң", "Иһ", "И ", "Иў", "ИӨ", "ИҰ", "ИЁ", "ИӘ", "И¬", "И®", "И°", "ИІ", ["Иә", "И»"], ["ИҪ", "Иҫ"], "ЙҒ", ["Йғ", "ЙҶ"], "ЙҲ", "ЙҠ", "ЙҢ", "ЙҺ", "Н°", "НІ", "Н¶", "Нҝ", "ОҶ", ["ОҲ", "ОҠ"], "ОҢ", ["ОҺ", "ОҸ"], ["О‘", "ОЎ"], ["ОЈ", "О«"], "ПҸ", ["П’", "П”"], "Пҳ", "Пҡ", "Пң", "Пһ", "П ", "Пў", "ПӨ", "ПҰ", "ПЁ", "ПӘ", "П¬", "П®", "Пҙ", "П·", ["П№", "Пә"], ["ПҪ", "РҜ"], "С ", "Сў", "СӨ", "СҰ", "СЁ", "СӘ", "С¬", "С®", "С°", "СІ", "Сҙ", "С¶", "Сё", "Сә", "Сј", "Сҫ", "ТҖ", "ТҠ", "ТҢ", "ТҺ", "Тҗ", "Т’", "Т”", "Т–", "Тҳ", "Тҡ", "Тң", "Тһ", "Т ", "Тў", "ТӨ", "ТҰ", "ТЁ", "ТӘ", "Т¬", "Т®", "Т°", "ТІ", "Тҙ", "Т¶", "Тё", "Тә", "Тј", "Тҫ", ["УҖ", "УҒ"], "Уғ", "У…", "УҮ", "Уү", "УӢ", "УҚ", "Уҗ", "У’", "У”", "У–", "Уҳ", "Уҡ", "Уң", "Уһ", "У ", "Уў", "УӨ", "УҰ", "УЁ", "УӘ", "У¬", "У®", "У°", "УІ", "Уҙ", "У¶", "Уё", "Уә", "Уј", "Уҫ", "ФҖ", "ФӮ", "Ф„", "ФҶ", "ФҲ", "ФҠ", "ФҢ", "ФҺ", "Фҗ", "Ф’", "Ф”", "Ф–", "Фҳ", "Фҡ", "Фң", "Фһ", "Ф ", "Фў", "ФӨ", "ФҰ", "ФЁ", "ФӘ", "Ф¬", "Ф®", ["Фұ", "Х–"], ["бӮ ", "бғ…"], "бғҮ", "бғҚ", ["бҺ ", "бҸө"], ["бІҗ", "бІә"], ["бІҪ", "бІҝ"], "бёҖ", "бёӮ", "бё„", "бёҶ", "бёҲ", "бёҠ", "бёҢ", "бёҺ", "бёҗ", "бё’", "бё”", "бё–", "бёҳ", "бёҡ", "бёң", "бёһ", "бё ", "бёў", "бёӨ", "бёҰ", "бёЁ", "бёӘ", "бё¬", "бё®", "бё°", "бёІ", "бёҙ", "бё¶", "бёё", "бёә", "бёј", "бёҫ", "б№Җ", "б№Ӯ", "б№„", "б№Ҷ", "б№Ҳ", "б№Ҡ", "б№Ң", "б№Һ", "б№җ", "б№’", "б№”", "б№–", "б№ҳ", "б№ҡ", "б№ң", "б№һ", "б№ ", "б№ў", "б№Ө", "б№Ұ", "б№Ё", "б№Ә", "б№¬", "б№®", "б№°", "б№І", "б№ҙ", "б№¶", "б№ё", "б№ә", "б№ј", "б№ҫ", "бәҖ", "бәӮ", "бә„", "бәҶ", "бәҲ", "бәҠ", "бәҢ", "бәҺ", "бәҗ", "бә’", "бә”", "бәһ", "бә ", "бәў", "бәӨ", "бәҰ", "бәЁ", "бәӘ", "бә¬", "бә®", "бә°", "бәІ", "бәҙ", "бә¶", "бәё", "бәә", "бәј", "бәҫ", "б»Җ", "б»Ӯ", "б»„", "б»Ҷ", "б»Ҳ", "б»Ҡ", "б»Ң", "б»Һ", "б»җ", "б»’", "б»”", "б»–", "б»ҳ", "б»ҡ", "б»ң", "б»һ", "б» ", "б»ў", "б»Ө", "б»Ұ", "б»Ё", "б»Ә", "б»¬", "б»®", "б»°", "б»І", "б»ҙ", "б»¶", "б»ё", "б»ә", "б»ј", "б»ҫ", ["бјҲ", "бјҸ"], ["бјҳ", "бјқ"], ["бјЁ", "бјҜ"], ["бјё", "бјҝ"], ["бҪҲ", "бҪҚ"], "бҪҷ", "бҪӣ", "бҪқ", "бҪҹ", ["бҪЁ", "бҪҜ"], ["бҫё", "бҫ»"], ["бҝҲ", "бҝӢ"], ["бҝҳ", "бҝӣ"], ["бҝЁ", "бҝ¬"], ["бҝё", "бҝ»"], "в„Ӯ", "в„Ү", ["в„Ӣ", "в„Қ"], ["в„җ", "в„’"], "в„•", ["в„ҷ", "в„қ"], "в„Ө", "в„Ұ", "в„Ё", ["в„Ә", "в„ӯ"], ["в„°", "в„і"], ["в„ҫ", "в„ҝ"], "в……", "вҶғ", ["в°Җ", "в°®"], "вұ ", ["вұў", "вұӨ"], "вұ§", "вұ©", "вұ«", ["вұӯ", "вұ°"], "вұІ", "вұө", ["вұҫ", "вІҖ"], "вІӮ", "вІ„", "вІҶ", "вІҲ", "вІҠ", "вІҢ", "вІҺ", "вІҗ", "вІ’", "вІ”", "вІ–", "вІҳ", "вІҡ", "вІң", "вІһ", "вІ ", "вІў", "вІӨ", "вІҰ", "вІЁ", "вІӘ", "вІ¬", "вІ®", "вІ°", "вІІ", "вІҙ", "вІ¶", "вІё", "вІә", "вІј", "вІҫ", "віҖ", "віӮ", "ві„", "віҶ", "віҲ", "віҠ", "віҢ", "віҺ", "віҗ", "ві’", "ві”", "ві–", "віҳ", "віҡ", "вің", "віһ", "ві ", "віў", "ві«", "віӯ", "віІ", "кҷҖ", "кҷӮ", "кҷ„", "кҷҶ", "кҷҲ", "кҷҠ", "кҷҢ", "кҷҺ", "кҷҗ", "кҷ’", "кҷ”", "кҷ–", "кҷҳ", "кҷҡ", "кҷң", "кҷһ", "кҷ ", "кҷў", "кҷӨ", "кҷҰ", "кҷЁ", "кҷӘ", "кҷ¬", "кҡҖ", "кҡӮ", "кҡ„", "кҡҶ", "кҡҲ", "кҡҠ", "кҡҢ", "кҡҺ", "кҡҗ", "кҡ’", "кҡ”", "кҡ–", "кҡҳ", "кҡҡ", "кңў", "кңӨ", "кңҰ", "кңЁ", "кңӘ", "кң¬", "кң®", "кңІ", "кңҙ", "кң¶", "кңё", "кңә", "кңј", "кңҫ", "кқҖ", "кқӮ", "кқ„", "кқҶ", "кқҲ", "кқҠ", "кқҢ", "кқҺ", "кқҗ", "кқ’", "кқ”", "кқ–", "кқҳ", "кқҡ", "кқң", "кқһ", "кқ ", "кқў", "кқӨ", "кқҰ", "кқЁ", "кқӘ", "кқ¬", "кқ®", "кқ№", "кқ»", ["кқҪ", "кқҫ"], "кһҖ", "кһӮ", "кһ„", "кһҶ", "кһӢ", "кһҚ", "кһҗ", "кһ’", "кһ–", "кһҳ", "кһҡ", "кһң", "кһһ", "кһ ", "кһў", "кһӨ", "кһҰ", "кһЁ", ["кһӘ", "кһ®"], ["кһ°", "кһҙ"], "кһ¶", "кһё", ["пјЎ", "пјә"]], !1, !1), h1 = M([["бӣ®", "бӣ°"], ["в… ", "вҶӮ"], ["вҶ…", "вҶҲ"], "гҖҮ", ["гҖЎ", "гҖ©"], ["гҖё", "гҖә"], ["кӣҰ", "кӣҜ"]], !1, !1), v1 = function() {
    return parseFloat(ne());
  }, d1 = function(u) {
    const A = u.join("");
    return A === "." || A === "_";
  }, $1 = function(u) {
    return new rr(u.join(""));
  }, m1 = function(u) {
    return u;
  }, _1 = function(u, A) {
    return u.arguments_.stepsPerCycle = A, u;
  }, w1 = function(u) {
    return u;
  }, y1 = function(u) {
    return u.arguments_.alignment = "polymeter_slowcat", u;
  }, b1 = function(u) {
    return (A) => A.options_.weight = (A.options_.weight ?? 1) + (u ?? 2) - 1;
  }, x1 = function(u) {
    return (A) => {
      const s = (A.options_.reps ?? 1) + (u ?? 2) - 1;
      A.options_.reps = s, A.options_.ops = A.options_.ops.filter((i) => i.type_ !== "replicate"), A.options_.ops.push({ type_: "replicate", arguments_: { amount: s } }), A.options_.weight = s;
    };
  }, N1 = function(u, A, s) {
    return (i) => i.options_.ops.push({ type_: "bjorklund", arguments_: { pulse: u, step: A, rotation: s } });
  }, M1 = function(u) {
    return (A) => A.options_.ops.push({ type_: "stretch", arguments_: { amount: u, type: "slow" } });
  }, k1 = function(u) {
    return (A) => A.options_.ops.push({ type_: "stretch", arguments_: { amount: u, type: "fast" } });
  }, L1 = function(u) {
    return (A) => A.options_.ops.push({ type_: "degradeBy", arguments_: { amount: u, seed: Fu++ } });
  }, P1 = function(u) {
    return (A) => A.options_.ops.push({ type_: "tail", arguments_: { element: u } });
  }, q1 = function(u) {
    return (A) => A.options_.ops.push({ type_: "range", arguments_: { element: u } });
  }, j1 = function(u, A) {
    const s = new tr(u, { ops: [], weight: 1, reps: 1 });
    for (const i of A)
      i(s);
    return s;
  }, S1 = function(u, A) {
    return new eu(A, "fastcat", void 0, !!u);
  }, R1 = function(u) {
    return { alignment: "stack", list: u };
  }, I1 = function(u) {
    return { alignment: "rand", list: u, seed: Fu++ };
  }, O1 = function(u) {
    return { alignment: "feet", list: u, seed: Fu++ };
  }, z1 = function(u, A) {
    return A && A.list.length > 0 ? new eu([u, ...A.list], A.alignment, A.seed) : u;
  }, Z1 = function(u, A) {
    return new eu(A ? [u, ...A.list] : [u], "polymeter");
  }, U1 = function(u) {
    return u;
  }, W1 = function(u) {
    return { name: "struct", args: { mini: u } };
  }, X1 = function(u) {
    return { name: "target", args: { name: u } };
  }, G1 = function(u, A, s) {
    return { name: "bjorklund", args: { pulse: u, step: parseInt(A) } };
  }, T1 = function(u) {
    return { name: "stretch", args: { amount: u } };
  }, Y1 = function(u) {
    return { name: "shift", args: { amount: "-" + u } };
  }, H1 = function(u) {
    return { name: "shift", args: { amount: u } };
  }, J1 = function(u) {
    return { name: "stretch", args: { amount: "1/" + u } };
  }, K1 = function(u) {
    return { name: "scale", args: { scale: u.join("") } };
  }, zu = function(u, A) {
    return A;
  }, Q1 = function(u, A) {
    return A.unshift(u), new eu(A, "slowcat");
  }, V1 = function(u) {
    return u;
  }, ue = function(u, A) {
    return new Ar(u.name, u.args, A);
  }, ee = function(u) {
    return u;
  }, re = function(u) {
    return u;
  }, Ae = function(u) {
    return new Cu("setcps", { value: u });
  }, te = function(u) {
    return new Cu("setcps", { value: u / 120 / 2 });
  }, se = function() {
    return new Cu("hush");
  }, t = 0, h = 0, K = [{ line: 1, column: 1 }], L = 0, nu = [], f = 0, Q;
  if ("startRule" in n) {
    if (!(n.startRule in C))
      throw new Error(`Can't start parsing from rule "` + n.startRule + '".');
    c = C[n.startRule];
  }
  function ne() {
    return r.substring(h, t);
  }
  function Zu() {
    return Eu(h, t);
  }
  function d(u, A) {
    return { type: "literal", text: u, ignoreCase: A };
  }
  function M(u, A, s) {
    return { type: "class", parts: u, inverted: A, ignoreCase: s };
  }
  function ie() {
    return { type: "end" };
  }
  function iu(u) {
    return { type: "other", description: u };
  }
  function Uu(u) {
    var A = K[u], s;
    if (A)
      return A;
    for (s = u - 1; !K[s]; )
      s--;
    for (A = K[s], A = {
      line: A.line,
      column: A.column
    }; s < u; )
      r.charCodeAt(s) === 10 ? (A.line++, A.column = 1) : A.column++, s++;
    return K[u] = A, A;
  }
  function Eu(u, A, s) {
    var i = Uu(u), l = Uu(A), _ = {
      source: E,
      start: {
        offset: u,
        line: i.line,
        column: i.column
      },
      end: {
        offset: A,
        line: l.line,
        column: l.column
      }
    };
    return s && E && typeof E.offset == "function" && (_.start = E.offset(_.start), _.end = E.offset(_.end)), _;
  }
  function o(u) {
    t < L || (t > L && (L = t, nu = []), nu.push(u));
  }
  function Ee(u, A, s) {
    return new Y(
      Y.buildMessage(u, A),
      u,
      A,
      s
    );
  }
  function Wu() {
    var u;
    return u = Ye(), u;
  }
  function S() {
    var u, A;
    return f++, u = t, Xu(), A = V(), A !== e ? (oe(), fe(), h = u, u = v1()) : (t = u, u = e), f--, u === e && f === 0 && o(y0), u;
  }
  function ae() {
    var u;
    return r.charCodeAt(t) === 46 ? (u = g, t++) : (u = e, f === 0 && o(su)), u;
  }
  function Ce() {
    var u;
    return l0.test(r.charAt(t)) ? (u = r.charAt(t), t++) : (u = e, f === 0 && o(b0)), u;
  }
  function Fe() {
    var u;
    return p0.test(r.charAt(t)) ? (u = r.charAt(t), t++) : (u = e, f === 0 && o(x0)), u;
  }
  function fe() {
    var u, A, s, i, l;
    if (u = t, A = Fe(), A !== e) {
      if (s = Xu(), s === e && (s = ce()), s === e && (s = null), i = [], l = z(), l !== e)
        for (; l !== e; )
          i.push(l), l = z();
      else
        i = e;
      i !== e ? (A = [A, s, i], u = A) : (t = u, u = e);
    } else
      t = u, u = e;
    return u;
  }
  function oe() {
    var u, A, s, i;
    if (u = t, A = ae(), A !== e) {
      if (s = [], i = z(), i !== e)
        for (; i !== e; )
          s.push(i), i = z();
      else
        s = e;
      s !== e ? (A = [A, s], u = A) : (t = u, u = e);
    } else
      t = u, u = e;
    return u;
  }
  function V() {
    var u, A, s, i;
    if (u = Be(), u === e)
      if (u = t, A = Ce(), A !== e) {
        for (s = [], i = z(); i !== e; )
          s.push(i), i = z();
        A = [A, s], u = A;
      } else
        t = u, u = e;
    return u;
  }
  function Xu() {
    var u;
    return r.charCodeAt(t) === 45 ? (u = $, t++) : (u = e, f === 0 && o(Pu)), u;
  }
  function ce() {
    var u;
    return r.charCodeAt(t) === 43 ? (u = v, t++) : (u = e, f === 0 && o(N0)), u;
  }
  function Be() {
    var u;
    return r.charCodeAt(t) === 48 ? (u = D, t++) : (u = e, f === 0 && o(M0)), u;
  }
  function z() {
    var u;
    return g0.test(r.charAt(t)) ? (u = r.charAt(t), t++) : (u = e, f === 0 && o(k0)), u;
  }
  function p() {
    var u, A;
    for (f++, u = [], ku.test(r.charAt(t)) ? (A = r.charAt(t), t++) : (A = e, f === 0 && o(qu)); A !== e; )
      u.push(A), ku.test(r.charAt(t)) ? (A = r.charAt(t), t++) : (A = e, f === 0 && o(qu));
    return f--, A = e, f === 0 && o(L0), u;
  }
  function Z() {
    var u, A, s, i;
    return u = t, A = p(), r.charCodeAt(t) === 44 ? (s = a, t++) : (s = e, f === 0 && o(P0)), s !== e ? (i = p(), A = [A, s, i], u = A) : (t = u, u = e), u;
  }
  function Gu() {
    var u, A, s, i;
    return u = t, A = p(), r.charCodeAt(t) === 124 ? (s = F, t++) : (s = e, f === 0 && o(q0)), s !== e ? (i = p(), A = [A, s, i], u = A) : (t = u, u = e), u;
  }
  function Tu() {
    var u, A, s, i;
    return u = t, A = p(), r.charCodeAt(t) === 46 ? (s = g, t++) : (s = e, f === 0 && o(su)), s !== e ? (i = p(), A = [A, s, i], u = A) : (t = u, u = e), u;
  }
  function U() {
    var u;
    return r.charCodeAt(t) === 34 ? (u = y, t++) : (u = e, f === 0 && o(j0)), u === e && (r.charCodeAt(t) === 39 ? (u = x, t++) : (u = e, f === 0 && o(S0))), u;
  }
  function uu() {
    var u;
    return f++, u = He(), u === e && (h0.test(r.charAt(t)) ? (u = r.charAt(t), t++) : (u = e, f === 0 && o(I0)), u === e && (r.charCodeAt(t) === 45 ? (u = $, t++) : (u = e, f === 0 && o(Pu)), u === e && (r.charCodeAt(t) === 35 ? (u = q, t++) : (u = e, f === 0 && o(O0)), u === e && (r.charCodeAt(t) === 46 ? (u = g, t++) : (u = e, f === 0 && o(su)), u === e && (r.charCodeAt(t) === 94 ? (u = j, t++) : (u = e, f === 0 && o(ju)), u === e && (r.charCodeAt(t) === 95 ? (u = Bu, t++) : (u = e, f === 0 && o(Su)))))))), f--, u === e && f === 0 && o(R0), u;
  }
  function Yu() {
    var u, A, s, i;
    if (u = t, p(), A = [], s = uu(), s !== e)
      for (; s !== e; )
        A.push(s), s = uu();
    else
      A = e;
    return A !== e ? (s = p(), h = t, i = d1(A), i ? i = e : i = void 0, i !== e ? (h = u, u = $1(A)) : (t = u, u = e)) : (t = u, u = e), u;
  }
  function De() {
    var u, A, s, i;
    return u = t, p(), r.charCodeAt(t) === 91 ? (A = Du, t++) : (A = e, f === 0 && o(Ru)), A !== e ? (p(), s = Ku(), s !== e ? (p(), r.charCodeAt(t) === 93 ? (i = lu, t++) : (i = e, f === 0 && o(Iu)), i !== e ? (p(), h = u, u = m1(s)) : (t = u, u = e)) : (t = u, u = e)) : (t = u, u = e), u;
  }
  function le() {
    var u, A, s, i, l;
    return u = t, p(), r.charCodeAt(t) === 123 ? (A = A0, t++) : (A = e, f === 0 && o(z0)), A !== e ? (p(), s = Qu(), s !== e ? (p(), r.charCodeAt(t) === 125 ? (i = t0, t++) : (i = e, f === 0 && o(Z0)), i !== e ? (l = pe(), l === e && (l = null), p(), h = u, u = _1(s, l)) : (t = u, u = e)) : (t = u, u = e)) : (t = u, u = e), u;
  }
  function pe() {
    var u, A, s;
    return u = t, r.charCodeAt(t) === 37 ? (A = s0, t++) : (A = e, f === 0 && o(U0)), A !== e ? (s = W(), s !== e ? (h = u, u = w1(s)) : (t = u, u = e)) : (t = u, u = e), u;
  }
  function ge() {
    var u, A, s, i;
    return u = t, p(), r.charCodeAt(t) === 60 ? (A = n0, t++) : (A = e, f === 0 && o(W0)), A !== e ? (p(), s = Qu(), s !== e ? (p(), r.charCodeAt(t) === 62 ? (i = i0, t++) : (i = e, f === 0 && o(X0)), i !== e ? (p(), h = u, u = y1(s)) : (t = u, u = e)) : (t = u, u = e)) : (t = u, u = e), u;
  }
  function W() {
    var u;
    return u = Yu(), u === e && (u = De(), u === e && (u = le(), u === e && (u = ge()))), u;
  }
  function Hu() {
    var u;
    return u = he(), u === e && (u = de(), u === e && (u = $e(), u === e && (u = me(), u === e && (u = ve(), u === e && (u = _e(), u === e && (u = we(), u === e && (u = ye()))))))), u;
  }
  function he() {
    var u, A, s;
    return u = t, p(), r.charCodeAt(t) === 64 ? (A = E0, t++) : (A = e, f === 0 && o(G0)), A === e && (r.charCodeAt(t) === 95 ? (A = Bu, t++) : (A = e, f === 0 && o(Su))), A !== e ? (s = S(), s === e && (s = null), h = u, u = b1(s)) : (t = u, u = e), u;
  }
  function ve() {
    var u, A, s;
    return u = t, p(), r.charCodeAt(t) === 33 ? (A = a0, t++) : (A = e, f === 0 && o(T0)), A !== e ? (s = S(), s === e && (s = null), h = u, u = x1(s)) : (t = u, u = e), u;
  }
  function de() {
    var u, A, s, i, l, _, k;
    return u = t, r.charCodeAt(t) === 40 ? (A = C0, t++) : (A = e, f === 0 && o(Y0)), A !== e ? (p(), s = H(), s !== e ? (p(), i = Z(), i !== e ? (p(), l = H(), l !== e ? (p(), Z(), p(), _ = H(), _ === e && (_ = null), p(), r.charCodeAt(t) === 41 ? (k = F0, t++) : (k = e, f === 0 && o(H0)), k !== e ? (h = u, u = N1(s, l, _)) : (t = u, u = e)) : (t = u, u = e)) : (t = u, u = e)) : (t = u, u = e)) : (t = u, u = e), u;
  }
  function $e() {
    var u, A, s;
    return u = t, r.charCodeAt(t) === 47 ? (A = f0, t++) : (A = e, f === 0 && o(J0)), A !== e ? (s = W(), s !== e ? (h = u, u = M1(s)) : (t = u, u = e)) : (t = u, u = e), u;
  }
  function me() {
    var u, A, s;
    return u = t, r.charCodeAt(t) === 42 ? (A = o0, t++) : (A = e, f === 0 && o(K0)), A !== e ? (s = W(), s !== e ? (h = u, u = k1(s)) : (t = u, u = e)) : (t = u, u = e), u;
  }
  function _e() {
    var u, A, s;
    return u = t, r.charCodeAt(t) === 63 ? (A = c0, t++) : (A = e, f === 0 && o(Q0)), A !== e ? (s = S(), s === e && (s = null), h = u, u = L1(s)) : (t = u, u = e), u;
  }
  function we() {
    var u, A, s;
    return u = t, r.charCodeAt(t) === 58 ? (A = B0, t++) : (A = e, f === 0 && o(V0)), A !== e ? (s = W(), s !== e ? (h = u, u = P1(s)) : (t = u, u = e)) : (t = u, u = e), u;
  }
  function ye() {
    var u, A, s;
    return u = t, r.substr(t, 2) === pu ? (A = pu, t += 2) : (A = e, f === 0 && o(u1)), A !== e ? (s = W(), s !== e ? (h = u, u = q1(s)) : (t = u, u = e)) : (t = u, u = e), u;
  }
  function H() {
    var u, A, s, i;
    if (u = t, A = W(), A !== e) {
      for (s = [], i = Hu(); i !== e; )
        s.push(i), i = Hu();
      h = u, u = j1(A, s);
    } else
      t = u, u = e;
    return u;
  }
  function I() {
    var u, A, s, i;
    if (u = t, r.charCodeAt(t) === 94 ? (A = j, t++) : (A = e, f === 0 && o(ju)), A === e && (A = null), s = [], i = H(), i !== e)
      for (; i !== e; )
        s.push(i), i = H();
    else
      s = e;
    return s !== e ? (h = u, u = S1(A, s)) : (t = u, u = e), u;
  }
  function Ju() {
    var u, A, s, i, l;
    if (u = t, A = [], s = t, i = Z(), i !== e ? (l = I(), l !== e ? s = l : (t = s, s = e)) : (t = s, s = e), s !== e)
      for (; s !== e; )
        A.push(s), s = t, i = Z(), i !== e ? (l = I(), l !== e ? s = l : (t = s, s = e)) : (t = s, s = e);
    else
      A = e;
    return A !== e && (h = u, A = R1(A)), u = A, u;
  }
  function be() {
    var u, A, s, i, l;
    if (u = t, A = [], s = t, i = Gu(), i !== e ? (l = I(), l !== e ? s = l : (t = s, s = e)) : (t = s, s = e), s !== e)
      for (; s !== e; )
        A.push(s), s = t, i = Gu(), i !== e ? (l = I(), l !== e ? s = l : (t = s, s = e)) : (t = s, s = e);
    else
      A = e;
    return A !== e && (h = u, A = I1(A)), u = A, u;
  }
  function xe() {
    var u, A, s, i, l;
    if (u = t, A = [], s = t, i = Tu(), i !== e ? (l = I(), l !== e ? s = l : (t = s, s = e)) : (t = s, s = e), s !== e)
      for (; s !== e; )
        A.push(s), s = t, i = Tu(), i !== e ? (l = I(), l !== e ? s = l : (t = s, s = e)) : (t = s, s = e);
    else
      A = e;
    return A !== e && (h = u, A = O1(A)), u = A, u;
  }
  function Ku() {
    var u, A, s;
    return u = t, A = I(), A !== e ? (s = Ju(), s === e && (s = be(), s === e && (s = xe())), s === e && (s = null), h = u, u = z1(A, s)) : (t = u, u = e), u;
  }
  function Qu() {
    var u, A, s;
    return u = t, A = I(), A !== e ? (s = Ju(), s === e && (s = null), h = u, u = Z1(A, s)) : (t = u, u = e), u;
  }
  function Ne() {
    var u, A, s, i;
    return u = t, p(), A = U(), A !== e ? (p(), s = Ku(), s !== e ? (p(), i = U(), i !== e ? (h = u, u = U1(s)) : (t = u, u = e)) : (t = u, u = e)) : (t = u, u = e), u;
  }
  function Me() {
    var u;
    return u = Ie(), u === e && (u = qe(), u === e && (u = Re(), u === e && (u = Le(), u === e && (u = Pe(), u === e && (u = ke(), u === e && (u = Se(), u === e && (u = je()))))))), u;
  }
  function ke() {
    var u, A, s;
    return u = t, r.substr(t, 6) === gu ? (A = gu, t += 6) : (A = e, f === 0 && o(e1)), A !== e ? (p(), s = X(), s !== e ? (h = u, u = W1(s)) : (t = u, u = e)) : (t = u, u = e), u;
  }
  function Le() {
    var u, A, s, i, l;
    return u = t, r.substr(t, 6) === hu ? (A = hu, t += 6) : (A = e, f === 0 && o(r1)), A !== e ? (p(), s = U(), s !== e ? (i = Yu(), i !== e ? (l = U(), l !== e ? (h = u, u = X1(i)) : (t = u, u = e)) : (t = u, u = e)) : (t = u, u = e)) : (t = u, u = e), u;
  }
  function Pe() {
    var u, A, s, i;
    return u = t, r.substr(t, 6) === vu ? (A = vu, t += 6) : (A = e, f === 0 && o(A1)), A !== e ? (p(), s = V(), s !== e ? (p(), i = V(), i !== e ? (p(), V(), h = u, u = G1(s, i)) : (t = u, u = e)) : (t = u, u = e)) : (t = u, u = e), u;
  }
  function qe() {
    var u, A, s;
    return u = t, r.substr(t, 4) === du ? (A = du, t += 4) : (A = e, f === 0 && o(t1)), A !== e ? (p(), s = S(), s !== e ? (h = u, u = T1(s)) : (t = u, u = e)) : (t = u, u = e), u;
  }
  function je() {
    var u, A, s;
    return u = t, r.substr(t, 4) === $u ? (A = $u, t += 4) : (A = e, f === 0 && o(s1)), A !== e ? (p(), s = S(), s !== e ? (h = u, u = Y1(s)) : (t = u, u = e)) : (t = u, u = e), u;
  }
  function Se() {
    var u, A, s;
    return u = t, r.substr(t, 4) === mu ? (A = mu, t += 4) : (A = e, f === 0 && o(n1)), A !== e ? (p(), s = S(), s !== e ? (h = u, u = H1(s)) : (t = u, u = e)) : (t = u, u = e), u;
  }
  function Re() {
    var u, A, s;
    return u = t, r.substr(t, 4) === _u ? (A = _u, t += 4) : (A = e, f === 0 && o(i1)), A !== e ? (p(), s = S(), s !== e ? (h = u, u = J1(s)) : (t = u, u = e)) : (t = u, u = e), u;
  }
  function Ie() {
    var u, A, s, i, l;
    if (u = t, r.substr(t, 5) === wu ? (A = wu, t += 5) : (A = e, f === 0 && o(E1)), A !== e)
      if (p(), s = U(), s !== e) {
        if (i = [], l = uu(), l !== e)
          for (; l !== e; )
            i.push(l), l = uu();
        else
          i = e;
        i !== e ? (l = U(), l !== e ? (h = u, u = K1(i)) : (t = u, u = e)) : (t = u, u = e);
      } else
        t = u, u = e;
    else
      t = u, u = e;
    return u;
  }
  function au() {
    var u, A, s, i;
    if (u = t, r.substr(t, 2) === yu ? (A = yu, t += 2) : (A = e, f === 0 && o(a1)), A !== e) {
      for (s = [], Lu.test(r.charAt(t)) ? (i = r.charAt(t), t++) : (i = e, f === 0 && o(Ou)); i !== e; )
        s.push(i), Lu.test(r.charAt(t)) ? (i = r.charAt(t), t++) : (i = e, f === 0 && o(Ou));
      A = [A, s], u = A;
    } else
      t = u, u = e;
    return u;
  }
  function Oe() {
    var u, A, s, i, l, _, k, G;
    if (u = t, r.substr(t, 3) === bu ? (A = bu, t += 3) : (A = e, f === 0 && o(C1)), A !== e)
      if (p(), r.charCodeAt(t) === 91 ? (s = Du, t++) : (s = e, f === 0 && o(Ru)), s !== e)
        if (p(), i = X(), i !== e) {
          for (l = [], _ = t, k = Z(), k !== e ? (G = X(), G !== e ? (h = _, _ = zu(i, G)) : (t = _, _ = e)) : (t = _, _ = e); _ !== e; )
            l.push(_), _ = t, k = Z(), k !== e ? (G = X(), G !== e ? (h = _, _ = zu(i, G)) : (t = _, _ = e)) : (t = _, _ = e);
          _ = p(), r.charCodeAt(t) === 93 ? (k = lu, t++) : (k = e, f === 0 && o(Iu)), k !== e ? (h = u, u = Q1(i, l)) : (t = u, u = e);
        } else
          t = u, u = e;
      else
        t = u, u = e;
    else
      t = u, u = e;
    return u;
  }
  function ze() {
    var u;
    return u = Oe(), u === e && (u = Ne()), u;
  }
  function X() {
    var u, A, s, i, l;
    if (u = t, A = ze(), A !== e) {
      for (p(), s = [], i = au(); i !== e; )
        s.push(i), i = au();
      h = u, u = V1(A);
    } else
      t = u, u = e;
    return u === e && (u = t, A = Me(), A !== e ? (p(), r.charCodeAt(t) === 36 ? (s = D0, t++) : (s = e, f === 0 && o(F1)), s !== e ? (i = p(), l = X(), l !== e ? (h = u, u = ue(A, l)) : (t = u, u = e)) : (t = u, u = e)) : (t = u, u = e)), u;
  }
  function Ze() {
    var u, A;
    return u = t, A = X(), A !== e && (h = u, A = ee(A)), u = A, u === e && (u = au()), u;
  }
  function Ue() {
    var u;
    return u = Ze(), u;
  }
  function We() {
    var u, A;
    return u = t, p(), A = Xe(), A === e && (A = Ge(), A === e && (A = Te())), A !== e ? (p(), h = u, u = re(A)) : (t = u, u = e), u;
  }
  function Xe() {
    var u, A, s;
    return u = t, r.substr(t, 6) === xu ? (A = xu, t += 6) : (A = e, f === 0 && o(f1)), A !== e ? (p(), s = S(), s !== e ? (h = u, u = Ae(s)) : (t = u, u = e)) : (t = u, u = e), u;
  }
  function Ge() {
    var u, A, s;
    return u = t, r.substr(t, 6) === Nu ? (A = Nu, t += 6) : (A = e, f === 0 && o(o1)), A !== e ? (p(), s = S(), s !== e ? (h = u, u = te(s)) : (t = u, u = e)) : (t = u, u = e), u;
  }
  function Te() {
    var u, A;
    return u = t, r.substr(t, 4) === Mu ? (A = Mu, t += 4) : (A = e, f === 0 && o(c1)), A !== e && (h = u, A = se()), u = A, u;
  }
  function Ye() {
    var u;
    return u = Ue(), u === e && (u = We()), u;
  }
  function He() {
    var u;
    return u = ur(), u === e && (u = Je(), u === e && (u = Ve(), u === e && (u = Ke(), u === e && (u = Qe(), u === e && (u = er()))))), u;
  }
  function Je() {
    var u;
    return v0.test(r.charAt(t)) ? (u = r.charAt(t), t++) : (u = e, f === 0 && o(B1)), u;
  }
  function Ke() {
    var u;
    return d0.test(r.charAt(t)) ? (u = r.charAt(t), t++) : (u = e, f === 0 && o(D1)), u;
  }
  function Qe() {
    var u;
    return $0.test(r.charAt(t)) ? (u = r.charAt(t), t++) : (u = e, f === 0 && o(l1)), u;
  }
  function Ve() {
    var u;
    return m0.test(r.charAt(t)) ? (u = r.charAt(t), t++) : (u = e, f === 0 && o(p1)), u;
  }
  function ur() {
    var u;
    return _0.test(r.charAt(t)) ? (u = r.charAt(t), t++) : (u = e, f === 0 && o(g1)), u;
  }
  function er() {
    var u;
    return w0.test(r.charAt(t)) ? (u = r.charAt(t), t++) : (u = e, f === 0 && o(h1)), u;
  }
  var rr = function(u) {
    this.type_ = "atom", this.source_ = u, this.location_ = Zu();
  }, eu = function(u, A, s, i) {
    this.type_ = "pattern", this.arguments_ = { alignment: A, tactus: i }, s !== void 0 && (this.arguments_.seed = s), this.source_ = u;
  }, Ar = function(u, A, s) {
    this.type_ = u, this.arguments_ = A, this.source_ = s;
  }, tr = function(u, A) {
    this.type_ = "element", this.source_ = u, this.options_ = A, this.location_ = Zu();
  }, Cu = function(u, A) {
    this.type_ = "command", this.name_ = u, this.options_ = A;
  }, Fu = 0;
  if (Q = c(), Q !== e && t === r.length)
    return Q;
  throw Q !== e && t < r.length && o(ie()), Ee(
    nu,
    L < r.length ? r.charAt(L) : null,
    L < r.length ? Eu(L, L + 1) : Eu(L, L)
  );
}
/**
 * @license Fraction.js v4.3.7 31/08/2023
 * https://www.xarg.org/2014/03/rational-numbers-in-javascript/
 *
 * Copyright (c) 2023, Robert Eisele (robert@raw.org)
 * Dual licensed under the MIT or GPL Version 2 licenses.
 **/
var ir = 2e3, B = {
  s: 1,
  n: 0,
  d: 1
};
function R(r, n) {
  if (isNaN(r = parseInt(r, 10)))
    throw Au();
  return r * n;
}
function b(r, n) {
  if (n === 0)
    throw ou();
  var e = Object.create(m.prototype);
  e.s = r < 0 ? -1 : 1, r = r < 0 ? -r : r;
  var E = T(r, n);
  return e.n = r / E, e.d = n / E, e;
}
function Vu(r) {
  for (var n = {}, e = r, E = 2, C = 4; C <= e; ) {
    for (; e % E === 0; )
      e /= E, n[E] = (n[E] || 0) + 1;
    C += 1 + 2 * E++;
  }
  return e !== r ? e > 1 && (n[e] = (n[e] || 0) + 1) : n[r] = (n[r] || 0) + 1, n;
}
var N = function(r, n) {
  var e = 0, E = 1, C = 1, c = 0, g = 0, $ = 0, v = 1, D = 1, a = 0, F = 1, y = 1, x = 1, q = 1e7, j;
  if (r != null)
    if (n !== void 0) {
      if (e = r, E = n, C = e * E, e % 1 !== 0 || E % 1 !== 0)
        throw Fr();
    } else
      switch (typeof r) {
        case "object": {
          if ("d" in r && "n" in r)
            e = r.n, E = r.d, "s" in r && (e *= r.s);
          else if (0 in r)
            e = r[0], 1 in r && (E = r[1]);
          else
            throw Au();
          C = e * E;
          break;
        }
        case "number": {
          if (r < 0 && (C = r, r = -r), r % 1 === 0)
            e = r;
          else if (r > 0) {
            for (r >= 1 && (D = Math.pow(10, Math.floor(1 + Math.log(r) / Math.LN10)), r /= D); F <= q && x <= q; )
              if (j = (a + y) / (F + x), r === j) {
                F + x <= q ? (e = a + y, E = F + x) : x > F ? (e = y, E = x) : (e = a, E = F);
                break;
              } else
                r > j ? (a += y, F += x) : (y += a, x += F), F > q ? (e = y, E = x) : (e = a, E = F);
            e *= D;
          } else
            (isNaN(r) || isNaN(n)) && (E = e = NaN);
          break;
        }
        case "string": {
          if (F = r.match(/\d+|./g), F === null)
            throw Au();
          if (F[a] === "-" ? (C = -1, a++) : F[a] === "+" && a++, F.length === a + 1 ? g = R(F[a++], C) : F[a + 1] === "." || F[a] === "." ? (F[a] !== "." && (c = R(F[a++], C)), a++, (a + 1 === F.length || F[a + 1] === "(" && F[a + 3] === ")" || F[a + 1] === "'" && F[a + 3] === "'") && (g = R(F[a], C), v = Math.pow(10, F[a].length), a++), (F[a] === "(" && F[a + 2] === ")" || F[a] === "'" && F[a + 2] === "'") && ($ = R(F[a + 1], C), D = Math.pow(10, F[a + 1].length) - 1, a += 3)) : F[a + 1] === "/" || F[a + 1] === ":" ? (g = R(F[a], C), v = R(F[a + 2], 1), a += 3) : F[a + 3] === "/" && F[a + 1] === " " && (c = R(F[a], C), g = R(F[a + 2], C), v = R(F[a + 4], 1), a += 5), F.length <= a) {
            E = v * D, C = /* void */
            e = $ + E * c + D * g;
            break;
          }
        }
        default:
          throw Au();
      }
  if (E === 0)
    throw ou();
  B.s = C < 0 ? -1 : 1, B.n = Math.abs(e), B.d = Math.abs(E);
};
function Er(r, n, e) {
  for (var E = 1; n > 0; r = r * r % e, n >>= 1)
    n & 1 && (E = E * r % e);
  return E;
}
function ar(r, n) {
  for (; n % 2 === 0; n /= 2)
    ;
  for (; n % 5 === 0; n /= 5)
    ;
  if (n === 1)
    return 0;
  for (var e = 10 % n, E = 1; e !== 1; E++)
    if (e = e * 10 % n, E > ir)
      return 0;
  return E;
}
function Cr(r, n, e) {
  for (var E = 1, C = Er(10, e, n), c = 0; c < 300; c++) {
    if (E === C)
      return c;
    E = E * 10 % n, C = C * 10 % n;
  }
  return 0;
}
function T(r, n) {
  if (!r)
    return n;
  if (!n)
    return r;
  for (; ; ) {
    if (r %= n, !r)
      return n;
    if (n %= r, !n)
      return r;
  }
}
function m(r, n) {
  if (N(r, n), this instanceof m)
    r = T(B.d, B.n), this.s = B.s, this.n = B.n / r, this.d = B.d / r;
  else
    return b(B.s * B.n, B.d);
}
var ou = function() {
  return new Error("Division by Zero");
}, Au = function() {
  return new Error("Invalid argument");
}, Fr = function() {
  return new Error("Parameters must be integer");
};
m.prototype = {
  s: 1,
  n: 0,
  d: 1,
  /**
   * Calculates the absolute value
   *
   * Ex: new Fraction(-4).abs() => 4
   **/
  abs: function() {
    return b(this.n, this.d);
  },
  /**
   * Inverts the sign of the current fraction
   *
   * Ex: new Fraction(-4).neg() => 4
   **/
  neg: function() {
    return b(-this.s * this.n, this.d);
  },
  /**
   * Adds two rational numbers
   *
   * Ex: new Fraction({n: 2, d: 3}).add("14.9") => 467 / 30
   **/
  add: function(r, n) {
    return N(r, n), b(
      this.s * this.n * B.d + B.s * this.d * B.n,
      this.d * B.d
    );
  },
  /**
   * Subtracts two rational numbers
   *
   * Ex: new Fraction({n: 2, d: 3}).add("14.9") => -427 / 30
   **/
  sub: function(r, n) {
    return N(r, n), b(
      this.s * this.n * B.d - B.s * this.d * B.n,
      this.d * B.d
    );
  },
  /**
   * Multiplies two rational numbers
   *
   * Ex: new Fraction("-17.(345)").mul(3) => 5776 / 111
   **/
  mul: function(r, n) {
    return N(r, n), b(
      this.s * B.s * this.n * B.n,
      this.d * B.d
    );
  },
  /**
   * Divides two rational numbers
   *
   * Ex: new Fraction("-17.(345)").inverse().div(3)
   **/
  div: function(r, n) {
    return N(r, n), b(
      this.s * B.s * this.n * B.d,
      this.d * B.n
    );
  },
  /**
   * Clones the actual object
   *
   * Ex: new Fraction("-17.(345)").clone()
   **/
  clone: function() {
    return b(this.s * this.n, this.d);
  },
  /**
   * Calculates the modulo of two rational numbers - a more precise fmod
   *
   * Ex: new Fraction('4.(3)').mod([7, 8]) => (13/3) % (7/8) = (5/6)
   **/
  mod: function(r, n) {
    if (isNaN(this.n) || isNaN(this.d))
      return new m(NaN);
    if (r === void 0)
      return b(this.s * this.n % this.d, 1);
    if (N(r, n), B.n === 0 && this.d === 0)
      throw ou();
    return b(
      this.s * (B.d * this.n) % (B.n * this.d),
      B.d * this.d
    );
  },
  /**
   * Calculates the fractional gcd of two rational numbers
   *
   * Ex: new Fraction(5,8).gcd(3,7) => 1/56
   */
  gcd: function(r, n) {
    return N(r, n), b(T(B.n, this.n) * T(B.d, this.d), B.d * this.d);
  },
  /**
   * Calculates the fractional lcm of two rational numbers
   *
   * Ex: new Fraction(5,8).lcm(3,7) => 15
   */
  lcm: function(r, n) {
    return N(r, n), B.n === 0 && this.n === 0 ? b(0, 1) : b(B.n * this.n, T(B.n, this.n) * T(B.d, this.d));
  },
  /**
   * Calculates the ceil of a rational number
   *
   * Ex: new Fraction('4.(3)').ceil() => (5 / 1)
   **/
  ceil: function(r) {
    return r = Math.pow(10, r || 0), isNaN(this.n) || isNaN(this.d) ? new m(NaN) : b(Math.ceil(r * this.s * this.n / this.d), r);
  },
  /**
   * Calculates the floor of a rational number
   *
   * Ex: new Fraction('4.(3)').floor() => (4 / 1)
   **/
  floor: function(r) {
    return r = Math.pow(10, r || 0), isNaN(this.n) || isNaN(this.d) ? new m(NaN) : b(Math.floor(r * this.s * this.n / this.d), r);
  },
  /**
   * Rounds a rational number
   *
   * Ex: new Fraction('4.(3)').round() => (4 / 1)
   **/
  round: function(r) {
    return r = Math.pow(10, r || 0), isNaN(this.n) || isNaN(this.d) ? new m(NaN) : b(Math.round(r * this.s * this.n / this.d), r);
  },
  /**
   * Rounds a rational number to a multiple of another rational number
   *
   * Ex: new Fraction('0.9').roundTo("1/8") => 7 / 8
   **/
  roundTo: function(r, n) {
    return N(r, n), b(this.s * Math.round(this.n * B.d / (this.d * B.n)) * B.n, B.d);
  },
  /**
   * Gets the inverse of the fraction, means numerator and denominator are exchanged
   *
   * Ex: new Fraction([-3, 4]).inverse() => -4 / 3
   **/
  inverse: function() {
    return b(this.s * this.d, this.n);
  },
  /**
   * Calculates the fraction to some rational exponent, if possible
   *
   * Ex: new Fraction(-1,2).pow(-3) => -8
   */
  pow: function(r, n) {
    if (N(r, n), B.d === 1)
      return B.s < 0 ? b(Math.pow(this.s * this.d, B.n), Math.pow(this.n, B.n)) : b(Math.pow(this.s * this.n, B.n), Math.pow(this.d, B.n));
    if (this.s < 0)
      return null;
    var e = Vu(this.n), E = Vu(this.d), C = 1, c = 1;
    for (var g in e)
      if (g !== "1") {
        if (g === "0") {
          C = 0;
          break;
        }
        if (e[g] *= B.n, e[g] % B.d === 0)
          e[g] /= B.d;
        else
          return null;
        C *= Math.pow(g, e[g]);
      }
    for (var g in E)
      if (g !== "1") {
        if (E[g] *= B.n, E[g] % B.d === 0)
          E[g] /= B.d;
        else
          return null;
        c *= Math.pow(g, E[g]);
      }
    return B.s < 0 ? b(c, C) : b(C, c);
  },
  /**
   * Check if two rational numbers are the same
   *
   * Ex: new Fraction(19.6).equals([98, 5]);
   **/
  equals: function(r, n) {
    return N(r, n), this.s * this.n * B.d === B.s * B.n * this.d;
  },
  /**
   * Check if two rational numbers are the same
   *
   * Ex: new Fraction(19.6).equals([98, 5]);
   **/
  compare: function(r, n) {
    N(r, n);
    var e = this.s * this.n * B.d - B.s * B.n * this.d;
    return (0 < e) - (e < 0);
  },
  simplify: function(r) {
    if (isNaN(this.n) || isNaN(this.d))
      return this;
    r = r || 1e-3;
    for (var n = this.abs(), e = n.toContinued(), E = 1; E < e.length; E++) {
      for (var C = b(e[E - 1], 1), c = E - 2; c >= 0; c--)
        C = C.inverse().add(e[c]);
      if (Math.abs(C.sub(n).valueOf()) < r)
        return C.mul(this.s);
    }
    return this;
  },
  /**
   * Check if two rational numbers are divisible
   *
   * Ex: new Fraction(19.6).divisible(1.5);
   */
  divisible: function(r, n) {
    return N(r, n), !(!(B.n * this.d) || this.n * B.d % (B.n * this.d));
  },
  /**
   * Returns a decimal representation of the fraction
   *
   * Ex: new Fraction("100.'91823'").valueOf() => 100.91823918239183
   **/
  valueOf: function() {
    return this.s * this.n / this.d;
  },
  /**
   * Returns a string-fraction representation of a Fraction object
   *
   * Ex: new Fraction("1.'3'").toFraction(true) => "4 1/3"
   **/
  toFraction: function(r) {
    var n, e = "", E = this.n, C = this.d;
    return this.s < 0 && (e += "-"), C === 1 ? e += E : (r && (n = Math.floor(E / C)) > 0 && (e += n, e += " ", E %= C), e += E, e += "/", e += C), e;
  },
  /**
   * Returns a latex representation of a Fraction object
   *
   * Ex: new Fraction("1.'3'").toLatex() => "\frac{4}{3}"
   **/
  toLatex: function(r) {
    var n, e = "", E = this.n, C = this.d;
    return this.s < 0 && (e += "-"), C === 1 ? e += E : (r && (n = Math.floor(E / C)) > 0 && (e += n, E %= C), e += "\\frac{", e += E, e += "}{", e += C, e += "}"), e;
  },
  /**
   * Returns an array of continued fraction elements
   *
   * Ex: new Fraction("7/8").toContinued() => [0,1,7]
   */
  toContinued: function() {
    var r, n = this.n, e = this.d, E = [];
    if (isNaN(n) || isNaN(e))
      return E;
    do
      E.push(Math.floor(n / e)), r = n % e, n = e, e = r;
    while (n !== 1);
    return E;
  },
  /**
   * Creates a string representation of a fraction with all digits
   *
   * Ex: new Fraction("100.'91823'").toString() => "100.(91823)"
   **/
  toString: function(r) {
    var n = this.n, e = this.d;
    if (isNaN(n) || isNaN(e))
      return "NaN";
    r = r || 15;
    var E = ar(n, e), C = Cr(n, e, E), c = this.s < 0 ? "-" : "";
    if (c += n / e | 0, n %= e, n *= 10, n && (c += "."), E) {
      for (var g = C; g--; )
        c += n / e | 0, n %= e, n *= 10;
      c += "(";
      for (var g = E; g--; )
        c += n / e | 0, n %= e, n *= 10;
      c += ")";
    } else
      for (var g = r; n && g--; )
        c += n / e | 0, n %= e, n *= 10;
    return c;
  }
};
class P {
  constructor(n, e) {
    this.begin = O(n), this.end = O(e);
  }
  get spanCycles() {
    const n = [];
    var e = this.begin;
    const E = this.end, C = E.sam();
    if (e.equals(E))
      return [new P(e, E)];
    for (; E.gt(e); ) {
      if (e.sam().equals(C)) {
        n.push(new P(e, this.end));
        break;
      }
      const c = e.nextSam();
      n.push(new P(e, c)), e = c;
    }
    return n;
  }
  get duration() {
    return this.end.sub(this.begin);
  }
  cycleArc() {
    const n = this.begin.cyclePos(), e = n.add(this.duration);
    return new P(n, e);
  }
  withTime(n) {
    return new P(n(this.begin), n(this.end));
  }
  withEnd(n) {
    return new P(this.begin, n(this.end));
  }
  withCycle(n) {
    const e = this.begin.sam(), E = e.add(n(this.begin.sub(e))), C = e.add(n(this.end.sub(e)));
    return new P(E, C);
  }
  intersection(n) {
    const e = this.begin.max(n.begin), E = this.end.min(n.end);
    if (!e.gt(E) && !(e.equals(E) && (e.equals(this.end) && this.begin.lt(this.end) || e.equals(n.end) && n.begin.lt(n.end))))
      return new P(e, E);
  }
  intersection_e(n) {
    const e = this.intersection(n);
    if (e == null)
      throw "TimeSpans do not intersect";
    return e;
  }
  midpoint() {
    return this.begin.add(this.duration.div(O(2)));
  }
  equals(n) {
    return this.begin.equals(n.begin) && this.end.equals(n.end);
  }
  show() {
    return this.begin.show() + " вҶ’ " + this.end.show();
  }
}
const fr = (r) => r.filter((n) => n != null);
m.prototype.sam = function() {
  return this.floor();
};
m.prototype.nextSam = function() {
  return this.sam().add(1);
};
m.prototype.wholeCycle = function() {
  return new P(this.sam(), this.nextSam());
};
m.prototype.cyclePos = function() {
  return this.sub(this.sam());
};
m.prototype.lt = function(r) {
  return this.compare(r) < 0;
};
m.prototype.gt = function(r) {
  return this.compare(r) > 0;
};
m.prototype.lte = function(r) {
  return this.compare(r) <= 0;
};
m.prototype.gte = function(r) {
  return this.compare(r) >= 0;
};
m.prototype.eq = function(r) {
  return this.compare(r) == 0;
};
m.prototype.ne = function(r) {
  return this.compare(r) != 0;
};
m.prototype.max = function(r) {
  return this.gt(r) ? this : r;
};
m.prototype.maximum = function(...r) {
  return r = r.map((n) => new m(n)), r.reduce((n, e) => e.max(n), this);
};
m.prototype.min = function(r) {
  return this.lt(r) ? this : r;
};
m.prototype.mulmaybe = function(r) {
  return r !== void 0 ? this.mul(r) : void 0;
};
m.prototype.divmaybe = function(r) {
  return r !== void 0 ? this.div(r) : void 0;
};
m.prototype.addmaybe = function(r) {
  return r !== void 0 ? this.add(r) : void 0;
};
m.prototype.submaybe = function(r) {
  return r !== void 0 ? this.sub(r) : void 0;
};
m.prototype.show = function() {
  return this.s * this.n + "/" + this.d;
};
m.prototype.or = function(r) {
  return this.eq(0) ? r : this;
};
const cu = (r) => m(r), ru = (...r) => {
  if (r = fr(r), r.length !== 0)
    return r.reduce(
      (n, e) => n === void 0 || e === void 0 ? void 0 : n.lcm(e),
      cu(1)
    );
};
cu._original = m;
const O = cu, u0 = 3e-4, or = (r, n) => (e, E) => {
  const g = r.source_[E].options_?.ops;
  if (g)
    for (const $ of g)
      switch ($.type_) {
        case "stretch": {
          const v = ["fast", "slow"], { type: D, amount: a } = $.arguments_;
          if (!v.includes(D))
            throw new Error(`mini: stretch: type must be one of ${v.join("|")} but got ${D}`);
          e = w.reify(e)[D](n(a));
          break;
        }
        case "replicate": {
          const { amount: v } = $.arguments_;
          e = w.reify(e), e = e._repeatCycles(v)._fast(v);
          break;
        }
        case "bjorklund": {
          $.arguments_.rotation ? e = e.euclidRot(n($.arguments_.pulse), n($.arguments_.step), n($.arguments_.rotation)) : e = e.euclid(n($.arguments_.pulse), n($.arguments_.step));
          break;
        }
        case "degradeBy": {
          e = w.reify(e)._degradeByWith(w.rand.early(u0 * $.arguments_.seed), $.arguments_.amount ?? 0.5);
          break;
        }
        case "tail": {
          const v = n($.arguments_.element);
          e = e.fmap((D) => (a) => Array.isArray(D) ? [...D, a] : [D, a]).appLeft(v);
          break;
        }
        case "range": {
          const v = n($.arguments_.element);
          e = w.reify(e);
          const D = (F, y, x = 1) => Array.from(
            { length: Math.abs(y - F) / x + 1 },
            (q, j) => F < y ? F + j * x : F - j * x
          );
          e = ((F, y) => F.squeezeBind((x) => y.bind((q) => w.fastcat(...D(x, q)))))(e, v);
          break;
        }
        default:
          console.warn(`operator "${$.type_}" not implemented`);
      }
  return e;
};
function J(r, n, e, E = 0) {
  e?.(r);
  const C = (c) => J(c, n, e, E);
  switch (r.type_) {
    case "pattern": {
      const c = r.source_.map((D) => C(D)).map(or(r, C)), g = r.arguments_.alignment, $ = c.filter((D) => D.__tactus_source);
      let v;
      switch (g) {
        case "stack": {
          v = w.stack(...c), $.length && (v.tactus = ru(...$.map((D) => O(D.tactus))));
          break;
        }
        case "polymeter_slowcat": {
          v = w.stack(...c.map((D) => D._slow(D.__weight))), $.length && (v.tactus = ru(...$.map((D) => O(D.tactus))));
          break;
        }
        case "polymeter": {
          const D = r.arguments_.stepsPerCycle ? C(r.arguments_.stepsPerCycle).fmap((F) => w.Fraction(F)) : w.pure(w.Fraction(c.length > 0 ? c[0].__weight : 1)), a = c.map((F) => F.fast(D.fmap((y) => y.div(F.__weight))));
          v = w.stack(...a);
          break;
        }
        case "rand": {
          v = w.chooseInWith(w.rand.early(u0 * r.arguments_.seed).segment(1), c), $.length && (v.tactus = ru(...$.map((D) => O(D.tactus))));
          break;
        }
        case "feet": {
          v = w.fastcat(...c);
          break;
        }
        default: {
          if (r.source_.some((a) => !!a.options_?.weight)) {
            const a = r.source_.reduce(
              (F, y) => F.add(y.options_?.weight || w.Fraction(1)),
              w.Fraction(0)
            );
            v = w.timeCat(
              ...r.source_.map((F, y) => [F.options_?.weight || w.Fraction(1), c[y]])
            ), v.__weight = a, v.tactus = a, $.length && (v.tactus = v.tactus.mul(ru(...$.map((F) => O(F.tactus)))));
          } else
            v = w.sequence(...c), v.tactus = c.length;
          r.arguments_.tactus && (v.__tactus_source = !0);
        }
      }
      return $.length && (v.__tactus_source = !0), v;
    }
    case "element":
      return C(r.source_);
    case "atom": {
      if (r.source_ === "~" || r.source_ === "-")
        return w.silence;
      if (!r.location_)
        return console.warn("no location for", r), r.source_;
      const c = isNaN(Number(r.source_)) ? r.source_ : Number(r.source_);
      if (E === -1)
        return w.pure(c);
      const [g, $] = e0(n, r, E);
      return w.pure(c).withLoc(g, $);
    }
    case "stretch":
      return C(r.source_).slow(C(r.arguments_.amount));
    default:
      return console.warn(`node type "${r.type_}" not implemented -> returning silence`), w.silence;
  }
}
const e0 = (r, n, e = 0) => {
  const { start: E, end: C } = n.location_, c = r?.split("").slice(E.offset, C.offset).join(""), [g = 0, $ = 0] = c ? c.split(n.source_).map((v) => v.split("").filter((D) => D === " ").length) : [];
  return [E.offset + g + e, C.offset - $ + e];
}, tu = (r, n = 0, e = r) => {
  try {
    return nr(r);
  } catch (E) {
    const C = [E.location.start.offset + n, E.location.end.offset + n], c = e.slice(0, C[0]).split(`
`).length;
    throw new Error(`[mini] parse error at line ${c}: ${E.message}`);
  }
}, cr = (r, n, e) => {
  const E = tu(r, n, e);
  let C = [];
  return J(
    E,
    r,
    (c) => {
      c.type_ === "atom" && C.push(c);
    },
    -1
  ), C;
}, Br = (r, n = 0, e) => cr(r, n, e).map((E) => e0(r, E, n)), r0 = (...r) => {
  const n = r.map((e) => {
    const E = `"${e}"`, C = tu(E);
    return J(C, E);
  });
  return w.sequence(...n);
}, Dr = (r, n) => {
  const e = `"${r}"`, E = tu(e);
  return J(E, e, null, n);
}, lr = (r) => {
  const n = tu(r);
  return J(n, r);
};
function pr(r) {
  return typeof r == "string" ? r0(r) : w.reify(r);
}
function gr() {
  w.setStringParser(r0);
}
export {
  Y as SyntaxError,
  e0 as getLeafLocation,
  Br as getLeafLocations,
  cr as getLeaves,
  lr as h,
  Dr as m,
  r0 as mini,
  tu as mini2ast,
  gr as miniAllStrings,
  pr as minify,
  nr as parse,
  J as patternifyAST
};
