// @strudel/core/dist/index.mjs
import T from "fraction.js";
const Ot = "strudel.log";
let te = 1e3, At, Tt;
function O(e, t, n = {}) {
  let s = performance.now();
  At === e && s - Tt < te || (At = e, Tt = s, console.log(`%c${e}`, "background-color: black;color:white;border-radius:15px"), typeof document < "u" && typeof CustomEvent < "u" && document.dispatchEvent(
    new CustomEvent(Ot, {
      detail: {
        message: e,
        type: t,
        data: n
      }
    })
  ));
}
O.key = Ot;
const Bi = (e) => /^[a-gA-G][#bs]*[0-9]$/.test(e), pt = (e) => /^[a-gA-G][#bsf]*[0-9]?$/.test(e), ee = (e) => {
  if (typeof e != "string")
    return [];
  const [t, n = "", s] = e.match(/^([a-gA-G])([#bsf]*)([0-9]*)$/)?.slice(1) || [];
  return t ? [t, n, s ? Number(s) : void 0] : [];
}, ne = { c: 0, d: 2, e: 4, f: 5, g: 7, a: 9, b: 11 }, se = { "#": 1, b: -1, s: 1, f: -1 }, st = (e, t = 3) => {
  const [n, s, r = t] = ee(e);
  if (!n)
    throw new Error('not a note: "' + e + '"');
  const o = ne[n.toLowerCase()], i = s?.split("").reduce((u, a) => u + se[a], 0) || 0;
  return (Number(r) + 1) * 12 + o + i;
}, K = (e) => Math.pow(2, (e - 69) / 12) * 440, re = (e) => 12 * Math.log(e / 440) / Math.LN2 + 69, xi = (e, t) => {
  if (typeof e != "object")
    throw new Error("valueToMidi: expected object value");
  let { freq: n, note: s } = e;
  if (typeof n == "number")
    return re(n);
  if (typeof s == "string")
    return st(s);
  if (typeof s == "number")
    return s;
  if (!t)
    throw new Error("valueToMidi: expected freq or note to be set");
  return t;
}, Oi = (e, t) => (e - t) * 1e3, oe = (e) => K(typeof e == "number" ? e : st(e)), ce = ["C", "Db", "D", "Eb", "E", "F", "Gb", "G", "Ab", "A", "Bb", "B"], zi = (e) => {
  const t = Math.floor(e / 12) - 1;
  return ce[e % 12] + t;
}, rt = (e, t) => (e % t + t) % t;
function ie(e, t = 0) {
  return isNaN(Number(e)) ? (O(`"${e}" is not a number, falling back to ${t}`, "warning"), t) : e;
}
const Mi = (e, t) => rt(Math.round(ie(e ?? 0, 0)), t), ji = (e) => {
  let { value: t, context: n } = e, s = t;
  if (typeof s == "object" && !Array.isArray(s) && (s = s.note || s.n || s.value, s === void 0))
    throw new Error(`cannot find a playable note for ${JSON.stringify(t)}`);
  if (typeof s == "number" && n.type !== "frequency")
    s = K(e.value);
  else if (typeof s == "number" && n.type === "frequency")
    s = e.value;
  else if (typeof s != "string" || !pt(s))
    throw new Error("not a note: " + JSON.stringify(s));
  return s;
}, Ei = (e) => {
  let { value: t, context: n } = e;
  if (typeof t == "object")
    return t.freq ? t.freq : oe(t.note || t.n || t.value);
  if (typeof t == "number" && n.type !== "frequency")
    t = K(e.value);
  else if (typeof t == "string" && pt(t))
    t = K(st(e.value));
  else if (typeof t != "number")
    throw new Error("not a note or frequency: " + t);
  return t;
}, ue = (e, t) => e.slice(t).concat(e.slice(0, t)), ae = (...e) => e.reduce(
  (t, n) => (...s) => t(n(...s)),
  (t) => t
), Ji = (...e) => ae(...e.reverse()), Y = (e) => e.filter((t) => t != null), V = (e) => [].concat(...e), U = (e) => e, Ni = (e, t) => e, zt = (e, t) => Array.from({ length: t - e + 1 }, (n, s) => s + e);
function _(e, t, n = e.length) {
  const s = function r(...o) {
    if (o.length >= n)
      return e.apply(this, o);
    {
      const i = function(...u) {
        return r.apply(this, o.concat(u));
      };
      return t && t(i, o), i;
    }
  };
  return t && t(s, []), s;
}
function Mt(e) {
  const t = Number(e);
  if (!isNaN(t))
    return t;
  if (pt(e))
    return st(e);
  throw new Error(`cannot parse as numeral: "${e}"`);
}
function jt(e, t) {
  return (...n) => e(...n.map(t));
}
function E(e) {
  return jt(e, Mt);
}
function le(e) {
  const t = Number(e);
  if (!isNaN(t))
    return t;
  const n = {
    pi: Math.PI,
    w: 1,
    h: 0.5,
    q: 0.25,
    e: 0.125,
    s: 0.0625,
    t: 1 / 3,
    f: 0.2,
    x: 1 / 6
  }[e];
  if (typeof n < "u")
    return n;
  throw new Error(`cannot parse as fractional: "${e}"`);
}
const $i = (e) => jt(e, le), Et = function(e, t) {
  return [t.slice(0, e), t.slice(e)];
}, Jt = (e, t, n) => t.map((s, r) => e(s, n[r])), fe = function(e) {
  const t = [];
  for (let n = 0; n < e.length - 1; ++n)
    t.push([e[n], e[n + 1]]);
  return t;
}, he = (e, t, n) => Math.min(Math.max(e, t), n), pe = ["Do", "Reb", "Re", "Mib", "Mi", "Fa", "Solb", "Sol", "Lab", "La", "Sib", "Si"], de = [
  "Sa",
  "Re",
  "Ga",
  "Ma",
  "Pa",
  "Dha",
  "Ni"
], me = ["C", "Db", "D", "Eb", "E", "F", "Gb", "G", "Ab", "A", "Hb", "H"], ye = [
  "Ni",
  "Pab",
  "Pa",
  "Voub",
  "Vou",
  "Ga",
  "Dib",
  "Di",
  "Keb",
  "Ke",
  "Zob",
  "Zo"
], ge = [
  "I",
  "Ro",
  "Ha",
  "Ni",
  "Ho",
  "He",
  "To"
], we = ["C", "Db", "D", "Eb", "E", "F", "Gb", "G", "Ab", "A", "Bb", "B"], Ri = (e, t = "letters") => {
  const s = (t === "solfeggio" ? pe : t === "indian" ? de : t === "german" ? me : t === "byzantine" ? ye : t === "japanese" ? ge : we)[e % 12], r = Math.floor(e / 12) - 1;
  return s + r;
};
function Pi(e) {
  var t = {};
  return e.filter(function(n) {
    return t.hasOwn(n) ? !1 : t[n] = !0;
  });
}
function Wi(e) {
  return e.sort().filter(function(t, n, s) {
    return !n || t != s[n - 1];
  });
}
function be(e) {
  return e.sort((t, n) => t.compare(n)).filter(function(t, n, s) {
    return !n || t.ne(s[n - 1]);
  });
}
function _e(e) {
  const t = new TextEncoder().encode(e);
  return btoa(String.fromCharCode(...t));
}
function ve(e) {
  const t = new Uint8Array(
    atob(e).split("").map((s) => s.charCodeAt(0))
  );
  return new TextDecoder().decode(t);
}
function Li(e) {
  return encodeURIComponent(_e(e));
}
function Ii(e) {
  return ve(decodeURIComponent(e));
}
function ke(e, t) {
  return Array.isArray(e) ? e.map(t) : Object.fromEntries(Object.entries(e).map(([n, s], r) => [n, t(s, n, r)]));
}
T.prototype.sam = function() {
  return this.floor();
};
T.prototype.nextSam = function() {
  return this.sam().add(1);
};
T.prototype.wholeCycle = function() {
  return new S(this.sam(), this.nextSam());
};
T.prototype.cyclePos = function() {
  return this.sub(this.sam());
};
T.prototype.lt = function(e) {
  return this.compare(e) < 0;
};
T.prototype.gt = function(e) {
  return this.compare(e) > 0;
};
T.prototype.lte = function(e) {
  return this.compare(e) <= 0;
};
T.prototype.gte = function(e) {
  return this.compare(e) >= 0;
};
T.prototype.eq = function(e) {
  return this.compare(e) == 0;
};
T.prototype.ne = function(e) {
  return this.compare(e) != 0;
};
T.prototype.max = function(e) {
  return this.gt(e) ? this : e;
};
T.prototype.maximum = function(...e) {
  return e = e.map((t) => new T(t)), e.reduce((t, n) => n.max(t), this);
};
T.prototype.min = function(e) {
  return this.lt(e) ? this : e;
};
T.prototype.mulmaybe = function(e) {
  return e !== void 0 ? this.mul(e) : void 0;
};
T.prototype.divmaybe = function(e) {
  return e !== void 0 ? this.div(e) : void 0;
};
T.prototype.addmaybe = function(e) {
  return e !== void 0 ? this.add(e) : void 0;
};
T.prototype.submaybe = function(e) {
  return e !== void 0 ? this.sub(e) : void 0;
};
T.prototype.show = function() {
  return this.s * this.n + "/" + this.d;
};
T.prototype.or = function(e) {
  return this.eq(0) ? e : this;
};
const d = (e) => T(e), qe = (...e) => {
  if (e = Y(e), e.length !== 0)
    return e.reduce((t, n) => t.gcd(n), d(1));
}, Z = (...e) => {
  if (e = Y(e), e.length !== 0)
    return e.reduce(
      (t, n) => t === void 0 || n === void 0 ? void 0 : t.lcm(n),
      d(1)
    );
};
d._original = T;
class S {
  constructor(t, n) {
    this.begin = d(t), this.end = d(n);
  }
  get spanCycles() {
    const t = [];
    var n = this.begin;
    const s = this.end, r = s.sam();
    if (n.equals(s))
      return [new S(n, s)];
    for (; s.gt(n); ) {
      if (n.sam().equals(r)) {
        t.push(new S(n, this.end));
        break;
      }
      const o = n.nextSam();
      t.push(new S(n, o)), n = o;
    }
    return t;
  }
  get duration() {
    return this.end.sub(this.begin);
  }
  cycleArc() {
    const t = this.begin.cyclePos(), n = t.add(this.duration);
    return new S(t, n);
  }
  withTime(t) {
    return new S(t(this.begin), t(this.end));
  }
  withEnd(t) {
    return new S(this.begin, t(this.end));
  }
  withCycle(t) {
    const n = this.begin.sam(), s = n.add(t(this.begin.sub(n))), r = n.add(t(this.end.sub(n)));
    return new S(s, r);
  }
  intersection(t) {
    const n = this.begin.max(t.begin), s = this.end.min(t.end);
    if (!n.gt(s) && !(n.equals(s) && (n.equals(this.end) && this.begin.lt(this.end) || n.equals(t.end) && t.begin.lt(t.end))))
      return new S(n, s);
  }
  intersection_e(t) {
    const n = this.intersection(t);
    if (n == null)
      throw "TimeSpans do not intersect";
    return n;
  }
  midpoint() {
    return this.begin.add(this.duration.div(d(2)));
  }
  equals(t) {
    return this.begin.equals(t.begin) && this.end.equals(t.end);
  }
  show() {
    return this.begin.show() + " → " + this.end.show();
  }
}
class C {
  /*
        Event class, representing a value active during the timespan
        'part'. This might be a fragment of an event, in which case the
        timespan will be smaller than the 'whole' timespan, otherwise the
        two timespans will be the same. The 'part' must never extend outside of the
        'whole'. If the event represents a continuously changing value
        then the whole will be returned as None, in which case the given
        value will have been sampled from the point halfway between the
        start and end of the 'part' timespan.
        The context is to store a list of source code locations causing the event.
  
        The word 'Event' is more or less a reserved word in javascript, hence this
        class is named called 'Hap'.
        */
  constructor(t, n, s, r = {}, o = !1) {
    this.whole = t, this.part = n, this.value = s, this.context = r, this.stateful = o, o && console.assert(typeof this.value == "function", "Stateful values must be functions");
  }
  get duration() {
    let t;
    return typeof this.value?.duration == "number" ? t = d(this.value.duration) : t = this.whole.end.sub(this.whole.begin), typeof this.value?.clip == "number" ? t.mul(this.value.clip) : t;
  }
  get endClipped() {
    return this.whole.begin.add(this.duration);
  }
  isActive(t) {
    return this.whole.begin <= t && this.endClipped >= t;
  }
  isInPast(t) {
    return t > this.endClipped;
  }
  isInNearPast(t, n) {
    return n - t <= this.endClipped;
  }
  isInFuture(t) {
    return t < this.whole.begin;
  }
  isInNearFuture(t, n) {
    return n < this.whole.begin && n > this.whole.begin - t;
  }
  isWithinTime(t, n) {
    return this.whole.begin <= n && this.endClipped >= t;
  }
  wholeOrPart() {
    return this.whole ? this.whole : this.part;
  }
  withSpan(t) {
    const n = this.whole ? t(this.whole) : void 0;
    return new C(n, t(this.part), this.value, this.context);
  }
  withValue(t) {
    return new C(this.whole, this.part, t(this.value), this.context);
  }
  hasOnset() {
    return this.whole != null && this.whole.begin.equals(this.part.begin);
  }
  hasTag(t) {
    return this.context.tags?.includes(t);
  }
  resolveState(t) {
    if (this.stateful && this.hasOnset()) {
      console.log("stateful");
      const n = this.value, [s, r] = n(t);
      return [s, new C(this.whole, this.part, r, this.context, !1)];
    }
    return [t, this];
  }
  spanEquals(t) {
    return this.whole == null && t.whole == null || this.whole.equals(t.whole);
  }
  equals(t) {
    return this.spanEquals(t) && this.part.equals(t.part) && // TODO would == be better ??
    this.value === t.value;
  }
  show(t = !1) {
    const n = typeof this.value == "object" ? t ? JSON.stringify(this.value).slice(1, -1).replaceAll('"', "").replaceAll(",", " ") : JSON.stringify(this.value) : this.value;
    var s = "";
    if (this.whole == null)
      s = "~" + this.part.show;
    else {
      var r = this.whole.begin.equals(this.part.begin) && this.whole.end.equals(this.part.end);
      this.whole.begin.equals(this.part.begin) || (s = this.whole.begin.show() + " ⇜ "), r || (s += "("), s += this.part.show(), r || (s += ")"), this.whole.end.equals(this.part.end) || (s += " ⇝ " + this.whole.end.show());
    }
    return "[ " + s + " | " + n + " ]";
  }
  showWhole(t = !1) {
    return `${this.whole == null ? "~" : this.whole.show()}: ${typeof this.value == "object" ? t ? JSON.stringify(this.value).slice(1, -1).replaceAll('"', "").replaceAll(",", " ") : JSON.stringify(this.value) : this.value}`;
  }
  combineContext(t) {
    const n = this;
    return { ...n.context, ...t.context, locations: (n.context.locations || []).concat(t.context.locations || []) };
  }
  setContext(t) {
    return new C(this.whole, this.part, this.value, t);
  }
  ensureObjectValue() {
    if (typeof this.value != "object")
      throw new Error(
        `expected hap.value to be an object, but got "${this.value}". Hint: append .note() or .s() to the end`,
        "error"
      );
  }
}
class X {
  constructor(t, n = {}) {
    this.span = t, this.controls = n;
  }
  // Returns new State with different span
  setSpan(t) {
    return new X(t, this.controls);
  }
  withSpan(t) {
    return this.setSpan(t(this.span));
  }
  // Returns new State with different controls
  setControls(t) {
    return new X(this.span, t);
  }
}
function Ae(e, t, n) {
  if (t?.value !== void 0 && Object.keys(t).length === 1)
    return O("[warn]: Can't do arithmetic on control pattern."), e;
  const s = Object.keys(e).filter((r) => Object.keys(t).includes(r));
  return Object.assign({}, e, t, Object.fromEntries(s.map((r) => [r, n(e[r], t[r])])));
}
_((e, t) => e * t);
_((e, t) => t.map(e));
function Te(e, t = 60) {
  let n = 0, s = d(0), r = [""], o = "";
  for (; r[0].length < t; ) {
    const i = e.queryArc(n, n + 1), u = i.filter((h) => h.hasOnset()).map((h) => h.duration), a = qe(...u), f = a.inverse();
    r = r.map((h) => h + "|"), o += "|";
    for (let h = 0; h < f; h++) {
      const [y, g] = [s, s.add(a)], w = i.filter((b) => b.whole.begin.lte(y) && b.whole.end.gte(g)), v = w.length - r.length;
      v > 0 && (r = r.concat(Array(v).fill(o))), r = r.map((b, B) => {
        const z = w[B];
        if (z) {
          const F = z.whole.begin.eq(y) ? "" + z.value : "-";
          return b + F;
        }
        return b + ".";
      }), o += ".", s = s.add(a);
    }
    n++;
  }
  return r.join(`
`);
}
let at, N = !0;
const Hi = function(e) {
  N = !!e;
}, Fi = (e) => at = e;
class p {
  /**
   * Create a pattern. As an end user, you will most likely not create a Pattern directly.
   *
   * @param {function} query - The function that maps a `State` to an array of `Hap`.
   * @noAutocomplete
   */
  constructor(t, n = void 0) {
    this.query = t, this._Pattern = !0, this.tactus = n;
  }
  get tactus() {
    return this.__tactus;
  }
  set tactus(t) {
    this.__tactus = t === void 0 ? void 0 : d(t);
  }
  setTactus(t) {
    return this.tactus = t, this;
  }
  withTactus(t) {
    return N ? new p(this.query, this.tactus === void 0 ? void 0 : t(this.tactus)) : this;
  }
  get hasTactus() {
    return this.tactus !== void 0;
  }
  //////////////////////////////////////////////////////////////////////
  // Haskell-style functor, applicative and monadic operations
  /**
   * Returns a new pattern, with the function applied to the value of
   * each hap. It has the alias `fmap`.
   * @synonyms fmap
   * @param {Function} func to to apply to the value
   * @returns Pattern
   * @example
   * "0 1 2".withValue(v => v + 10).log()
   */
  withValue(t) {
    const n = new p((s) => this.query(s).map((r) => r.withValue(t)));
    return n.tactus = this.tactus, n;
  }
  /**
   * see `withValue`
   * @noAutocomplete
   */
  fmap(t) {
    return this.withValue(t);
  }
  /**
   * Assumes 'this' is a pattern of functions, and given a function to
   * resolve wholes, applies a given pattern of values to that
   * pattern of functions.
   * @param {Function} whole_func
   * @param {Function} func
   * @noAutocomplete
   * @returns Pattern
   */
  appWhole(t, n) {
    const s = this, r = function(o) {
      const i = s.query(o), u = n.query(o), a = function(f, h) {
        const y = f.part.intersection(h.part);
        if (y != null)
          return new C(
            t(f.whole, h.whole),
            y,
            f.value(h.value),
            h.combineContext(f)
          );
      };
      return V(
        i.map((f) => Y(u.map((h) => a(f, h))))
      );
    };
    return new p(r);
  }
  /**
   * When this method is called on a pattern of functions, it matches its haps
   * with those in the given pattern of values.  A new pattern is returned, with
   * each matching value applied to the corresponding function.
   *
   * In this `_appBoth` variant, where timespans of the function and value haps
   * are not the same but do intersect, the resulting hap has a timespan of the
   * intersection. This applies to both the part and the whole timespan.
   * @param {Pattern} pat_val
   * @noAutocomplete
   * @returns Pattern
   */
  appBoth(t) {
    const n = this, s = function(o, i) {
      if (!(o == null || i == null))
        return o.intersection_e(i);
    }, r = n.appWhole(s, t);
    return N && (r.tactus = Z(t.tactus, n.tactus)), r;
  }
  /**
   * As with `appBoth`, but the `whole` timespan is not the intersection,
   * but the timespan from the function of patterns that this method is called
   * on. In practice, this means that the pattern structure, including onsets,
   * are preserved from the pattern of functions (often referred to as the left
   * hand or inner pattern).
   * @param {Pattern} pat_val
   * @noAutocomplete
   * @returns Pattern
   */
  appLeft(t) {
    const n = this, s = function(o) {
      const i = [];
      for (const u of n.query(o)) {
        const a = t.query(o.setSpan(u.wholeOrPart()));
        for (const f of a) {
          const h = u.whole, y = u.part.intersection(f.part);
          if (y) {
            const g = u.value(f.value), w = f.combineContext(u), v = new C(h, y, g, w);
            i.push(v);
          }
        }
      }
      return i;
    }, r = new p(s);
    return r.tactus = this.tactus, r;
  }
  /**
   * As with `appLeft`, but `whole` timespans are instead taken from the
   * pattern of values, i.e. structure is preserved from the right hand/outer
   * pattern.
   * @param {Pattern} pat_val
   * @noAutocomplete
   * @returns Pattern
   */
  appRight(t) {
    const n = this, s = function(o) {
      const i = [];
      for (const u of t.query(o)) {
        const a = n.query(o.setSpan(u.wholeOrPart()));
        for (const f of a) {
          const h = u.whole, y = f.part.intersection(u.part);
          if (y) {
            const g = f.value(u.value), w = u.combineContext(f), v = new C(h, y, g, w);
            i.push(v);
          }
        }
      }
      return i;
    }, r = new p(s);
    return r.tactus = t.tactus, r;
  }
  bindWhole(t, n) {
    const s = this, r = function(o) {
      const i = function(a, f) {
        return new C(
          t(a.whole, f.whole),
          f.part,
          f.value,
          Object.assign({}, a.context, f.context, {
            locations: (a.context.locations || []).concat(f.context.locations || [])
          })
        );
      }, u = function(a) {
        return n(a.value).query(o.setSpan(a.part)).map((f) => i(a, f));
      };
      return V(s.query(o).map((a) => u(a)));
    };
    return new p(r);
  }
  bind(t) {
    const n = function(s, r) {
      if (!(s == null || r == null))
        return s.intersection_e(r);
    };
    return this.bindWhole(n, t);
  }
  join() {
    return this.bind(U);
  }
  outerBind(t) {
    return this.bindWhole((n) => n, t).setTactus(this.tactus);
  }
  outerJoin() {
    return this.outerBind(U);
  }
  innerBind(t) {
    return this.bindWhole((n, s) => s, t);
  }
  innerJoin() {
    return this.innerBind(U);
  }
  // Flatterns patterns of patterns, by retriggering/resetting inner patterns at onsets of outer pattern haps
  resetJoin(t = !1) {
    const n = this;
    return new p((s) => n.discreteOnly().query(s).map((r) => r.value.late(t ? r.whole.begin : r.whole.begin.cyclePos()).query(s).map(
      (o) => new C(
        // Supports continuous haps in the inner pattern
        o.whole ? o.whole.intersection(r.whole) : void 0,
        o.part.intersection(r.part),
        o.value
      ).setContext(r.combineContext(o))
    ).filter((o) => o.part)).flat());
  }
  restartJoin() {
    return this.resetJoin(!0);
  }
  // Like the other joins above, joins a pattern of patterns of values, into a flatter
  // pattern of values. In this case it takes whole cycles of the inner pattern to fit each event
  // in the outer pattern.
  squeezeJoin() {
    const t = this;
    function n(s) {
      const r = t.discreteOnly().query(s);
      function o(u) {
        const f = u.value._focusSpan(u.wholeOrPart()).query(s.setSpan(u.part));
        function h(y, g) {
          let w;
          if (g.whole && y.whole && (w = g.whole.intersection(y.whole), !w))
            return;
          const v = g.part.intersection(y.part);
          if (!v)
            return;
          const b = g.combineContext(y);
          return new C(w, v, g.value, b);
        }
        return f.map((y) => h(u, y));
      }
      return V(r.map(o)).filter((u) => u);
    }
    return new p(n);
  }
  squeezeBind(t) {
    return this.fmap(t).squeezeJoin();
  }
  //////////////////////////////////////////////////////////////////////
  // Utility methods mainly for internal use
  /**
   * Query haps inside the given time span.
   *
   * @param {Fraction | number} begin from time
   * @param {Fraction | number} end to time
   * @returns Hap[]
   * @example
   * const pattern = sequence('a', ['b', 'c'])
   * const haps = pattern.queryArc(0, 1)
   * console.log(haps)
   * silence
   * @noAutocomplete
   */
  queryArc(t, n, s = {}) {
    try {
      return this.query(new X(new S(t, n), s));
    } catch (r) {
      return O(`[query]: ${r.message}`, "error"), [];
    }
  }
  /**
   * Returns a new pattern, with queries split at cycle boundaries. This makes
   * some calculations easier to express, as all haps are then constrained to
   * happen within a cycle.
   * @returns Pattern
   * @noAutocomplete
   */
  splitQueries() {
    const t = this, n = (s) => V(s.span.spanCycles.map((r) => t.query(s.setSpan(r))));
    return new p(n);
  }
  /**
   * Returns a new pattern, where the given function is applied to the query
   * timespan before passing it to the original pattern.
   * @param {Function} func the function to apply
   * @returns Pattern
   * @noAutocomplete
   */
  withQuerySpan(t) {
    return new p((n) => this.query(n.withSpan(t)));
  }
  withQuerySpanMaybe(t) {
    const n = this;
    return new p((s) => {
      const r = s.withSpan(t);
      return r.span ? n.query(r) : [];
    });
  }
  /**
   * As with `withQuerySpan`, but the function is applied to both the
   * begin and end time of the query timespan.
   * @param {Function} func the function to apply
   * @returns Pattern
   * @noAutocomplete
   */
  withQueryTime(t) {
    return new p((n) => this.query(n.withSpan((s) => s.withTime(t))));
  }
  /**
   * Similar to `withQuerySpan`, but the function is applied to the timespans
   * of all haps returned by pattern queries (both `part` timespans, and where
   * present, `whole` timespans).
   * @param {Function} func
   * @returns Pattern
   * @noAutocomplete
   */
  withHapSpan(t) {
    return new p((n) => this.query(n).map((s) => s.withSpan(t)));
  }
  /**
   * As with `withHapSpan`, but the function is applied to both the
   * begin and end time of the hap timespans.
   * @param {Function} func the function to apply
   * @returns Pattern
   * @noAutocomplete
   */
  withHapTime(t) {
    return this.withHapSpan((n) => n.withTime(t));
  }
  /**
   * Returns a new pattern with the given function applied to the list of haps returned by every query.
   * @param {Function} func
   * @returns Pattern
   * @noAutocomplete
   */
  withHaps(t) {
    const n = new p((s) => t(this.query(s), s));
    return n.tactus = this.tactus, n;
  }
  /**
   * As with `withHaps`, but applies the function to every hap, rather than every list of haps.
   * @param {Function} func
   * @returns Pattern
   * @noAutocomplete
   */
  withHap(t) {
    return this.withHaps((n) => n.map(t));
  }
  /**
   * Returns a new pattern with the context field set to every hap set to the given value.
   * @param {*} context
   * @returns Pattern
   * @noAutocomplete
   */
  setContext(t) {
    return this.withHap((n) => n.setContext(t));
  }
  /**
   * Returns a new pattern with the given function applied to the context field of every hap.
   * @param {Function} func
   * @returns Pattern
   * @noAutocomplete
   */
  withContext(t) {
    const n = this.withHap((s) => s.setContext(t(s.context)));
    return this.__pure !== void 0 && (n.__pure = this.__pure, n.__pure_loc = this.__pure_loc), n;
  }
  /**
   * Returns a new pattern with the context field of every hap set to an empty object.
   * @returns Pattern
   * @noAutocomplete
   */
  stripContext() {
    return this.withHap((t) => t.setContext({}));
  }
  /**
   * Returns a new pattern with the given location information added to the
   * context of every hap.
   * @param {Number} start start offset
   * @param {Number} end end offset
   * @returns Pattern
   * @noAutocomplete
   */
  withLoc(t, n) {
    const s = {
      start: t,
      end: n
    }, r = this.withContext((o) => {
      const i = (o.locations || []).concat([s]);
      return { ...o, locations: i };
    });
    return this.__pure && (r.__pure = this.__pure, r.__pure_loc = s), r;
  }
  /**
   * Returns a new Pattern, which only returns haps that meet the given test.
   * @param {Function} hap_test - a function which returns false for haps to be removed from the pattern
   * @returns Pattern
   * @noAutocomplete
   */
  filterHaps(t) {
    return new p((n) => this.query(n).filter(t));
  }
  /**
   * As with `filterHaps`, but the function is applied to values
   * inside haps.
   * @param {Function} value_test
   * @returns Pattern
   * @noAutocomplete
   */
  filterValues(t) {
    return new p((n) => this.query(n).filter((s) => t(s.value))).setTactus(this.tactus);
  }
  /**
   * Returns a new pattern, with haps containing undefined values removed from
   * query results.
   * @returns Pattern
   * @noAutocomplete
   */
  removeUndefineds() {
    return this.filterValues((t) => t != null);
  }
  /**
   * Returns a new pattern, with all haps without onsets filtered out. A hap
   * with an onset is one with a `whole` timespan that begins at the same time
   * as its `part` timespan.
   * @returns Pattern
   * @noAutocomplete
   */
  onsetsOnly() {
    return this.filterHaps((t) => t.hasOnset());
  }
  /**
   * Returns a new pattern, with 'continuous' haps (those without 'whole'
   * timespans) removed from query results.
   * @returns Pattern
   * @noAutocomplete
   */
  discreteOnly() {
    return this.filterHaps((t) => t.whole);
  }
  /**
   * Combines adjacent haps with the same value and whole.  Only
   * intended for use in tests.
   * @noAutocomplete
   */
  defragmentHaps() {
    return this.discreteOnly().withHaps((n) => {
      const s = [];
      for (var r = 0; r < n.length; ++r) {
        for (var o = !0, i = n[r]; o; ) {
          const f = JSON.stringify(n[r].value);
          for (var u = !1, a = r + 1; a < n.length; a++) {
            const h = n[a];
            if (i.whole.equals(h.whole)) {
              if (i.part.begin.eq(h.part.end)) {
                if (f === JSON.stringify(h.value)) {
                  i = new C(i.whole, new S(h.part.begin, i.part.end), i.value), n.splice(a, 1), u = !0;
                  break;
                }
              } else if (h.part.begin.eq(i.part.end) && f == JSON.stringify(h.value)) {
                i = new C(i.whole, new S(i.part.begin, h.part.end), i.value), n.splice(a, 1), u = !0;
                break;
              }
            }
          }
          o = u;
        }
        s.push(i);
      }
      return s;
    });
  }
  /**
   * Queries the pattern for the first cycle, returning Haps. Mainly of use when
   * debugging a pattern.
   * @param {Boolean} with_context - set to true, otherwise the context field
   * will be stripped from the resulting haps.
   * @returns [Hap]
   * @noAutocomplete
   */
  firstCycle(t = !1) {
    var n = this;
    return t || (n = n.stripContext()), n.query(new X(new S(d(0), d(1))));
  }
  /**
   * Accessor for a list of values returned by querying the first cycle.
   * @noAutocomplete
   */
  get firstCycleValues() {
    return this.firstCycle().map((t) => t.value);
  }
  /**
   * More human-readable version of the `firstCycleValues` accessor.
   * @noAutocomplete
   */
  get showFirstCycle() {
    return this.firstCycle().map(
      (t) => `${t.value}: ${t.whole.begin.toFraction()} - ${t.whole.end.toFraction()}`
    );
  }
  /**
   * Returns a new pattern, which returns haps sorted in temporal order. Mainly
   * of use when comparing two patterns for equality, in tests.
   * @returns Pattern
   * @noAutocomplete
   */
  sortHapsByPart() {
    return this.withHaps(
      (t) => t.sort(
        (n, s) => n.part.begin.sub(s.part.begin).or(n.part.end.sub(s.part.end)).or(n.whole.begin.sub(s.whole.begin).or(n.whole.end.sub(s.whole.end)))
      )
    );
  }
  asNumber() {
    return this.fmap(Mt);
  }
  //////////////////////////////////////////////////////////////////////
  // Operators - see 'make composers' later..
  _opIn(t, n) {
    return this.fmap(n).appLeft(m(t));
  }
  _opOut(t, n) {
    return this.fmap(n).appRight(m(t));
  }
  _opMix(t, n) {
    return this.fmap(n).appBoth(m(t));
  }
  _opSqueeze(t, n) {
    const s = m(t);
    return this.fmap((r) => s.fmap((o) => n(r)(o))).squeezeJoin();
  }
  _opSqueezeOut(t, n) {
    const s = this;
    return m(t).fmap((o) => s.fmap((i) => n(i)(o))).squeezeJoin();
  }
  _opReset(t, n) {
    return m(t).fmap((r) => this.fmap((o) => n(o)(r))).resetJoin();
  }
  _opRestart(t, n) {
    return m(t).fmap((r) => this.fmap((o) => n(o)(r))).restartJoin();
  }
  //////////////////////////////////////////////////////////////////////
  // End-user methods.
  // Those beginning with an underscore (_) are 'patternified',
  // i.e. versions are created without the underscore, that are
  // magically transformed to accept patterns for all their arguments.
  //////////////////////////////////////////////////////////////////////
  // Methods without corresponding toplevel functions
  /**
   * Layers the result of the given function(s). Like `superimpose`, but without the original pattern:
   * @name layer
   * @memberof Pattern
   * @synonyms apply
   * @returns Pattern
   * @example
   * "<0 2 4 6 ~ 4 ~ 2 0!3 ~!5>*8"
   *   .layer(x=>x.add("0,2"))
   *   .scale('C minor').note()
   */
  layer(...t) {
    return x(...t.map((n) => n(this)));
  }
  /**
   * Superimposes the result of the given function(s) on top of the original pattern:
   * @name superimpose
   * @memberof Pattern
   * @returns Pattern
   * @example
   * "<0 2 4 6 ~ 4 ~ 2 0!3 ~!5>*8"
   *   .superimpose(x=>x.add(2))
   *   .scale('C minor').note()
   */
  superimpose(...t) {
    return this.stack(...t.map((n) => n(this)));
  }
  //////////////////////////////////////////////////////////////////////
  // Multi-pattern functions
  /**
   * Stacks the given pattern(s) to the current pattern.
   * @name stack
   * @memberof Pattern
   * @example
   * s("hh*4").stack(
   *   note("c4(5,8)")
   * )
   */
  stack(...t) {
    return x(this, ...t);
  }
  sequence(...t) {
    return I(this, ...t);
  }
  /**
   * Appends the given pattern(s) to the current pattern.
   * @name seq
   * @memberof Pattern
   * @synonyms sequence, fastcat
   * @example
   * s("hh*4").seq(
   *   note("c4(5,8)")
   * )
   */
  seq(...t) {
    return I(this, ...t);
  }
  /**
   * Appends the given pattern(s) to the next cycle.
   * @name cat
   * @memberof Pattern
   * @synonyms slowcat
   * @example
   * s("hh*4").cat(
   *   note("c4(5,8)")
   * )
   */
  cat(...t) {
    return Me(this, ...t);
  }
  fastcat(...t) {
    return H(this, ...t);
  }
  slowcat(...t) {
    return D(this, ...t);
  }
  //////////////////////////////////////////////////////////////////////
  // Context methods - ones that deal with metadata
  onTrigger(t, n = !0) {
    return this.withHap(
      (s) => s.setContext({
        ...s.context,
        onTrigger: (...r) => {
          s.context.onTrigger?.(...r), t(...r);
        },
        // if dominantTrigger is set to true, the default output (webaudio) will be disabled
        // when using multiple triggers, you cannot flip this flag to false again!
        // example: x.csound('CooLSynth').log() as well as x.log().csound('CooLSynth') should work the same
        dominantTrigger: s.context.dominantTrigger || n
      })
    );
  }
  log(t = (s, r) => `[hap] ${r.showWhole(!0)}`, n = (s, r) => ({ hap: r })) {
    return this.onTrigger((...s) => {
      O(t(...s), void 0, n(...s));
    }, !1);
  }
  logValues(t = U) {
    return this.log((n, s) => t(s.value));
  }
  //////////////////////////////////////////////////////////////////////
  // Visualisation
  drawLine() {
    return console.log(Te(this)), this;
  }
}
function Ce(e, t) {
  let n = [];
  return t.forEach((s) => {
    const r = n.findIndex(([o]) => e(s, o));
    r === -1 ? n.push([s]) : n[r].push(s);
  }), n;
}
const Se = (e, t) => e.spanEquals(t);
p.prototype.collect = function() {
  return this.withHaps(
    (e) => Ce(Se, e).map((t) => new C(t[0].whole, t[0].part, t, {}))
  );
};
p.prototype.arpWith = function(e) {
  return this.collect().fmap((t) => m(e(t))).innerJoin().withHap((t) => new C(t.whole, t.part, t.value.value, t.combineContext(t.value)));
};
p.prototype.arp = function(e) {
  return this.arpWith((t) => e.fmap((n) => t[n % t.length]));
};
function et(e) {
  return !Array.isArray(e) && typeof e == "object";
}
function Be(e, t, n) {
  return et(e) || et(t) ? (et(e) || (e = { value: e }), et(t) || (t = { value: t }), Ae(e, t, n)) : n(e, t);
}
(function() {
  const e = {
    set: [(n, s) => s],
    keep: [(n) => n],
    keepif: [(n, s) => s ? n : void 0],
    // numerical functions
    /**
     *
     * Assumes a pattern of numbers. Adds the given number to each item in the pattern.
     * @name add
     * @memberof Pattern
     * @example
     * // Here, the triad 0, 2, 4 is shifted by different amounts
     * n("0 2 4".add("<0 3 4 0>")).scale("C:major")
     * // Without add, the equivalent would be:
     * // n("<[0 2 4] [3 5 7] [4 6 8] [0 2 4]>").scale("C:major")
     * @example
     * // You can also use add with notes:
     * note("c3 e3 g3".add("<0 5 7 0>"))
     * // Behind the scenes, the notes are converted to midi numbers:
     * // note("48 52 55".add("<0 5 7 0>"))
     */
    add: [E((n, s) => n + s)],
    // support string concatenation
    /**
     *
     * Like add, but the given numbers are subtracted.
     * @name sub
     * @memberof Pattern
     * @example
     * n("0 2 4".sub("<0 1 2 3>")).scale("C4:minor")
     * // See add for more information.
     */
    sub: [E((n, s) => n - s)],
    /**
     *
     * Multiplies each number by the given factor.
     * @name mul
     * @memberof Pattern
     * @example
     * "<1 1.5 [1.66, <2 2.33>]>*4".mul(150).freq()
     */
    mul: [E((n, s) => n * s)],
    /**
     *
     * Divides each number by the given factor.
     * @name div
     * @memberof Pattern
     */
    div: [E((n, s) => n / s)],
    mod: [E(rt)],
    pow: [E(Math.pow)],
    band: [E((n, s) => n & s)],
    bor: [E((n, s) => n | s)],
    bxor: [E((n, s) => n ^ s)],
    blshift: [E((n, s) => n << s)],
    brshift: [E((n, s) => n >> s)],
    // TODO - force numerical comparison if both look like numbers?
    lt: [(n, s) => n < s],
    gt: [(n, s) => n > s],
    lte: [(n, s) => n <= s],
    gte: [(n, s) => n >= s],
    eq: [(n, s) => n == s],
    eqt: [(n, s) => n === s],
    ne: [(n, s) => n != s],
    net: [(n, s) => n !== s],
    and: [(n, s) => n && s],
    or: [(n, s) => n || s],
    //  bitwise ops
    func: [(n, s) => s(n)]
  }, t = ["In", "Out", "Mix", "Squeeze", "SqueezeOut", "Reset", "Restart"];
  for (const [n, [s, r]] of Object.entries(e)) {
    p.prototype["_" + n] = function(o) {
      return this.fmap((i) => s(i, o));
    }, Object.defineProperty(p.prototype, n, {
      // a getter that returns a function, so 'pat' can be
      // accessed by closures that are methods of that function..
      get: function() {
        const o = this, i = (...u) => o[n].in(...u);
        for (const u of t)
          i[u.toLowerCase()] = function(...a) {
            var f = o;
            a = I(a), r && (f = r(f), a = r(a));
            var h;
            return n === "keepif" ? (h = f["_op" + u](a, (y) => (g) => s(y, g)), h = h.removeUndefineds()) : h = f["_op" + u](a, (y) => (g) => Be(y, g, s)), h;
          };
        return i.squeezein = i.squeeze, i;
      }
    });
    for (const o of t)
      p.prototype[o.toLowerCase()] = function(...i) {
        return this.set[o.toLowerCase()](i);
      };
  }
  p.prototype.struct = function(...n) {
    return this.keepif.out(...n);
  }, p.prototype.structAll = function(...n) {
    return this.keep.out(...n);
  }, p.prototype.mask = function(...n) {
    return this.keepif.in(...n);
  }, p.prototype.maskAll = function(...n) {
    return this.keep.in(...n);
  }, p.prototype.reset = function(...n) {
    return this.keepif.reset(...n);
  }, p.prototype.resetAll = function(...n) {
    return this.keep.reset(...n);
  }, p.prototype.restart = function(...n) {
    return this.keepif.restart(...n);
  }, p.prototype.restartAll = function(...n) {
    return this.keep.restart(...n);
  };
})();
const Vi = x, Gi = x, Di = Pt, tt = (e) => new p(() => [], e), A = tt(1), J = tt(0);
function G(e) {
  function t(s) {
    return s.span.spanCycles.map((r) => new C(d(r.begin).wholeCycle(), r, e));
  }
  const n = new p(t, 1);
  return n.__pure = e, n;
}
function Nt(e) {
  return e instanceof p || e?._Pattern;
}
function m(e) {
  return Nt(e) ? e : at && typeof e == "string" ? at(e) : G(e);
}
function x(...e) {
  e = e.map((s) => Array.isArray(s) ? I(...s) : m(s));
  const t = (s) => V(e.map((r) => r.query(s))), n = new p(t);
  return N && (n.tactus = Z(...e.map((s) => s.tactus))), n;
}
function dt(e, t) {
  if (t = t.map((o) => Array.isArray(o) ? I(...o) : m(o)), t.length === 0)
    return A;
  if (t.length === 1)
    return t[0];
  const [n, ...s] = t.map((o) => o.tactus), r = N ? n.maximum(...s) : void 0;
  return x(...e(r, t));
}
function xe(...e) {
  return dt(
    (t, n) => n.map((s) => s.tactus.eq(t) ? s : $(s, tt(t.sub(s.tactus)))),
    e
  );
}
function Oe(...e) {
  return dt(
    (t, n) => n.map((s) => s.tactus.eq(t) ? s : $(tt(t.sub(s.tactus)), s)),
    e
  );
}
function ze(...e) {
  return dt(
    (t, n) => n.map((s) => {
      if (s.tactus.eq(t))
        return s;
      const r = tt(t.sub(s.tactus).div(2));
      return $(r, s, r);
    }),
    e
  );
}
function Qi(e, ...t) {
  const [n, ...s] = t.map((i) => i.tactus), r = n.maximum(...s), o = {
    centre: ze,
    left: xe,
    right: Oe,
    expand: x,
    repeat: (...i) => Re(r, ...i)
  };
  return e.inhabit(o).fmap((i) => i(...t)).innerJoin().setTactus(r);
}
function D(...e) {
  if (e = e.map((s) => Array.isArray(s) ? H(...s) : m(s)), e.length == 1)
    return e[0];
  const t = function(s) {
    const r = s.span, o = rt(r.begin.sam(), e.length), i = e[o];
    if (!i)
      return [];
    const u = r.begin.floor().sub(r.begin.div(e.length).floor());
    return i.withHapTime((a) => a.add(u)).query(s.setSpan(r.withTime((a) => a.sub(u))));
  }, n = N ? Z(...e.map((s) => s.tactus)) : void 0;
  return new p(t).splitQueries().setTactus(n);
}
function $t(...e) {
  e = e.map(m);
  const t = function(n) {
    const s = Math.floor(n.span.begin) % e.length;
    return e[s]?.query(n) || [];
  };
  return new p(t).splitQueries();
}
function Me(...e) {
  return D(...e);
}
function Ui(...e) {
  const t = e.reduce((n, [s]) => n + s, 0);
  return e = e.map(([n, s]) => [n, s.fast(n)]), $(...e).slow(t);
}
function H(...e) {
  let t = D(...e);
  return e.length > 1 && (t = t._fast(e.length), t.tactus = e.length), t;
}
function I(...e) {
  return H(...e);
}
function je(...e) {
  return H(...e);
}
function lt(e) {
  return Array.isArray(e) ? e.length == 0 ? [A, 0] : e.length == 1 ? lt(e[0]) : [H(...e.map((t) => lt(t)[0])), e.length] : [m(e), 1];
}
const Ki = _((e, t) => m(t).mask(e)), Xi = _((e, t) => m(t).struct(e)), Yi = _((e, t) => m(t).superimpose(...e)), Zi = _((e, t) => m(t).set(e)), tu = _((e, t) => m(t).keep(e)), eu = _((e, t) => m(t).keepif(e)), nu = _((e, t) => m(t).add(e)), su = _((e, t) => m(t).sub(e)), ru = _((e, t) => m(t).mul(e)), ou = _((e, t) => m(t).div(e)), cu = _((e, t) => m(t).mod(e)), iu = _((e, t) => m(t).pow(e)), uu = _((e, t) => m(t).band(e)), au = _((e, t) => m(t).bor(e)), lu = _((e, t) => m(t).bxor(e)), fu = _((e, t) => m(t).blshift(e)), hu = _((e, t) => m(t).brshift(e)), pu = _((e, t) => m(t).lt(e)), du = _((e, t) => m(t).gt(e)), mu = _((e, t) => m(t).lte(e)), yu = _((e, t) => m(t).gte(e)), gu = _((e, t) => m(t).eq(e)), wu = _((e, t) => m(t).eqt(e)), bu = _((e, t) => m(t).ne(e)), _u = _((e, t) => m(t).net(e)), vu = _((e, t) => m(t).and(e)), ku = _((e, t) => m(t).or(e)), qu = _((e, t) => m(t).func(e));
function l(e, t, n = !0, s = !1, r = (o) => o.innerJoin()) {
  if (Array.isArray(e)) {
    const u = {};
    for (const a of e)
      u[a] = l(a, t, n);
    return u;
  }
  const o = t.length;
  var i;
  return n ? i = function(...u) {
    u = u.map(m);
    const a = u[u.length - 1];
    let f;
    if (o === 1)
      f = t(a);
    else {
      const h = u.slice(0, -1);
      if (h.every((y) => y.__pure != null)) {
        const y = h.map((w) => w.__pure), g = h.filter((w) => w.__pure_loc).map((w) => w.__pure_loc);
        f = t(...y, a), f = f.withContext((w) => {
          const v = (w.locations || []).concat(g);
          return { ...w, locations: v };
        });
      } else {
        const [y, ...g] = h;
        let w = (...v) => t(...v, a);
        w = _(w, null, o - 1), f = r(g.reduce((v, b) => v.appLeft(b), y.fmap(w)));
      }
    }
    return s && (f.tactus = a.tactus), f;
  } : i = function(...u) {
    u = u.map(m);
    const a = t(...u);
    return s && (a.tactus = u[u.length - 1].tactus), a;
  }, p.prototype[e] = function(...u) {
    if (o === 2 && u.length !== 1)
      u = [I(...u)];
    else if (o !== u.length + 1)
      throw new Error(`.${e}() expects ${o - 1} inputs but got ${u.length}.`);
    return u = u.map(m), i(...u, this);
  }, o > 1 && (p.prototype["_" + e] = function(...u) {
    return t(...u, this);
  }), _(i, null, o);
}
function ot(e, t, n = !0, s = !1, r = (o) => o.stepJoin()) {
  return l(e, t, n, s, r);
}
const Au = l("round", function(e) {
  return e.asNumber().fmap((t) => Math.round(t));
}), Tu = l("floor", function(e) {
  return e.asNumber().fmap((t) => Math.floor(t));
}), Cu = l("ceil", function(e) {
  return e.asNumber().fmap((t) => Math.ceil(t));
}), Su = l("toBipolar", function(e) {
  return e.fmap((t) => t * 2 - 1);
}), Bu = l("fromBipolar", function(e) {
  return e.fmap((t) => (t + 1) / 2);
}), xu = l("range", function(e, t, n) {
  return n.mul(t - e).add(e);
}), Ou = l("rangex", function(e, t, n) {
  return n._range(Math.log(e), Math.log(t)).fmap(Math.exp);
}), zu = l("range2", function(e, t, n) {
  return n.fromBipolar()._range(e, t);
}), Mu = l(
  "ratio",
  (e) => e.fmap((t) => Array.isArray(t) ? t.slice(1).reduce((n, s) => n / s, t[0]) : t)
), ju = l("compress", function(e, t, n) {
  return e = d(e), t = d(t), e.gt(t) || e.gt(1) || t.gt(1) || e.lt(0) || t.lt(0) ? A : n._fastGap(d(1).div(t.sub(e)))._late(e);
}), { compressSpan: Eu, compressspan: Ju } = l(["compressSpan", "compressspan"], function(e, t) {
  return t._compress(e.begin, e.end);
}), { fastGap: Nu, fastgap: $u } = l(["fastGap", "fastgap"], function(e, t) {
  const n = function(r) {
    const o = r.begin.sam(), i = r.begin.sub(o).mul(e).min(1), u = r.end.sub(o).mul(e).min(1);
    if (!(i >= 1))
      return new S(o.add(i), o.add(u));
  }, s = function(r) {
    const o = r.part.begin, i = r.part.end, u = o.sam(), a = o.sub(u).div(e).min(1), f = i.sub(u).div(e).min(1), h = new S(u.add(a), u.add(f)), y = r.whole ? new S(
      h.begin.sub(o.sub(r.whole.begin).div(e)),
      h.end.add(r.whole.end.sub(i).div(e))
    ) : void 0;
    return new C(y, h, r.value, r.context);
  };
  return t.withQuerySpanMaybe(n).withHap(s).splitQueries();
}), Ru = l("focus", function(e, t, n) {
  return e = d(e), t = d(t), n._fast(d(1).div(t.sub(e))).late(e.cyclePos());
}), { focusSpan: Pu, focusspan: Wu } = l(["focusSpan", "focusspan"], function(e, t) {
  return t._focus(e.begin, e.end);
}), Lu = l("ply", function(e, t) {
  const n = t.fmap((s) => G(s)._fast(e)).squeezeJoin();
  return N && (n.tactus = d(e).mulmaybe(t.tactus)), n;
}), { fast: Iu, density: Hu } = l(
  ["fast", "density"],
  function(e, t) {
    return e === 0 ? A : (e = d(e), t.withQueryTime((s) => s.mul(e)).withHapTime((s) => s.div(e)).setTactus(t.tactus));
  },
  !0,
  !0
), Fu = l("hurry", function(e, t) {
  return t._fast(e).mul(G({ speed: e }));
}), { slow: Vu, sparsity: Gu } = l(["slow", "sparsity"], function(e, t) {
  return e === 0 ? A : t._fast(d(1).div(e));
}), Du = l("inside", function(e, t, n) {
  return t(n._slow(e))._fast(e);
}), Qu = l("outside", function(e, t, n) {
  return t(n._fast(e))._slow(e);
}), Uu = l("lastOf", function(e, t, n) {
  const s = Array(e - 1).fill(n);
  return s.push(t(n)), $t(...s);
}), { firstOf: Ku, every: Xu } = l(["firstOf", "every"], function(e, t, n) {
  const s = Array(e - 1).fill(n);
  return s.unshift(t(n)), $t(...s);
}), Yu = l("apply", function(e, t) {
  return e(t);
}), Zu = l("cpm", function(e, t) {
  return t._fast(e / 60 / 1);
}), ta = l(
  "early",
  function(e, t) {
    return e = d(e), t.withQueryTime((n) => n.add(e)).withHapTime((n) => n.sub(e));
  },
  !0,
  !0
), Ee = l(
  "late",
  function(e, t) {
    return e = d(e), t._early(d(0).sub(e));
  },
  !0,
  !0
), ea = l("zoom", function(e, t, n) {
  if (t = d(t), e = d(e), e.gte(t))
    return J;
  const s = t.sub(e), r = N ? n.tactus.mulmaybe(s) : void 0;
  return n.withQuerySpan((o) => o.withCycle((i) => i.mul(s).add(e))).withHapSpan((o) => o.withCycle((i) => i.sub(e).div(s))).splitQueries().setTactus(r);
}), { zoomArc: na, zoomarc: sa } = l(["zoomArc", "zoomarc"], function(e, t) {
  return t.zoom(e.begin, e.end);
}), ra = l(
  "linger",
  function(e, t) {
    return e == 0 ? A : e < 0 ? t._zoom(e.add(1), 1)._slow(e) : t._zoom(0, e)._slow(e);
  },
  !0,
  !0
), oa = l("segment", function(e, t) {
  return t.struct(G(!0)._fast(e)).setTactus(e);
}), ca = l("swingBy", (e, t, n) => n.inside(t, Ee(je(0, e / 2)))), ia = l("swing", (e, t) => t.swingBy(1 / 3, e)), { invert: ua, inv: aa } = l(
  ["invert", "inv"],
  function(e) {
    return e.fmap((t) => !t);
  },
  !0,
  !0
), la = l("when", function(e, t, n) {
  return e ? t(n) : n;
}), fa = l("off", function(e, t, n) {
  return x(n, t(n.late(e)));
}), ha = l("brak", function(e) {
  return e.when(D(!1, !0), (t) => H(t, A)._late(0.25));
}), Je = l(
  "rev",
  function(e) {
    const t = function(n) {
      const s = n.span, r = s.begin.sam(), o = s.begin.nextSam(), i = function(a) {
        const f = a.withTime((y) => r.add(o.sub(y))), h = f.begin;
        return f.begin = f.end, f.end = h, f;
      };
      return e.query(n.setSpan(i(s))).map((a) => a.withSpan(i));
    };
    return new p(t).splitQueries();
  },
  !1,
  !0
), pa = l("pressBy", function(e, t) {
  return t.fmap((n) => G(n).compress(e, 1)).squeezeJoin();
}), da = l("press", function(e) {
  return e._pressBy(0.5);
});
p.prototype.hush = function() {
  return A;
};
const ma = l(
  "palindrome",
  function(e) {
    return e.lastOf(2, Je);
  },
  !0,
  !0
), { juxBy: ya, juxby: ga } = l(["juxBy", "juxby"], function(e, t, n) {
  e /= 2;
  const s = function(i, u, a) {
    return u in i ? i[u] : a;
  }, r = n.withValue((i) => Object.assign({}, i, { pan: s(i, "pan", 0.5) - e })), o = t(n.withValue((i) => Object.assign({}, i, { pan: s(i, "pan", 0.5) + e })));
  return x(r, o).setTactus(N ? Z(r.tactus, o.tactus) : void 0);
}), wa = l("jux", function(e, t) {
  return t._juxBy(1, e, t);
}), { echoWith: ba, echowith: _a, stutWith: va, stutwith: ka } = l(
  ["echoWith", "echowith", "stutWith", "stutwith"],
  function(e, t, n, s) {
    return x(...zt(0, e - 1).map((r) => n(s.late(d(t).mul(r)), r)));
  }
), qa = l("echo", function(e, t, n, s) {
  return s._echoWith(e, t, (r, o) => r.gain(Math.pow(n, o)));
}), Aa = l("stut", function(e, t, n, s) {
  return s._echoWith(e, n, (r, o) => r.gain(Math.pow(t, o)));
}), mt = function(e, t, n = !1) {
  return e = d(e), D(
    ...zt(0, e.sub(1)).map(
      (s) => n ? t.late(d(s).div(e)) : t.early(d(s).div(e))
    )
  );
}, Ta = l(
  "iter",
  function(e, t) {
    return mt(e, t, !1);
  },
  !0,
  !0
), { iterBack: Ca, iterback: Sa } = l(
  ["iterBack", "iterback"],
  function(e, t) {
    return mt(e, t, !0);
  },
  !0,
  !0
), { repeatCycles: Ba } = l(
  "repeatCycles",
  function(e, t) {
    return new p(function(n) {
      const s = n.span.begin.sam(), r = s.div(e).sam(), o = s.sub(r);
      return n = n.withSpan((i) => i.withTime((u) => u.sub(o))), t.query(n).map((i) => i.withSpan((u) => u.withTime((a) => a.add(o))));
    }).splitQueries();
  },
  !0,
  !0
), yt = function(e, t, n, s = !1, r = !1) {
  const o = Array(e - 1).fill(!1);
  o.unshift(!0);
  const i = mt(e, I(...o), !s);
  return r || (n = n.repeatCycles(e)), n.when(i, t);
}, { chunk: xa, slowchunk: Oa, slowChunk: za } = l(["chunk", "slowchunk", "slowChunk"], function(e, t, n) {
  return yt(e, t, n, !1, !1);
}), { chunkBack: Ma, chunkback: ja } = l(["chunkBack", "chunkback"], function(e, t, n) {
  return yt(e, t, n, !0);
}), { fastchunk: Ea, fastChunk: Ja } = l(["fastchunk", "fastChunk"], function(e, t, n) {
  return yt(e, t, n, !1, !0);
}), Na = l(
  "bypass",
  function(e, t) {
    return e = !!parseInt(e), e ? A : t;
  },
  !0,
  !0
), $a = l("ribbon", (e, t, n) => n.early(e).restart(G(1).slow(t))), Ra = l("hsla", (e, t, n, s, r) => r.color(`hsla(${e}turn,${t * 100}%,${n * 100}%,${s})`)), Pa = l("hsl", (e, t, n, s) => s.color(`hsl(${e}turn,${t * 100}%,${n * 100}%)`));
p.prototype.tag = function(e) {
  return this.withContext((t) => ({ ...t, tags: (t.tags || []).concat([e]) }));
};
p.prototype.stepJoin = function() {
  const e = this, t = $(...Ct(St(e.queryArc(0, 1)))).tactus, n = function(s) {
    const o = e.early(s.span.begin.sam()).query(s.setSpan(new S(d(0), d(1))));
    return $(...Ct(St(o))).query(s);
  };
  return new p(n, t);
};
function Ct(e) {
  const t = e.filter((o, i) => i.hasTactus).reduce((o, i) => o.add(i), d(0)), n = Y(e.map((o, i) => i.tactus)).reduce(
    (o, i) => o.add(i),
    d(0)
  ), s = t.eq(0) ? void 0 : n.div(t);
  function r(o, i) {
    return i.tactus === void 0 ? [o.mulmaybe(s), i] : [i.tactus, i];
  }
  return e.map((o) => r(...o));
}
function St(e) {
  const t = V(e.map((r) => [r.part.begin, r.part.end])), n = be([d(0), d(1), ...t]);
  return fe(n).map((r) => [
    r[1].sub(r[0]),
    x(...Ne(new S(...r), e).map((o) => o.value.withHap((i) => i.setContext(i.combineContext(o)))))
  ]);
}
function Ne(e, t) {
  return Y(t.map((n) => $e(e, n)));
}
function $e(e, t) {
  const n = e.intersection(t.part);
  if (n != null)
    return new C(t.whole, n, t.value, t.context);
}
const Wa = l("steps", function(e, t) {
  return t.tactus === void 0 ? t : t.tactus.eq(d(0)) ? J : t.fast(d(e).div(t.tactus));
});
function Rt(e, ...t) {
  const n = t.map((r) => lt(r));
  if (n.length == 0)
    return A;
  e == 0 && (e = n[0][1]);
  const s = [];
  for (const r of n)
    r[1] != 0 && (e == r[1] ? s.push(r[0]) : s.push(r[0]._fast(d(e).div(d(r[1])))));
  return x(...s);
}
function Re(e, ...t) {
  return t.length == 0 ? A : Array.isArray(t[0]) ? Rt(e, ...t) : Pt(...t).steps(e);
}
function Pt(...e) {
  if (Array.isArray(e[0]))
    return Rt(0, ...e);
  if (e = e.filter((o) => o.hasTactus), e.length == 0)
    return A;
  const t = e[0].tactus;
  if (t.eq(d(0)))
    return J;
  const [n, ...s] = e, r = x(n, ...s.map((o) => o._slow(o.tactus.div(t))));
  return r.tactus = t, r;
}
function $(...e) {
  if (e.length === 0)
    return J;
  const t = (i) => Array.isArray(i) ? i : [i.tactus, i];
  if (e = e.map(t), e.find((i) => i[0] === void 0)) {
    const i = e.map((a) => a[0]).filter((a) => a !== void 0);
    if (i.length === 0)
      return H(...e.map((a) => a[1]));
    if (i.length === e.length)
      return J;
    const u = i.reduce((a, f) => a.add(f), d(0)).div(i.length);
    for (let a of e)
      a[0] === void 0 && (a[0] = u);
  }
  if (e.length == 1) {
    const i = m(e[0][1]);
    return i.tactus = e[0][0], i;
  }
  const n = e.map((i) => i[0]).reduce((i, u) => i.add(u), d(0));
  let s = d(0);
  const r = [];
  for (const [i, u] of e) {
    if (d(i).eq(0))
      continue;
    const a = s.add(i);
    r.push(m(u)._compress(s.div(n), a.div(n))), s = a;
  }
  const o = x(...r);
  return o.tactus = n, o;
}
const La = $, Pe = $;
function Ia(...e) {
  e = e.map((r) => Array.isArray(r) ? r.map(m) : [m(r)]);
  const t = Z(...e.map((r) => d(r.length)));
  let n = [];
  for (let r = 0; r < t; ++r)
    n.push(...e.map((o) => o.length == 0 ? A : o[r % o.length]));
  n = n.filter((r) => r.hasTactus && r.tactus > 0);
  const s = n.reduce((r, o) => r.add(o.tactus), d(0));
  return n = $(...n), n.tactus = s, n;
}
const Ha = ot("s_add", function(e, t) {
  if (!t.hasTactus || t.tactus.lte(0) || (e = d(e), e.eq(0)))
    return J;
  const n = e < 0;
  n && (e = e.abs());
  const s = e.div(t.tactus);
  return s.lte(0) ? J : s.gte(1) ? t : n ? t.zoom(d(1).sub(s), 1) : t.zoom(0, s);
}), Fa = ot("s_sub", function(e, t) {
  return t.hasTactus ? (e = d(e), e.lt(0) ? t.s_add(d(0).sub(t.tactus.add(e))) : t.s_add(t.tactus.sub(e))) : J;
}), Va = ot("s_expand", function(e, t) {
  return t.withTactus((n) => n.mul(d(e)));
}), Ga = ot("s_contract", function(e, t) {
  return t.withTactus((n) => n.div(d(e)));
});
p.prototype.s_taperlist = function(e, t) {
  const n = this;
  if (!n.hasTactus)
    return [n];
  if (t = t - 1, t === 0)
    return [n];
  const s = [], r = e > 0;
  e = d(Math.abs(e));
  const o = n.tactus.sub(e.mul(d(t))).max(d(0));
  for (let i = 0; i < t; ++i)
    s.push(n.zoom(0, o.add(e.mul(d(i))).div(n.tactus)));
  return s.push(n), r && s.reverse(), s;
};
const Da = (e, t, n) => n.s_taperlist(e, t), Qa = l(
  "s_taper",
  function(e, t, n) {
    if (!n.hasTactus)
      return J;
    const s = n.s_taperlist(e, t), r = $(...s);
    return r.tactus = s.reduce((o, i) => o.add(i.tactus), d(0)), r;
  },
  !0,
  !1,
  (e) => e.stepJoin()
);
p.prototype.s_tour = function(...e) {
  return $(
    ...[].concat(
      ...e.map((t, n) => [...e.slice(0, e.length - n), this, ...e.slice(e.length - n)]),
      this,
      ...e
    )
  );
};
const Ua = function(e, ...t) {
  return e.s_tour(...t);
}, Ka = l("chop", function(e, t) {
  const s = Array.from({ length: e }, (o, i) => i).map((o) => ({ begin: o / e, end: (o + 1) / e })), r = function(o) {
    return I(s.map((i) => Object.assign({}, o, i)));
  };
  return t.squeezeBind(r).setTactus(N ? d(e).mulmaybe(t.tactus) : void 0);
}), Xa = l("striate", function(e, t) {
  const s = Array.from({ length: e }, (o, i) => i).map((o) => ({ begin: o / e, end: (o + 1) / e })), r = D(...s);
  return t.set(r)._fast(e).setTactus(N ? d(e).mulmaybe(t.tactus) : void 0);
}), Wt = function(e, t, n = 0.5) {
  return t.speed(1 / e * n).unit("c").slow(e);
}, We = l(
  "slice",
  function(e, t, n) {
    return e.innerBind(
      (s) => t.outerBind(
        (r) => n.outerBind((o) => {
          o = o instanceof Object ? o : { s: o };
          const i = Array.isArray(s) ? s[r] : r / s, u = Array.isArray(s) ? s[r + 1] : (r + 1) / s;
          return G({ begin: i, end: u, _slices: s, ...o });
        })
      )
    ).setTactus(t.tactus);
  },
  !1
  // turns off auto-patternification
), Ya = l(
  "splice",
  function(e, t, n) {
    const s = We(e, t, n);
    return new p((r) => {
      const o = r.controls._cps || 1;
      return s.query(r).map(
        (u) => u.withValue((a) => ({
          speed: o / a._slices / u.whole.duration * (a.speed || 1),
          unit: "c",
          ...a
        }))
      );
    }).setTactus(t.tactus);
  },
  !1
  // turns off auto-patternification
), { loopAt: Za, loopat: tl } = l(["loopAt", "loopat"], function(e, t) {
  return new p((n) => Wt(e, t, n.controls._cps).query(n));
}), el = l(
  "fit",
  (e) => e.withHaps(
    (t, n) => t.map(
      (s) => s.withValue((r) => ({
        ...r,
        speed: (n.controls._cps || 1) / s.whole.duration,
        unit: "c"
      }))
    )
  )
), { loopAtCps: nl, loopatcps: sl } = l(["loopAtCps", "loopatcps"], function(e, t, n) {
  return Wt(e, n, t);
}), rl = (e) => G(1).withValue(() => m(e())).innerJoin();
let Bt = (e) => e < 0.5 ? 1 : 1 - (e - 0.5) / 0.5, Le = (e, t, n) => {
  t = m(t), e = m(e), n = m(n);
  let s = t.fmap((o) => ({ gain: Bt(o) })), r = t.fmap((o) => ({ gain: Bt(1 - o) }));
  return x(e.mul(s), n.mul(r));
};
p.prototype.xfade = function(e, t) {
  return Le(this, e, t);
};
function gt(e) {
  let t = Array.isArray(e);
  e = t ? e : [e];
  const n = e[0], s = (i) => {
    let u;
    if (typeof i == "object" && i.value !== void 0 && (u = { ...i }, i = i.value, delete u.value), t && Array.isArray(i)) {
      const a = u || {};
      return i.forEach((f, h) => {
        h < e.length && (a[e[h]] = f);
      }), a;
    } else
      return u ? (u[n] = i, u) : { [n]: i };
  }, r = (...i) => I(...i).withValue(s), o = function(...i) {
    return i.length ? this.set(r(...i)) : this.fmap(s);
  };
  return p.prototype[n] = o, r;
}
function c(e, ...t) {
  const n = Array.isArray(e) ? e[0] : e;
  let s = {};
  return s[n] = gt(e), t.forEach((r) => {
    s[r] = s[n], p.prototype[r] = p.prototype[n];
  }), s;
}
const { s: Ie, sound: He } = c(["s", "n", "gain"], "sound"), { source: Fe, src: Ve } = c("source", "src"), { n: Ge } = c("n"), { note: De } = c(["note", "n"]), { accelerate: Qe } = c("accelerate"), { velocity: Ue } = c("velocity"), { gain: Ke } = c("gain"), { postgain: Xe } = c("postgain"), { amp: Ye } = c("amp"), { attack: Ze, att: tn } = c("attack", "att"), { fmh: en } = c(["fmh", "fmi"], "fmh"), { fmi: nn, fm: sn } = c(["fmi", "fmh"], "fm"), { fmenv: rn } = c("fmenv"), { fmattack: on } = c("fmattack"), { fmdecay: cn } = c("fmdecay"), { fmsustain: un } = c("fmsustain"), { fmrelease: an } = c("fmrelease"), { fmvelocity: ln } = c("fmvelocity"), { bank: fn } = c("bank"), { analyze: hn } = c("analyze"), { fft: pn } = c("fft"), { decay: dn, dec: mn } = c("decay", "dec"), { sustain: yn, sus: gn } = c("sustain", "sus"), { release: wn, rel: bn } = c("release", "rel"), { hold: _n } = c("hold"), { bandf: vn, bpf: kn, bp: qn } = c(["bandf", "bandq", "bpenv"], "bpf", "bp"), { bandq: An, bpq: Tn } = c("bandq", "bpq"), { begin: Cn } = c("begin"), { end: Sn } = c("end"), { loop: Bn } = c("loop"), { loopBegin: xn, loopb: On } = c("loopBegin", "loopb"), { loopEnd: zn, loope: Mn } = c("loopEnd", "loope"), { crush: jn } = c("crush"), { coarse: En } = c("coarse"), { drive: Jn } = c("drive"), { channels: Nn, ch: $n } = c("channels", "ch"), { phaserrate: Rn, phasr: Pn } = c("phaserrate", "phasr"), { phaser: Wn, ph: Ln } = c(["phaser", "phaserdepth", "phasercenter", "phasersweep"], "ph"), { phasersweep: In, phs: Hn } = c("phasersweep", "phs"), { phasercenter: Fn, phc: Vn } = c("phasercenter", "phc"), { phaserdepth: Gn, phd: Dn, phasdp: Qn } = c("phaserdepth", "phd", "phasdp"), { channel: Un } = c("channel"), { cut: Kn } = c("cut"), { cutoff: Xn, ctf: Yn, lpf: Zn, lp: ts } = c(["cutoff", "resonance", "lpenv"], "ctf", "lpf", "lp"), { lpenv: es, lpe: ns } = c("lpenv", "lpe"), { hpenv: ss, hpe: rs } = c("hpenv", "hpe"), { bpenv: os, bpe: cs } = c("bpenv", "bpe"), { lpattack: is, lpa: us } = c("lpattack", "lpa"), { hpattack: as, hpa: ls } = c("hpattack", "hpa"), { bpattack: fs, bpa: hs } = c("bpattack", "bpa"), { lpdecay: ps, lpd: ds } = c("lpdecay", "lpd"), { hpdecay: ms, hpd: ys } = c("hpdecay", "hpd"), { bpdecay: gs, bpd: ws } = c("bpdecay", "bpd"), { lpsustain: bs, lps: _s } = c("lpsustain", "lps"), { hpsustain: vs, hps: ks } = c("hpsustain", "hps"), { bpsustain: qs, bps: As } = c("bpsustain", "bps"), { lprelease: Ts, lpr: Cs } = c("lprelease", "lpr"), { hprelease: Ss, hpr: Bs } = c("hprelease", "hpr"), { bprelease: xs, bpr: Os } = c("bprelease", "bpr"), { ftype: zs } = c("ftype"), { fanchor: Ms } = c("fanchor"), { vib: js, vibrato: Es, v: Js } = c(["vib", "vibmod"], "vibrato", "v"), { noise: Ns } = c("noise"), { vibmod: $s, vmod: Rs } = c(["vibmod", "vib"], "vmod"), { hcutoff: Ps, hpf: Ws, hp: Ls } = c(["hcutoff", "hresonance", "hpenv"], "hpf", "hp"), { hresonance: Is, hpq: Hs } = c("hresonance", "hpq"), { resonance: Fs, lpq: Vs } = c("resonance", "lpq"), { djf: Gs } = c("djf"), { delay: Ds } = c(["delay", "delaytime", "delayfeedback"]), { delayfeedback: Qs, delayfb: Us, dfb: Ks } = c("delayfeedback", "delayfb", "dfb"), { delaytime: Xs, delayt: Ys, dt: Zs } = c("delaytime", "delayt", "dt"), { lock: tr } = c("lock"), { detune: er, det: nr } = c("detune", "det"), { unison: sr } = c("unison"), { spread: rr } = c("spread"), { dry: or } = c("dry"), { fadeTime: cr, fadeOutTime: ir } = c("fadeTime", "fadeOutTime"), { fadeInTime: ur } = c("fadeInTime"), { freq: ar } = c("freq"), { pattack: lr, patt: fr } = c("pattack", "patt"), { pdecay: hr, pdec: pr } = c("pdecay", "pdec"), { psustain: dr, psus: mr } = c("psustain", "psus"), { prelease: yr, prel: gr } = c("prelease", "prel"), { penv: wr } = c("penv"), { pcurve: br } = c("pcurve"), { panchor: _r } = c("panchor"), { gate: vr, gat: kr } = c("gate", "gat"), { leslie: qr } = c("leslie"), { lrate: Ar } = c("lrate"), { lsize: Tr } = c("lsize"), { activeLabel: Cr } = c("activeLabel"), { label: Sr } = c(["label", "activeLabel"]), { degree: Br } = c("degree"), { mtranspose: xr } = c("mtranspose"), { ctranspose: Or } = c("ctranspose"), { harmonic: zr } = c("harmonic"), { stepsPerOctave: Mr } = c("stepsPerOctave"), { octaveR: jr } = c("octaveR"), { nudge: Er } = c("nudge"), { octave: Jr } = c("octave"), { orbit: Nr } = c("orbit"), { overgain: $r } = c("overgain"), { overshape: Rr } = c("overshape"), { pan: Pr } = c("pan"), { panspan: Wr } = c("panspan"), { pansplay: Lr } = c("pansplay"), { panwidth: Ir } = c("panwidth"), { panorient: Hr } = c("panorient"), { rate: Fr } = c("rate"), { slide: Vr } = c("slide"), { semitone: Gr } = c("semitone"), { voice: Dr } = c("voice"), { chord: Qr } = c("chord"), { dictionary: Ur, dict: Kr } = c("dictionary", "dict"), { anchor: Xr } = c("anchor"), { offset: Yr } = c("offset"), { octaves: Zr } = c("octaves"), { mode: to } = c(["mode", "anchor"]), { room: eo } = c(["room", "size"]), { roomlp: no, rlp: so } = c("roomlp", "rlp"), { roomdim: ro, rdim: oo } = c("roomdim", "rdim"), { roomfade: co, rfade: io } = c("roomfade", "rfade"), { ir: uo, iresponse: ao } = c(["ir", "i"], "iresponse"), { roomsize: lo, size: fo, sz: ho, rsize: po } = c("roomsize", "size", "sz", "rsize"), { shape: mo } = c(["shape", "shapevol"]), { distort: yo, dist: go } = c(["distort", "distortvol"], "dist"), { compressor: wo } = c([
  "compressor",
  "compressorRatio",
  "compressorKnee",
  "compressorAttack",
  "compressorRelease"
]), { compressorKnee: bo } = c("compressorKnee"), { compressorRatio: _o } = c("compressorRatio"), { compressorAttack: vo } = c("compressorAttack"), { compressorRelease: ko } = c("compressorRelease"), { speed: qo } = c("speed"), { unit: Ao } = c("unit"), { squiz: To } = c("squiz"), { vowel: Co } = c("vowel"), { waveloss: So } = c("waveloss"), { density: Bo } = c("density"), { expression: xo } = c("expression"), { sustainpedal: Oo } = c("sustainpedal"), { tremolodepth: zo, tremdp: Mo } = c("tremolodepth", "tremdp"), { tremolorate: jo, tremr: Eo } = c("tremolorate", "tremr"), { fshift: Jo } = c("fshift"), { fshiftnote: No } = c("fshiftnote"), { fshiftphase: $o } = c("fshiftphase"), { triode: Ro } = c("triode"), { krush: Po } = c("krush"), { kcutoff: Wo } = c("kcutoff"), { octer: Lo } = c("octer"), { octersub: Io } = c("octersub"), { octersubsub: Ho } = c("octersubsub"), { ring: Fo } = c("ring"), { ringf: Vo } = c("ringf"), { ringdf: Go } = c("ringdf"), { freeze: Do } = c("freeze"), { xsdelay: Qo } = c("xsdelay"), { tsdelay: Uo } = c("tsdelay"), { real: Ko } = c("real"), { imag: Xo } = c("imag"), { enhance: Yo } = c("enhance"), { partials: Zo } = c("partials"), { comb: tc } = c("comb"), { smear: ec } = c("smear"), { scram: nc } = c("scram"), { binshift: sc } = c("binshift"), { hbrick: rc } = c("hbrick"), { lbrick: oc } = c("lbrick"), { midichan: cc } = c("midichan"), { control: ic } = c("control"), { ccn: uc } = c("ccn"), { ccv: ac } = c("ccv"), { polyTouch: lc } = c("polyTouch"), { midibend: fc } = c("midibend"), { miditouch: hc } = c("miditouch"), { ctlNum: pc } = c("ctlNum"), { frameRate: dc } = c("frameRate"), { frames: mc } = c("frames"), { hours: yc } = c("hours"), { midicmd: gc } = c("midicmd"), { minutes: wc } = c("minutes"), { progNum: bc } = c("progNum"), { seconds: _c } = c("seconds"), { songPtr: vc } = c("songPtr"), { uid: kc } = c("uid"), { val: qc } = c("val"), { cps: Ac } = c("cps"), { clip: Tc, legato: Cc } = c("clip", "legato"), { duration: Sc, dur: Bc } = c("duration", "dur"), { zrand: xc } = c("zrand"), { curve: Oc } = c("curve"), { deltaSlide: zc } = c("deltaSlide"), { pitchJump: Mc } = c("pitchJump"), { pitchJumpTime: jc } = c("pitchJumpTime"), { lfo: Ec, repeatTime: Jc } = c("lfo", "repeatTime"), { znoise: Nc } = c("znoise"), { zmod: $c } = c("zmod"), { zcrush: Rc } = c("zcrush"), { zdelay: Pc } = c("zdelay"), { tremolo: Wc } = c("tremolo"), { zzfx: Lc } = c("zzfx"), { color: Ic, colour: Hc } = c(["color", "colour"]);
let Fc = (...e) => e.reduce((t, n) => Object.assign(t, { [n]: gt(n) }), {});
const Vc = l("adsr", (e, t) => {
  e = Array.isArray(e) ? e : [e];
  const [n, s, r, o] = e;
  return t.set({ attack: n, decay: s, sustain: r, release: o });
}), Gc = l("ad", (e, t) => {
  e = Array.isArray(e) ? e : [e];
  const [n, s = n] = e;
  return t.attack(n).decay(s);
}), Dc = l("ds", (e, t) => {
  e = Array.isArray(e) ? e : [e];
  const [n, s = 0] = e;
  return t.set({ decay: n, sustain: s });
}), Qc = l("ar", (e, t) => {
  e = Array.isArray(e) ? e : [e];
  const [n, s = n] = e;
  return t.set({ attack: n, release: s });
}), ol = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  accelerate: Qe,
  activeLabel: Cr,
  ad: Gc,
  adsr: Vc,
  amp: Ye,
  analyze: hn,
  anchor: Xr,
  ar: Qc,
  att: tn,
  attack: Ze,
  bandf: vn,
  bandq: An,
  bank: fn,
  begin: Cn,
  binshift: sc,
  bp: qn,
  bpa: hs,
  bpattack: fs,
  bpd: ws,
  bpdecay: gs,
  bpe: cs,
  bpenv: os,
  bpf: kn,
  bpq: Tn,
  bpr: Os,
  bprelease: xs,
  bps: As,
  bpsustain: qs,
  ccn: uc,
  ccv: ac,
  ch: $n,
  channel: Un,
  channels: Nn,
  chord: Qr,
  clip: Tc,
  coarse: En,
  color: Ic,
  colour: Hc,
  comb: tc,
  compressor: wo,
  compressorAttack: vo,
  compressorKnee: bo,
  compressorRatio: _o,
  compressorRelease: ko,
  control: ic,
  cps: Ac,
  createParam: gt,
  createParams: Fc,
  crush: jn,
  ctf: Yn,
  ctlNum: pc,
  ctranspose: Or,
  curve: Oc,
  cut: Kn,
  cutoff: Xn,
  dec: mn,
  decay: dn,
  degree: Br,
  delay: Ds,
  delayfb: Us,
  delayfeedback: Qs,
  delayt: Ys,
  delaytime: Xs,
  deltaSlide: zc,
  density: Bo,
  det: nr,
  detune: er,
  dfb: Ks,
  dict: Kr,
  dictionary: Ur,
  dist: go,
  distort: yo,
  djf: Gs,
  drive: Jn,
  dry: or,
  ds: Dc,
  dt: Zs,
  dur: Bc,
  duration: Sc,
  end: Sn,
  enhance: Yo,
  expression: xo,
  fadeInTime: ur,
  fadeOutTime: ir,
  fadeTime: cr,
  fanchor: Ms,
  fft: pn,
  fm: sn,
  fmattack: on,
  fmdecay: cn,
  fmenv: rn,
  fmh: en,
  fmi: nn,
  fmrelease: an,
  fmsustain: un,
  fmvelocity: ln,
  frameRate: dc,
  frames: mc,
  freeze: Do,
  freq: ar,
  fshift: Jo,
  fshiftnote: No,
  fshiftphase: $o,
  ftype: zs,
  gain: Ke,
  gat: kr,
  gate: vr,
  harmonic: zr,
  hbrick: rc,
  hcutoff: Ps,
  hold: _n,
  hours: yc,
  hp: Ls,
  hpa: ls,
  hpattack: as,
  hpd: ys,
  hpdecay: ms,
  hpe: rs,
  hpenv: ss,
  hpf: Ws,
  hpq: Hs,
  hpr: Bs,
  hprelease: Ss,
  hps: ks,
  hpsustain: vs,
  hresonance: Is,
  imag: Xo,
  ir: uo,
  iresponse: ao,
  kcutoff: Wo,
  krush: Po,
  label: Sr,
  lbrick: oc,
  legato: Cc,
  leslie: qr,
  lfo: Ec,
  lock: tr,
  loop: Bn,
  loopBegin: xn,
  loopEnd: zn,
  loopb: On,
  loope: Mn,
  lp: ts,
  lpa: us,
  lpattack: is,
  lpd: ds,
  lpdecay: ps,
  lpe: ns,
  lpenv: es,
  lpf: Zn,
  lpq: Vs,
  lpr: Cs,
  lprelease: Ts,
  lps: _s,
  lpsustain: bs,
  lrate: Ar,
  lsize: Tr,
  midibend: fc,
  midichan: cc,
  midicmd: gc,
  miditouch: hc,
  minutes: wc,
  mode: to,
  mtranspose: xr,
  n: Ge,
  noise: Ns,
  note: De,
  nudge: Er,
  octave: Jr,
  octaveR: jr,
  octaves: Zr,
  octer: Lo,
  octersub: Io,
  octersubsub: Ho,
  offset: Yr,
  orbit: Nr,
  overgain: $r,
  overshape: Rr,
  pan: Pr,
  panchor: _r,
  panorient: Hr,
  panspan: Wr,
  pansplay: Lr,
  panwidth: Ir,
  partials: Zo,
  patt: fr,
  pattack: lr,
  pcurve: br,
  pdec: pr,
  pdecay: hr,
  penv: wr,
  ph: Ln,
  phasdp: Qn,
  phaser: Wn,
  phasercenter: Fn,
  phaserdepth: Gn,
  phaserrate: Rn,
  phasersweep: In,
  phasr: Pn,
  phc: Vn,
  phd: Dn,
  phs: Hn,
  pitchJump: Mc,
  pitchJumpTime: jc,
  polyTouch: lc,
  postgain: Xe,
  prel: gr,
  prelease: yr,
  progNum: bc,
  psus: mr,
  psustain: dr,
  rate: Fr,
  rdim: oo,
  real: Ko,
  registerControl: c,
  rel: bn,
  release: wn,
  repeatTime: Jc,
  resonance: Fs,
  rfade: io,
  ring: Fo,
  ringdf: Go,
  ringf: Vo,
  rlp: so,
  room: eo,
  roomdim: ro,
  roomfade: co,
  roomlp: no,
  roomsize: lo,
  rsize: po,
  s: Ie,
  scram: nc,
  seconds: _c,
  semitone: Gr,
  shape: mo,
  size: fo,
  slide: Vr,
  smear: ec,
  songPtr: vc,
  sound: He,
  source: Fe,
  speed: qo,
  spread: rr,
  squiz: To,
  src: Ve,
  stepsPerOctave: Mr,
  sus: gn,
  sustain: yn,
  sustainpedal: Oo,
  sz: ho,
  tremdp: Mo,
  tremolo: Wc,
  tremolodepth: zo,
  tremolorate: jo,
  tremr: Eo,
  triode: Ro,
  tsdelay: Uo,
  uid: kc,
  unison: sr,
  unit: Ao,
  v: Js,
  val: qc,
  velocity: Ue,
  vib: js,
  vibmod: $s,
  vibrato: Es,
  vmod: Rs,
  voice: Dr,
  vowel: Co,
  waveloss: So,
  xsdelay: Qo,
  zcrush: Rc,
  zdelay: Pc,
  zmod: $c,
  znoise: Nc,
  zrand: xc,
  zzfx: Lc
}, Symbol.toStringTag, { value: "Module" })), Uc = function(e, t) {
  const [n, s] = e, [r, o] = t, [i, u] = Et(s, r);
  return [
    [s, n - s],
    [Jt((a, f) => a.concat(f), i, o), u]
  ];
}, Kc = function(e, t) {
  const [n, s] = e, [r, o] = t, [i, u] = Et(n, o);
  return [
    [n, s - n],
    [Jt((f, h) => f.concat(h), r, i), u]
  ];
}, Lt = function(e, t) {
  const [n, s] = e;
  return Math.min(n, s) <= 1 ? [e, t] : Lt(...n > s ? Uc(e, t) : Kc(e, t));
}, Xc = function(e, t) {
  const n = e < 0;
  e = Math.abs(e);
  const s = t - e, r = Array(e).fill([1]), o = Array(s).fill([0]), i = Lt([e, s], [r, o]), u = V(i[1][0]).concat(V(i[1][1]));
  return n ? u.map((a) => a === 0 ? 1 : 0) : u;
}, wt = function(e, t, n) {
  const s = Xc(e, t);
  return n ? ue(s, -n) : s;
}, cl = l("euclid", function(e, t, n) {
  return n.struct(wt(e, t, 0));
}), { euclidrot: il, euclidRot: ul } = l(["euclidrot", "euclidRot"], function(e, t, n, s) {
  return s.struct(wt(e, t, n));
}), It = function(e, t, n, s) {
  if (e < 1)
    return A;
  const o = wt(e, t, n).join("").split("1").slice(1).map((i) => [i.length + 1, !0]);
  return s.struct(Pe(...o));
}, al = l(["euclidLegato"], function(e, t, n) {
  return It(e, t, 0, n);
}), ll = l(["euclidLegatoRot"], function(e, t, n, s) {
  return It(e, t, n, s);
});
function Yc(e, t, n = 0.05, s = 0.1, r = 0.1, o = globalThis.setInterval, i = globalThis.clearInterval, u = !0) {
  let a = 0, f = 0, h = 10 ** 4, y = 0.01;
  const g = (P) => n = P(n);
  r = r || s / 2;
  const w = () => {
    const P = e(), Q = P + s + r;
    for (f === 0 && (f = P + y); f < Q; )
      f = u ? Math.round(f * h) / h : f, t(f, n, a, P), f += n, a++;
  };
  let v;
  const b = () => {
    B(), w(), v = o(w, s * 1e3);
  }, B = () => {
    v !== void 0 && i(v), v = void 0;
  };
  return { setDuration: g, start: b, stop: () => {
    a = 0, f = 0, B();
  }, pause: () => B(), duration: n, interval: s, getPhase: () => f, minLatency: y };
}
function fl(e) {
  return new p((t) => [new C(void 0, t.span, e)]);
}
const W = (e) => {
  const t = (n) => [new C(void 0, n.span, e(n.span.midpoint()))];
  return new p(t);
}, Ht = W((e) => 1 - e % 1), Zc = Ht.toBipolar(), bt = W((e) => e % 1), ti = bt.toBipolar(), Ft = W((e) => Math.sin(Math.PI * 2 * e)), ei = Ft.fromBipolar(), hl = ei._early(d(1).div(4)), pl = Ft._early(d(1).div(4)), ni = W((e) => Math.floor(e * 2 % 2)), dl = ni.toBipolar(), ml = H(Ht, bt), yl = H(Zc, ti), si = W(U);
let _t = 0, vt = 0;
typeof window < "u" && document.addEventListener("mousemove", (e) => {
  _t = e.clientY / document.body.clientHeight, vt = e.clientX / document.body.clientWidth;
});
const gl = W(() => _t), wl = W(() => _t), bl = W(() => vt), _l = W(() => vt), ri = (e) => {
  const t = e << 13 ^ e, n = t >> 17 ^ t;
  return n << 5 ^ n;
}, oi = (e) => e - Math.trunc(e), ci = (e) => ri(Math.trunc(oi(e / 300) * 536870912)), ii = (e) => e % 536870912 / 536870912, ft = (e) => Math.abs(ii(ci(e))), vl = (e) => bt.range(0, e).floor().segment(e), R = W(ft), kl = R.toBipolar(), Vt = (e) => R.fmap((t) => t < e), ql = (e) => m(e).fmap(Vt).innerJoin(), Al = Vt(0.5), ui = (e) => R.fmap((t) => Math.trunc(t * e)), Tl = (e) => m(e).fmap(ui).innerJoin(), L = function(e, t, n = !0) {
  const s = Array.isArray(e), r = Object.keys(e).length;
  return e = ke(e, m), r === 0 ? A : t.fmap((o) => {
    let i = o;
    return s && (i = n ? Math.round(i) % r : he(Math.round(i), 0, e.length - 1)), e[i];
  });
}, ai = function(e, t) {
  return Array.isArray(t) && ([t, e] = [e, t]), li(e, t);
}, li = l("pick", function(e, t) {
  return L(e, t, !1).innerJoin();
}), fi = l("pickmod", function(e, t) {
  return L(e, t, !0).innerJoin();
}), Cl = l("pickF", function(e, t, n) {
  return n.apply(ai(e, t));
}), Sl = l("pickmodF", function(e, t, n) {
  return n.apply(fi(e, t));
}), Bl = l("pickOut", function(e, t) {
  return L(e, t, !1).outerJoin();
}), xl = l("pickmodOut", function(e, t) {
  return L(e, t, !0).outerJoin();
}), Ol = l("pickRestart", function(e, t) {
  return L(e, t, !1).restartJoin();
}), zl = l("pickmodRestart", function(e, t) {
  return L(e, t, !0).restartJoin();
}), Ml = l("pickReset", function(e, t) {
  return L(e, t, !1).resetJoin();
}), jl = l("pickmodReset", function(e, t) {
  return L(e, t, !0).resetJoin();
}), { inhabit: El, pickSqueeze: Jl } = l(["inhabit", "pickSqueeze"], function(e, t) {
  return L(e, t, !1).squeezeJoin();
}), { inhabitmod: Nl, pickmodSqueeze: $l } = l(["inhabitmod", "pickmodSqueeze"], function(e, t) {
  return L(e, t, !0).squeezeJoin();
}), Rl = (e, t) => (t = t.map(m), t.length == 0 ? A : e.fmap((n) => {
  const s = rt(Math.round(n), t.length);
  return t[s];
}).squeezeJoin()), Gt = (e, t) => (t = t.map(m), t.length == 0 ? A : e.range(0, t.length).fmap((n) => {
  const s = Math.min(Math.max(Math.floor(n), 0), t.length - 1);
  return t[s];
})), kt = (e, t) => Gt(e, t).outerJoin(), hi = (e, t) => Gt(e, t).innerJoin(), Pl = (...e) => kt(R, e);
p.prototype.choose = function(...e) {
  return kt(this, e);
};
p.prototype.choose2 = function(...e) {
  return kt(this.fromBipolar(), e);
};
const pi = (...e) => hi(R.segment(1), e), Wl = pi, Dt = function(e, ...t) {
  const n = t.map((u) => m(u[0])), s = [];
  let r = 0;
  for (const u of t)
    r += u[1], s.push(r);
  const o = r, i = function(u) {
    const a = u * o;
    return n[s.findIndex((f) => f > a, s)];
  };
  return e.fmap(i);
}, di = (...e) => Dt(...e).outerJoin(), Ll = (...e) => di(R, ...e), mi = (...e) => Dt(R.segment(1), ...e).innerJoin(), Il = mi, yi = (e) => {
  const t = e.fmap(Math.floor), n = e.fmap((o) => Math.floor(o) + 1), s = (o) => 6 * o ** 5 - 15 * o ** 4 + 10 * o ** 3, r = (o) => (i) => (u) => i + s(o) * (u - i);
  return e.sub(t).fmap(r).appBoth(t.fmap(ft)).appBoth(n.fmap(ft));
}, Hl = yi(si.fmap((e) => Number(e))), Fl = l(
  "degradeByWith",
  (e, t, n) => n.fmap((s) => (r) => s).appLeft(e.filterValues((s) => s > t))
), Vl = l("degradeBy", function(e, t) {
  return t._degradeByWith(R, e);
}), Gl = l("degrade", (e) => e._degradeBy(0.5)), Dl = l("undegradeBy", function(e, t) {
  return t._degradeByWith(
    R.fmap((n) => 1 - n),
    e
  );
}), Ql = l("undegrade", (e) => e._undegradeBy(0.5)), Ul = l("sometimesBy", function(e, t, n) {
  return m(e).fmap((s) => x(n._degradeBy(s), t(n._undegradeBy(1 - s)))).innerJoin();
}), Kl = l("sometimes", function(e, t) {
  return t._sometimesBy(0.5, e);
}), Xl = l("someCyclesBy", function(e, t, n) {
  return m(e).fmap(
    (s) => x(
      n._degradeByWith(R._segment(1), s),
      t(n._degradeByWith(R.fmap((r) => 1 - r)._segment(1), 1 - s))
    )
  ).innerJoin();
}), Yl = l("someCycles", function(e, t) {
  return t._someCyclesBy(0.5, e);
}), Zl = l("often", function(e, t) {
  return t.sometimesBy(0.75, e);
}), tf = l("rarely", function(e, t) {
  return t.sometimesBy(0.25, e);
}), ef = l("almostNever", function(e, t) {
  return t.sometimesBy(0.1, e);
}), nf = l("almostAlways", function(e, t) {
  return t.sometimesBy(0.9, e);
}), sf = l("never", function(e, t) {
  return t;
}), rf = l("always", function(e, t) {
  return e(t);
});
let nt;
try {
  nt = window?.speechSynthesis;
} catch {
  console.warn("cannot use window: not in browser?");
}
let xt = nt?.getVoices();
function gi(e, t, n) {
  nt.cancel();
  const s = new SpeechSynthesisUtterance(e);
  s.lang = t, xt = nt.getVoices();
  const r = xt.filter((o) => o.lang.includes(t));
  typeof n == "number" ? s.voice = r[n % r.length] : typeof n == "string" && (s.voice = r.find((o) => o.name === o)), speechSynthesis.speak(s);
}
const of = l("speak", function(e, t, n) {
  return n.onTrigger((s, r) => {
    gi(r.value, e, t);
  });
}), wi = async (...e) => {
  const t = await Promise.allSettled(e), n = t.filter((s) => s.status === "fulfilled").map((s) => s.value);
  return t.forEach((s, r) => {
    s.status === "rejected" && console.warn(`evalScope: module with index ${r} could not be loaded:`, s.reason);
  }), n.forEach((s) => {
    Object.entries(s).forEach(([r, o]) => {
      globalThis[r] = o;
    });
  }), n;
};
function bi(e, t = {}) {
  const { wrapExpression: n = !0, wrapAsync: s = !0 } = t;
  n && (e = `{${e}}`), s && (e = `(async ()=>${e})()`);
  const r = `"use strict";return (${e})`;
  return Function(r)();
}
const _i = async (e, t) => {
  let n = {};
  if (t) {
    const o = t(e);
    e = o.output, n = o;
  }
  return { mode: "javascript", pattern: await bi(e, { wrapExpression: !!t }), meta: n };
};
class vi {
  constructor({ onTrigger: t, onToggle: n, getTime: s }) {
    this.started = !1, this.cps = 0.5, this.lastTick = 0, this.getTime = s, this.time_at_last_tick_message = 0, this.num_cycles_at_cps_change = 0, this.onToggle = n, this.latency = 0.1, this.cycle = 0, this.id = Math.round(Date.now() * Math.random()), this.worker_time_dif, this.worker = new SharedWorker(new URL("/assets/clockworker--4w5RvGG.js", import.meta.url)), this.worker.port.start(), this.channel = new BroadcastChannel("strudeltick");
    let r = 0;
    const o = 20, i = 10 ** 3, u = (h, y, g) => {
      const w = s() - (h + y) + g;
      if (this.worker_time_dif == null)
        this.worker_time_dif = w;
      else {
        const b = Math.round((this.worker_time_dif * r + w * 1) / (r + 1) * i) / i;
        b != this.worker_time_dif && (r = 4), this.worker_time_dif = b;
      }
      r = Math.min(r + 1, o);
    }, a = (h) => {
      const {
        num_cycles_at_cps_change: y,
        cps: g,
        num_seconds_at_cps_change: w,
        num_seconds_since_cps_change: v,
        begin: b,
        end: B,
        tickdeadline: z,
        cycle: M
      } = h;
      this.cps = g, this.cycle = M, u(w, v, z), f(b, B, y, w), this.time_at_last_tick_message = this.getTime();
    }, f = (h, y, g, w) => {
      if (this.started === !1)
        return;
      this.pattern.queryArc(h, y, { _cps: this.cps }).forEach((b) => {
        if (b.hasOnset()) {
          const B = (b.whole.begin - g) / this.cps + w + this.latency + this.worker_time_dif, z = b.duration / this.cps;
          t?.(b, 0, z, this.cps, B);
        }
      });
    };
    this.channel.onmessage = (h) => {
      if (!this.started)
        return;
      const { payload: y, type: g } = h.data;
      switch (g) {
        case "tick":
          a(y);
      }
    };
  }
  sendMessage(t, n) {
    this.worker.port.postMessage({ type: t, payload: n, id: this.id });
  }
  now() {
    const t = (this.getTime() - this.time_at_last_tick_message) * this.cps;
    return this.cycle + t;
  }
  setCps(t = 1) {
    this.sendMessage("cpschange", { cps: t });
  }
  setCycle(t) {
    this.sendMessage("setcycle", { cycle: t });
  }
  setStarted(t) {
    this.sendMessage("toggle", { started: t }), this.started = t, this.onToggle?.(t);
  }
  start() {
    O("[cyclist] start"), this.setStarted(!0);
  }
  stop() {
    this.worker_time_dif = null, O("[cyclist] stop"), this.setStarted(!1);
  }
  setPattern(t, n = !1) {
    this.pattern = t, n && !this.started && this.start();
  }
  log(t, n, s) {
    const r = s.filter((o) => o.hasOnset());
    console.log(`${t.toFixed(4)} - ${n.toFixed(4)} ${Array(r.length).fill("I").join("")}`);
  }
}
class ki {
  constructor({ interval: t, onTrigger: n, onToggle: s, onError: r, getTime: o, latency: i = 0.1, setInterval: u, clearInterval: a }) {
    this.started = !1, this.cps = 0.5, this.num_ticks_since_cps_change = 0, this.lastTick = 0, this.lastBegin = 0, this.lastEnd = 0, this.getTime = o, this.num_cycles_at_cps_change = 0, this.seconds_at_cps_change, this.onToggle = s, this.latency = i, this.clock = Yc(
      o,
      // called slightly before each cycle
      (f, h, y, g) => {
        this.num_ticks_since_cps_change === 0 && (this.num_cycles_at_cps_change = this.lastEnd, this.seconds_at_cps_change = f), this.num_ticks_since_cps_change++;
        const v = this.num_ticks_since_cps_change * h * this.cps;
        try {
          const b = this.lastEnd;
          this.lastBegin = b;
          const B = this.num_cycles_at_cps_change + v;
          if (this.lastEnd = B, this.lastTick = f, f < g) {
            console.log("skip query: too late");
            return;
          }
          this.pattern.queryArc(b, B, { _cps: this.cps }).forEach((M) => {
            if (M.hasOnset()) {
              const F = (M.whole.begin - this.num_cycles_at_cps_change) / this.cps + this.seconds_at_cps_change + i, P = M.duration / this.cps, Q = F - f;
              n?.(M, Q, P, this.cps, F);
            }
          });
        } catch (b) {
          O(`[cyclist] error: ${b.message}`), r?.(b);
        }
      },
      t,
      // duration of each cycle
      0.1,
      0.1,
      u,
      a
    );
  }
  now() {
    if (!this.started)
      return 0;
    const t = this.getTime() - this.lastTick - this.clock.duration;
    return this.lastBegin + t * this.cps;
  }
  setStarted(t) {
    this.started = t, this.onToggle?.(t);
  }
  start() {
    if (this.num_ticks_since_cps_change = 0, this.num_cycles_at_cps_change = 0, !this.pattern)
      throw new Error("Scheduler: no pattern set! call .setPattern first.");
    O("[cyclist] start"), this.clock.start(), this.setStarted(!0);
  }
  pause() {
    O("[cyclist] pause"), this.clock.pause(), this.setStarted(!1);
  }
  stop() {
    O("[cyclist] stop"), this.clock.stop(), this.lastEnd = 0, this.setStarted(!1);
  }
  setPattern(t, n = !1) {
    this.pattern = t, n && !this.started && this.start();
  }
  setCps(t = 0.5) {
    this.cps !== t && (this.cps = t, this.num_ticks_since_cps_change = 0);
  }
  log(t, n, s) {
    const r = s.filter((o) => o.hasOnset());
    console.log(`${t.toFixed(4)} - ${n.toFixed(4)} ${Array(r.length).fill("I").join("")}`);
  }
}
let ht;
function cf() {
  if (!ht)
    throw new Error("no time set! use setTime to define a time source");
  return ht();
}
function qi(e) {
  ht = e;
}
function uf({
  defaultOutput: e,
  onEvalError: t,
  beforeEval: n,
  afterEval: s,
  getTime: r,
  transpiler: o,
  onToggle: i,
  editPattern: u,
  onUpdateState: a,
  sync: f = !1,
  setInterval: h,
  clearInterval: y
}) {
  const g = {
    schedulerError: void 0,
    evalError: void 0,
    code: "// LOADING",
    activeCode: "// LOADING",
    pattern: void 0,
    miniLocations: [],
    widgets: [],
    pending: !1,
    started: !1
  }, w = (k) => {
    Object.assign(g, k), g.isDirty = g.code !== g.activeCode, g.error = g.evalError || g.schedulerError, a?.(g);
  }, v = {
    onTrigger: Ai({ defaultOutput: e, getTime: r }),
    getTime: r,
    onToggle: (k) => {
      w({ started: k }), i?.(k);
    },
    setInterval: h,
    clearInterval: y
  }, b = f && typeof SharedWorker < "u" ? new vi(v) : new ki(v);
  let B = {}, z = 0, M;
  const F = function() {
    return B = {}, z = 0, M = void 0, A;
  }, P = (k, q = !0) => {
    k = u?.(k) || k, b.setPattern(k, q);
  };
  qi(() => b.now());
  const Q = () => b.stop(), Qt = () => b.start(), Ut = () => b.pause(), Kt = () => b.toggle(), ct = (k) => b.setCps(k), qt = (k) => b.setCps(k / 60), Xt = function(k) {
    return M = k, A;
  }, Yt = () => {
    p.prototype.p = function(q) {
      return q.startsWith("_") || q.endsWith("_") ? A : (q === "$" && (q = `$${z}`, z++), B[q] = this, this);
    }, p.prototype.q = function(q) {
      return A;
    };
    try {
      for (let q = 1; q < 10; ++q)
        Object.defineProperty(p.prototype, `d${q}`, {
          get() {
            return this.p(q);
          },
          configurable: !0
        }), Object.defineProperty(p.prototype, `p${q}`, {
          get() {
            return this.p(q);
          },
          configurable: !0
        }), p.prototype[`q${q}`] = A;
    } catch (q) {
      console.warn("injectPatternMethods: error:", q);
    }
    const k = l("cpm", function(q, it) {
      return it._fast(q / 60 / b.cps);
    });
    return wi({
      all: Xt,
      hush: F,
      cpm: k,
      setCps: ct,
      setcps: ct,
      setCpm: qt,
      setcpm: qt
    });
  };
  return { scheduler: b, evaluate: async (k, q = !0, it = !0) => {
    if (!k)
      throw new Error("no code to evaluate");
    try {
      w({ code: k, pending: !0 }), await Yt(), await n?.({ code: k }), it && F();
      let { pattern: j, meta: ut } = await _i(k, o);
      if (Object.keys(B).length && (j = x(...Object.values(B))), M && (j = M(j)), !Nt(j)) {
        const Zt = `got "${typeof evaluated}" instead of pattern`;
        throw new Error(Zt + (typeof evaluated == "function" ? ", did you forget to call a function?" : "."));
      }
      return O("[eval] code updated"), P(j, q), w({
        miniLocations: ut?.miniLocations || [],
        widgets: ut?.widgets || [],
        activeCode: k,
        pattern: j,
        evalError: void 0,
        schedulerError: void 0,
        pending: !1
      }), s?.({ code: k, pattern: j, meta: ut }), j;
    } catch (j) {
      O(`[eval] error: ${j.message}`, "error"), console.error(j), w({ evalError: j, pending: !1 }), t?.(j);
    }
  }, start: Qt, stop: Q, pause: Ut, setCps: ct, setPattern: P, setCode: (k) => w({ code: k }), toggle: Kt, state: g };
}
const Ai = ({ getTime: e, defaultOutput: t }) => async (n, s, r, o, i) => {
  try {
    (!n.context.onTrigger || !n.context.dominantTrigger) && await t(n, s, r, o, i), n.context.onTrigger && await n.context.onTrigger(e() + s, n, e(), o, i);
  } catch (u) {
    O(`[cyclist] error: ${u.message}`, "error");
  }
}, af = function(e, t = {}) {
  const n = document.getElementById("code"), s = "background-image:url(" + e + ");background-size:contain;";
  n.style = s;
  const { className: r } = n, o = (a, f) => {
    ({
      style: () => n.style = s + ";" + f,
      className: () => n.className = f + " " + r
    })[a]();
  }, i = Object.entries(t).filter(([a, f]) => typeof f == "function");
  Object.entries(t).filter(([a, f]) => typeof f == "string").forEach(([a, f]) => o(a, f)), i.length;
}, lf = () => {
  const e = document.getElementById("code");
  e && (e.style = "");
};
O("🌀 @strudel/core loaded 🌀");
globalThis._strudelLoaded && console.warn(
  `@strudel/core was loaded more than once...
This might happen when you have multiple versions of strudel installed. 
Please check with "npm ls @strudel/core".`
);
globalThis._strudelLoaded = !0;
export {
  ki as Cyclist,
  d as Fraction,
  C as Hap,
  p as Pattern,
  X as State,
  S as TimeSpan,
  Gt as __chooseWith,
  Vt as _brandBy,
  Ne as _fitslice,
  ui as _irand,
  $e as _match,
  rt as _mod,
  Rt as _polymeterListSteps,
  Ct as _retime,
  St as _slices,
  Qe as accelerate,
  Cr as activeLabel,
  Gc as ad,
  nu as add,
  Vc as adsr,
  nf as almostAlways,
  ef as almostNever,
  rf as always,
  Ye as amp,
  hn as analyze,
  Xr as anchor,
  vu as and,
  Yu as apply,
  Qc as ar,
  Ui as arrange,
  tn as att,
  Ze as attack,
  af as backgroundImage,
  uu as band,
  vn as bandf,
  An as bandq,
  fn as bank,
  ve as base64ToUnicode,
  Cn as begin,
  sc as binshift,
  Xc as bjork,
  fu as blshift,
  au as bor,
  qn as bp,
  hs as bpa,
  fs as bpattack,
  ws as bpd,
  gs as bpdecay,
  cs as bpe,
  os as bpenv,
  kn as bpf,
  Tn as bpq,
  Os as bpr,
  xs as bprelease,
  As as bps,
  qs as bpsustain,
  ha as brak,
  Al as brand,
  ql as brandBy,
  hu as brshift,
  lu as bxor,
  Na as bypass,
  Hi as calculateTactus,
  Me as cat,
  uc as ccn,
  ac as ccv,
  Cu as ceil,
  $n as ch,
  Un as channel,
  Nn as channels,
  Pl as choose,
  pi as chooseCycles,
  hi as chooseInWith,
  kt as chooseWith,
  Ka as chop,
  Qr as chord,
  xa as chunk,
  Ma as chunkBack,
  ja as chunkback,
  he as clamp,
  lf as cleanupUi,
  Tc as clip,
  En as coarse,
  Li as code2hash,
  Ic as color,
  Hc as colour,
  tc as comb,
  Ji as compose,
  ju as compress,
  Eu as compressSpan,
  wo as compressor,
  vo as compressorAttack,
  bo as compressorKnee,
  _o as compressorRatio,
  ko as compressorRelease,
  Ju as compressspan,
  Ni as constant,
  ic as control,
  ol as controls,
  hl as cosine,
  pl as cosine2,
  Zu as cpm,
  Ac as cps,
  Yc as createClock,
  gt as createParam,
  Fc as createParams,
  jn as crush,
  Yn as ctf,
  pc as ctlNum,
  Or as ctranspose,
  _ as curry,
  Oc as curve,
  Kn as cut,
  Xn as cutoff,
  mn as dec,
  dn as decay,
  Gl as degrade,
  Vl as degradeBy,
  Fl as degradeByWith,
  Br as degree,
  Ds as delay,
  Us as delayfb,
  Qs as delayfeedback,
  Ys as delayt,
  Xs as delaytime,
  zc as deltaSlide,
  nr as det,
  er as detune,
  Ks as dfb,
  Kr as dict,
  Ur as dictionary,
  go as dist,
  yo as distort,
  ou as div,
  Gs as djf,
  Te as drawLine,
  Jn as drive,
  or as dry,
  Dc as ds,
  Zs as dt,
  Bc as dur,
  Sc as duration,
  ta as early,
  qa as echo,
  ba as echoWith,
  _a as echowith,
  Sn as end,
  Yo as enhance,
  gu as eq,
  wu as eqt,
  cl as euclid,
  al as euclidLegato,
  ll as euclidLegatoRot,
  ul as euclidRot,
  il as euclidrot,
  wi as evalScope,
  _i as evaluate,
  Xu as every,
  xo as expression,
  ur as fadeInTime,
  ir as fadeOutTime,
  cr as fadeTime,
  Ms as fanchor,
  Iu as fast,
  Ja as fastChunk,
  Nu as fastGap,
  H as fastcat,
  Ea as fastchunk,
  $u as fastgap,
  pn as fft,
  Ku as firstOf,
  el as fit,
  V as flatten,
  Tu as floor,
  sn as fm,
  on as fmattack,
  cn as fmdecay,
  rn as fmenv,
  en as fmh,
  nn as fmi,
  an as fmrelease,
  un as fmsustain,
  ln as fmvelocity,
  Ru as focus,
  Pu as focusSpan,
  Wu as focusspan,
  $i as fractionalArgs,
  dc as frameRate,
  mc as frames,
  Do as freeze,
  ar as freq,
  re as freqToMidi,
  Bu as fromBipolar,
  Jo as fshift,
  No as fshiftnote,
  $o as fshiftphase,
  zs as ftype,
  qu as func,
  Ke as gain,
  tt as gap,
  kr as gat,
  vr as gate,
  Oi as getEventOffsetMs,
  oe as getFreq,
  Ei as getFrequency,
  ji as getPlayableNoteValue,
  Mi as getSoundIndex,
  cf as getTime,
  Ai as getTrigger,
  du as gt,
  yu as gte,
  zr as harmonic,
  Ii as hash2code,
  rc as hbrick,
  Ps as hcutoff,
  _n as hold,
  yc as hours,
  Ls as hp,
  ls as hpa,
  as as hpattack,
  ys as hpd,
  ms as hpdecay,
  rs as hpe,
  ss as hpenv,
  Ws as hpf,
  Hs as hpq,
  Bs as hpr,
  Ss as hprelease,
  ks as hps,
  vs as hpsustain,
  Is as hresonance,
  Pa as hsl,
  Ra as hsla,
  Fu as hurry,
  U as id,
  Xo as imag,
  El as inhabit,
  Nl as inhabitmod,
  Du as inside,
  aa as inv,
  ua as invert,
  uo as ir,
  Tl as irand,
  ao as iresponse,
  pt as isNote,
  Bi as isNoteWithOctave,
  Nt as isPattern,
  Ht as isaw,
  Zc as isaw2,
  Ta as iter,
  Ca as iterBack,
  Sa as iterback,
  wa as jux,
  ya as juxBy,
  ga as juxby,
  Wo as kcutoff,
  tu as keep,
  eu as keepif,
  Po as krush,
  Sr as label,
  Uu as lastOf,
  Ee as late,
  oc as lbrick,
  Cc as legato,
  qr as leslie,
  Ec as lfo,
  ra as linger,
  zt as listRange,
  tr as lock,
  Ot as logKey,
  O as logger,
  Bn as loop,
  Za as loopAt,
  nl as loopAtCps,
  xn as loopBegin,
  zn as loopEnd,
  tl as loopat,
  sl as loopatcps,
  On as loopb,
  Mn as loope,
  ts as lp,
  us as lpa,
  is as lpattack,
  ds as lpd,
  ps as lpdecay,
  ns as lpe,
  es as lpenv,
  Zn as lpf,
  Vs as lpq,
  Cs as lpr,
  Ts as lprelease,
  _s as lps,
  bs as lpsustain,
  Ar as lrate,
  Tr as lsize,
  pu as lt,
  mu as lte,
  jt as mapArgs,
  Ki as mask,
  zi as midi2note,
  K as midiToFreq,
  fc as midibend,
  cc as midichan,
  gc as midicmd,
  hc as miditouch,
  wc as minutes,
  cu as mod,
  to as mode,
  _l as mouseX,
  wl as mouseY,
  bl as mousex,
  gl as mousey,
  xr as mtranspose,
  ru as mul,
  Ge as n,
  ie as nanFallback,
  bu as ne,
  _u as net,
  sf as never,
  Ns as noise,
  De as note,
  st as noteToMidi,
  J as nothing,
  Er as nudge,
  E as numeralArgs,
  ke as objectMap,
  Jr as octave,
  jr as octaveR,
  Zr as octaves,
  Lo as octer,
  Io as octersub,
  Ho as octersubsub,
  fa as off,
  Yr as offset,
  Zl as often,
  ku as or,
  Nr as orbit,
  Qu as outside,
  $r as overgain,
  Rr as overshape,
  fe as pairs,
  ma as palindrome,
  Pr as pan,
  _r as panchor,
  Hr as panorient,
  Wr as panspan,
  Lr as pansplay,
  Ir as panwidth,
  le as parseFractional,
  Mt as parseNumeral,
  Zo as partials,
  fr as patt,
  lr as pattack,
  br as pcurve,
  pr as pdec,
  hr as pdecay,
  wr as penv,
  Hl as perlin,
  yi as perlinWith,
  Ln as ph,
  Qn as phasdp,
  Wn as phaser,
  Fn as phasercenter,
  Gn as phaserdepth,
  Rn as phaserrate,
  In as phasersweep,
  Pn as phasr,
  Vn as phc,
  Dn as phd,
  Hn as phs,
  ai as pick,
  Cl as pickF,
  Bl as pickOut,
  Ml as pickReset,
  Ol as pickRestart,
  Jl as pickSqueeze,
  fi as pickmod,
  Sl as pickmodF,
  xl as pickmodOut,
  jl as pickmodReset,
  zl as pickmodRestart,
  $l as pickmodSqueeze,
  ae as pipe,
  Mc as pitchJump,
  jc as pitchJumpTime,
  Lu as ply,
  Di as pm,
  lc as polyTouch,
  Vi as polyrhythm,
  Xe as postgain,
  iu as pow,
  Gi as pr,
  gr as prel,
  yr as prelease,
  da as press,
  pa as pressBy,
  bc as progNum,
  mr as psus,
  dr as psustain,
  G as pure,
  R as rand,
  kl as rand2,
  Wl as randcat,
  xu as range,
  zu as range2,
  Ou as rangex,
  tf as rarely,
  Fr as rate,
  Mu as ratio,
  oo as rdim,
  Ko as real,
  rl as ref,
  l as register,
  c as registerControl,
  m as reify,
  bn as rel,
  wn as release,
  Y as removeUndefineds,
  Ba as repeatCycles,
  Jc as repeatTime,
  uf as repl,
  Fs as resonance,
  Je as rev,
  io as rfade,
  $a as ribbon,
  Fo as ring,
  Go as ringdf,
  Vo as ringf,
  so as rlp,
  eo as room,
  ro as roomdim,
  co as roomfade,
  no as roomlp,
  lo as roomsize,
  ue as rotate,
  Au as round,
  po as rsize,
  vl as run,
  Ie as s,
  Ha as s_add,
  Ia as s_alt,
  $ as s_cat,
  Ga as s_contract,
  Va as s_expand,
  Pt as s_polymeter,
  Re as s_polymeterSteps,
  Fa as s_sub,
  Qa as s_taper,
  Da as s_taperlist,
  Ua as s_tour,
  bt as saw,
  ti as saw2,
  nc as scram,
  _c as seconds,
  oa as segment,
  Gr as semitone,
  je as seq,
  I as sequence,
  Zi as set,
  Fi as setStringParser,
  qi as setTime,
  mo as shape,
  W as signal,
  A as silence,
  ei as sine,
  Ft as sine2,
  fo as size,
  We as slice,
  Vr as slide,
  Vu as slow,
  za as slowChunk,
  D as slowcat,
  $t as slowcatPrime,
  Oa as slowchunk,
  ec as smear,
  Ri as sol2note,
  Yl as someCycles,
  Xl as someCyclesBy,
  Kl as sometimes,
  Ul as sometimesBy,
  vc as songPtr,
  He as sound,
  Fe as source,
  Gu as sparsity,
  of as speak,
  qo as speed,
  Ya as splice,
  Et as splitAt,
  rr as spread,
  ni as square,
  dl as square2,
  Rl as squeeze,
  To as squiz,
  Ve as src,
  x as stack,
  Qi as stackBy,
  ze as stackCentre,
  xe as stackLeft,
  Oe as stackRight,
  fl as steady,
  Wa as steps,
  Mr as stepsPerOctave,
  Xa as striate,
  Xi as struct,
  Aa as stut,
  va as stutWith,
  ka as stutwith,
  su as sub,
  Yi as superimpose,
  gn as sus,
  yn as sustain,
  Oo as sustainpedal,
  ia as swing,
  ca as swingBy,
  ho as sz,
  si as time,
  Pe as timeCat,
  La as timecat,
  Su as toBipolar,
  ee as tokenizeNote,
  Mo as tremdp,
  Wc as tremolo,
  zo as tremolodepth,
  jo as tremolorate,
  Eo as tremr,
  ml as tri,
  yl as tri2,
  Ro as triode,
  Uo as tsdelay,
  kc as uid,
  Ql as undegrade,
  Dl as undegradeBy,
  _e as unicodeToBase64,
  Pi as uniq,
  Wi as uniqsort,
  be as uniqsortr,
  sr as unison,
  Ao as unit,
  Js as v,
  qc as val,
  xi as valueToMidi,
  Ue as velocity,
  js as vib,
  $s as vibmod,
  Es as vibrato,
  Rs as vmod,
  Dr as voice,
  Co as vowel,
  So as waveloss,
  Ll as wchoose,
  mi as wchooseCycles,
  la as when,
  Il as wrandcat,
  Le as xfade,
  Qo as xsdelay,
  Rc as zcrush,
  Pc as zdelay,
  Jt as zipWith,
  $c as zmod,
  Nc as znoise,
  ea as zoom,
  na as zoomArc,
  sa as zoomarc,
  xc as zrand,
  Lc as zzfx
};
